---
title: "Extended library EPIC"
author: "Lucas A Salas"
date: "August 28, 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

workdir<-"I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data"

#library(S4Vectors)#minfi is not calling the program automatically
library(sesame)
library(minfi)
library(ENmix)
library(pheatmap)
library(RColorBrewer)
library(ggplot2)
library(scatterpie)
library(reshape2)
library(gridExtra)
library(reshape)
library(FlowSorted.Blood.EPIC)
library(FlowSorted.BloodExtended.EPIC)
library(limma)
library(BiocParallel)
library(doFuture)
library(BiocParallel.FutureParam)
library(future.apply)
library(wateRmelon)
library(IDOL)
```

## R Markdown

Generating the basic objects here a combination of minfi and sesame were used

```{r preprocess idats, eval=FALSE, include=FALSE}
sheet <- read.metharray.sheet(base = "I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data", pattern = "phenotype_extended_base2.csv")
workdir<-"I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data"
epicData <- read.metharray.exp(targets = sheet,extended = TRUE, force = TRUE)
MsetEx<-preprocessNoob(epicData)
save(epicData, MsetEx, file= "Raw_data_mix.RData")
#sesame is used to 
load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/Raw_data_mix.RData")
    IDATprefixes <- epicData$filenames#searchIDATprefixes(getwd(), recursive=TRUE)
    names(IDATprefixes)<-colnames(epicData)
    register(FutureParam())
    cl <- parallel::makeCluster(detectCores(), type = "SOCK")
    plan(cluster, workers = cl,.cleanup = TRUE)
    p<-FutureParam()
    tmp<-openSesame(IDATprefixes,
                    mask = TRUE,
                    sum.TypeI = TRUE, 
                    platform = "EPIC", 
                    BPPARAM = p)
    parallel::stopCluster(cl)
    plan(sequential)
    gc()
    closeAllConnections()
    register(FutureParam())
    cl <- parallel::makeCluster(detectCores(), type = "SOCK")
    plan(cluster, workers = cl,.cleanup = TRUE)
    p<-FutureParam()
    tmp2 <- openSesame(IDATprefixes,
                       mask = TRUE,
                    sum.TypeI = TRUE, 
                    platform = "EPIC", 
                    BPPARAM = p,
                    what="sigset")
    parallel::stopCluster(cl)
    plan(sequential)
    gc()
    closeAllConnections()
    register(FutureParam())
    cl <- parallel::makeCluster(detectCores(), type = "SOCK")
    plan(cluster, workers = cl,.cleanup = TRUE)
    p<-FutureParam()
    tmp3<-SigSetsToRGChannelSet(tmp2, BPPARAM = p)
    parallel::stopCluster(cl)
    plan(sequential)
    gc()
    closeAllConnections()
    print(table(sheet$CellType))
    pheno<-as.data.frame(pData(epicData))
    

    qc<-foreach(j = 1:length(names(tmp2))) %do% {#
      sesameQC(tmp2[[j]], betas = tmp[,j])
    }

    names(qc)<-names(tmp2)
    df1<-as.data.frame(data.table::rbindlist(qc))
    rownames(df1)<-names(tmp2)
    df1<-df1[rownames(df1)%in%rownames(pheno),]
    df1<-df1[rownames(pheno),]
    identical(rownames(pheno), rownames(df1))
    pheno<-cbind.data.frame(pheno, df1)
    tmp<-tmp[,colnames(tmp)%in%rownames(pheno)]
    tmp<-tmp[,rownames(pheno)]
    identical(rownames(pheno), colnames(tmp))
    tmp3<-tmp3[,colnames(tmp3)%in%rownames(pheno)]
    tmp3<-tmp3[,rownames(pheno)]
    identical(rownames(pheno), colnames(tmp3))
    pheno$karyotype<-NA
    karyotype<-foreach(k = 1:length(names(tmp2))) %do% {
      inferSexKaryotypes(tmp2[[k]])
    }
    pheno$karyotype<-unlist(karyotype)
    print(table(pheno$karyotype, pheno$Sex))
    boxplot(pheno$frac_na, main= paste("Fraction of masked probes"))
    pData(tmp3)<-DataFrame(pheno, row.names = rownames(pheno))
    #extract OOB p-values
    pvalsesame<-matrix(data = NA,nrow = length(pOOBAH(tmp2[[1]], 
                                                      return.pval = TRUE)),
                       ncol=length(rownames(pheno)), 
                       dimnames = list(names(pOOBAH(tmp2[[1]], 
                                                      return.pval = TRUE)),
                                       rownames(pheno)))
    for ( l in rownames(pheno)){
      pvalsesame[,l]<-pOOBAH(tmp2[[l]], return.pval = TRUE)
      }
    pvalsesame[1:5,1:5]
    pvalsesame<-pvalsesame[,rownames(pheno)]
    identical (rownames(pheno), colnames(pvalsesame))
    #calculating CNV segments
    ssets.normal <- sesameDataGet('EPIC.5.normal')
    segmentmeth<-function(m,sset, sset.normal, refversion ="hg38"){
      seg <- cnSegmentation(sset[[m]], ssets.normal, refversion = refversion)
      seg$seg.signals$Sample_ID<-names(sset)[[m]]
      seg2<-list(seg$seg.signals)
      names(seg2)<-m
      return(seg2)
    }
    
    seg0<-foreach(m = 1:length(names(tmp2))) %do% {
      segmentmeth(m, tmp2, ssets.normal, refversion = "hg38")
    }
    for (i  in 1: length(names(tmp2))){
      if (i==1){
        seg<-as.data.frame(seg0[[i]])
        colnames(seg)<-c("ID", "chrom", "loc.start", "loc.end", "num.mark", 
"seg.mean", "seg.sd", "seg.median", "seg.mad", "pval", 
"lcl", "ucl", "Sample_ID")
      } else{
        seg1<-as.data.frame(seg0[[i]])
        colnames(seg1)<-c("ID", "chrom", "loc.start", "loc.end", "num.mark", 
"seg.mean", "seg.sd", "seg.median", "seg.mad", "pval", 
"lcl", "ucl", "Sample_ID")
        seg<-rbind.data.frame(seg,seg1)
      }
    }
    seg<-as.data.frame(data.table::rbindlist(list(seg0)))
    
    
    # # #ExtraSNPs using Type I probes
    tmp1<-  getAFTypeIbySumAlleles(tmp2[[1]])
    extraSNPAFs <- matrix(data = NA,nrow = length(tmp1),
                   ncol=length(rownames(pheno)), dimnames = list(names(tmp1),
                                                           rownames(pheno)))
    for ( n in rownames(pheno)){
      extraSNPAFs[,n]<-getAFTypeIbySumAlleles(tmp2[[n]])
      }
i="Ext_deconv"
    assign(x = paste0(i, "_betas"),value = tmp)
    assign(x = paste0(i, "_rgset"),value = tmp3)
    assign(x = paste0(i, "_seg"),value = seg)
    assign(x = paste0(i, "_pOOBHA"),value = pvalsesame)
    assign(x = paste0(i, "_SNP"),value = extraSNPAFs)
    assign(x = paste0(i, "_pheno"),value = pheno)
    save(list = c(paste0(i, "_betas"),
                  paste0(i, "_rgset"),
                  paste0(i, "_seg"),
                  paste0(i, "_pOOBHA"),
                  paste0(i, "_SNP"),
                  paste0(i, "_pheno")), file= paste0(i,"_sesame.RData"))
    rm(list = ls()[!ls()%in%c(paste0(i, "_betas"),
                  paste0(i, "_rgset"),
                  paste0(i, "_seg"),
                  paste0(i, "_pOOBHA"),
                  paste0(i, "_SNP"),
                  paste0(i, "_pheno"))])

```

```{r exclude problematic probes, fig.height=11, fig.width=11, cache=FALSE, eval=TRUE, echo=TRUE, include=TRUE}
load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/Ext_deconv_sesame.RData")
annot<-as.data.frame(readRDS("D:/OneDrive/OneDrive - Dartmouth College/Crossreactive/EPIC.hg19.manifest.rds"))
#head(annot)

annot<-annot[rownames(annot)%in% rownames(Ext_deconv_betas),]

#filter crossreactive, problematic probes, X, Y and ch
Ext_deconv_betas2<-Ext_deconv_betas[rownames(Ext_deconv_betas)%in%rownames(annot[annot$seqnames!="chrY" & 
                                            annot$seqnames!="chrX" & 
                                            annot$probeType=="cg" &
                                            annot$MASK_general==FALSE,]),]

dim(Ext_deconv_betas2)

Ext_deconv_betas3<-Ext_deconv_betas2[complete.cases(Ext_deconv_betas2),]

dim(Ext_deconv_betas3)

```

Modified functions of ENmix, sesame, and IDOL are used

```{r basic functions, echo=FALSE,include=FALSE}
plotCtrl2<-function (rgSet, IDorder = NULL) 
{
    if (!is(rgSet, "rgDataSet") & !is(rgSet, "RGChannelSet")) {
        stop("object needs to be of class 'rgDataSet' or 'RGChannelSet'")
    }
    if (!is.null(IDorder)) {
        if (sum(!(IDorder %in% colnames(rgSet))) > 0) {
            idmissing = IDorder[!(IDorder %in% colnames(rgSet))]
            cat("The IDs were not found in the input data: ", 
                idmissing, "\n")
            stop("Wrong ids in IDorder, please check")
        }
        else {
            rgSet = rgSet[, IDorder]
        }
    }
    if (is(rgSet, "rgDataSet")) {
        ctrls <- getCGinfo(rgSet, type = "ctrl")
    }
    else if (is(rgSet, "RGChannelSet")) {
        ctrls <- getProbeInfo(rgSet, type = "Control")
    }
    ctrls <- ctrls[ctrls$Address %in% rownames(rgSet), ]
    ctrl_r <- assays(rgSet)$Red[ctrls$Address, ]
    ctrl_g <- assays(rgSet)$Green[ctrls$Address, ]
    contlid <- c("STAINING", "EXTENSION", "HYBRIDIZATION", 
                 "TARGET REMOVAL", "BISULFITE CONVERSION I", 
                 "BISULFITE CONVERSION II", "SPECIFICITY I", 
                 "SPECIFICITY II", "NON-POLYMORPHIC", "NEGATIVE", 
                 "NORM_A", "NORM_C", "NORM_G", "NORM_T", 
                 "NORM_ACGT")
    col <- as.vector(ctrls$Color)
    col[col == "-99"] <- NA
    col[col == "Aqua"] <- "aquamarine2"
    col[col == "Crimson"] <- "firebrick2"
    col[col == "Fuchsia"] <- "deeppink1"
    col[col == "Indigo"] <- "darkviolet"
    col[col == "Lime"] <- "yellowgreen"
    col[col == "Olive"] <- "darkolivegreen"
    col[col == "Silver"] <- "azure4"
    col[col == "Teal"] <- "cyan4"
    ctrls$Color <- col
    for (ctype in contlid) {
        fn <- ctype
        fn <- gsub(" ", "_", fn)
        cat("Plotting ", fn, ".jpg", "\n")
        par(mfrow = c(1, 2))
        if (ctype == "NORM_ACGT") {
            cc <- ctrls[ctrls$Type %in% c("NORM_A", "NORM_C", 
                                          "NORM_G", "NORM_T"), ]
        }
        else {
            cc <- ctrls[ctrls$Type %in% ctype, ]
        }
        red <- ctrl_r[cc$Address, ]
        grn <- ctrl_g[cc$Address, ]
        ymax <- max(red, grn) * 1.01
        if (ctype == "NEGATIVE") {
            par(mar = c(5, 4, 4, 2))
            colnames(red) <- 1:ncol(red)
            colnames(grn) <- 1:ncol(grn)
            boxplot(grn, ylim = c(0, ymax), main = paste(ctype, 
                                                         " Green", sep = ""), bty = "o", 
                    xlab = "Sample", ylab = "Intensity", 
                    cex.lab = 1.2)
            boxplot(red, ylim = c(0, ymax), main = paste(ctype, 
                                                         " Red", sep = ""), bty = "o", 
                    xlab = "Sample", ylab = "Intensity", 
                    cex.lab = 1.2)
        }
        else if (ctype %in% c("NORM_A", "NORM_C", 
                              "NORM_G", "NORM_T")) {
            par(mar = c(5, 4, 4, 1))
            colnames(red) <- 1:ncol(red)
            colnames(grn) <- 1:ncol(grn)
            boxplot(grn, ylim = c(0, ymax), col = unique(as.vector(cc$Color)), 
                    main = paste(ctype, " Green", sep = ""), 
                    bty = "o", xlab = "Sample", ylab = "Intensity", 
                    cex.lab = 1.2)
            boxplot(red, ylim = c(0, ymax), col = unique(as.vector(cc$Color)), 
                    main = paste(ctype, " Red", sep = ""), 
                    bty = "o", xlab = "Sample", ylab = "Intensity", 
                    cex.lab = 1.2)
        }
        else if (ctype == "NORM_ACGT") {
            par(mar = c(5, 4, 4, 8.2))
            label <- c("NORM_A", "NORM_C", "NORM_G", 
                       "NORM_T")
            colcode <- c("Red", "Green", "Blue", 
                         "Purple")
            idx <- t(replicate(nrow(red), 1:ncol(red)))
            loc <- cc$Color
            loc[loc == "Red"] <- -0.2
            loc[loc == "Green"] <- -0.1
            loc[loc == "Blue"] <- 0.1
            loc[loc == "Purple"] = 0.2
            loc = as.numeric(loc)
            loc1 <- matrix(rep(loc, ncol(red)), ncol = ncol(red))
            idx = idx + loc1
            plot(idx, grn, col = as.vector(cc$Color), ylim = c(0, 
                                                               ymax), main = paste(ctype, " Green", sep = ""), 
                 bty = "o", xlab = "Sample", ylab = "Intensity", 
                 cex.lab = 1.2)
            legend(x = (ncol(grn) + 0.2) * 1.05, y = ymax * 0.8, 
                   xjust = 0, yjust = 0.5, bty = "o", legend = label, 
                   col = colcode, xpd = TRUE, pch = 15, cex = 0.8)
            plot(idx, red, col = as.vector(cc$Color), ylim = c(0, 
                                                               ymax), main = paste(ctype, " Red", sep = ""), 
                 bty = "o", xlab = "Sample", ylab = "Intensity", 
                 cex.lab = 1.2)
            legend(x = (ncol(red) + 0.2) * 1.05, y = ymax * 0.8, 
                   xjust = 0, yjust = 0.5, bty = "o", legend = label, 
                   col = colcode, xpd = TRUE, pch = 15, cex = 0.8)
        }
        else {
            par(mar = c(5, 4, 4, 8.2))
            idx <- t(replicate(nrow(red), 1:ncol(red)))
            plot(idx, grn, col = as.vector(cc$Color), ylim = c(0, 
                                                               ymax), main = paste(ctype, " Green", sep = ""), 
                 bty = "o", xlab = "Sample", ylab = "Intensity", 
                 cex.lab = 1.2)
            legend(x = ncol(grn) * 1.05, y = ymax * 0.8, xjust = 0, 
                   yjust = 0.5, bty = "o", legend = as.vector(cc$ExtendedType), 
                   col = as.vector(cc$Color), xpd = TRUE, pch = 15, 
                   cex = 0.8)
            plot(idx, red, col = as.vector(cc$Color), ylim = c(0, 
                                                               ymax), main = paste(ctype, " Red", sep = ""), 
                 bty = "o", xlab = "Sample", ylab = "Intensity", 
                 cex.lab = 1.2)
            legend(x = ncol(red) * 1.045, y = ymax * 0.8, xjust = 0, 
                   yjust = 0.5, bty = "o", legend = as.vector(cc$ExtendedType), 
                   col = as.vector(cc$Color), xpd = TRUE, pch = 15, 
                   cex = 0.8)
        }
        
    }
}
QCinfo3<-function (rgSet, detPthre = 1e-06, detPtype = "negative", detpval=NULL,
    nbthre = 3, samplethre = 0.05, CpGthre = 0.05, bisulthre = NULL, 
    outlier = TRUE, distplot = TRUE) 
{
    if (!is(rgSet, "rgDataSet") & !is(rgSet, "RGChannelSetExtended") & !is(rgSet, "RGChannelSet")) 
        stop("[QCinfo] The input should be an object of 'rgDataSet' or 'RGChannelSetExtended' or 'RGChannelSet'")
    if (is(rgSet, "rgDataSet")) {
        cginfo = getCGinfo(rgSet)
        typeI <- cginfo[cginfo$Infinium_Design_Type %in% c("I", 
            "snpI"), ]
        typeIred = typeI[typeI$Color_Channel == "Red", 
            ]
        typeIgrn = typeI[typeI$Color_Channel == "Grn", 
            ]
        typeII <- cginfo[cginfo$Infinium_Design_Type %in% c("II", 
            "snpII"), ]
        locusNames = c(typeIred$Name, typeIgrn$Name, typeII$Name)
        if(!is.null(detpval)){
          detP<-get(detpval)
        }else{
          detP <- calcdetP(rgSet, detPtype = detPtype)
        }
        ctrls <- metadata(rgSet)$ictrl
    }
    else if (is(rgSet, "RGChannelSetExtended") | is(rgSet, "RGChannelSet")) {
        typeI <- getProbeInfo(rgSet, type = "I")
        typeII <- getProbeInfo(rgSet, type = "II")
        locusNames <- getManifestInfo(rgSet, "locusNames")
        if(!is.null(detpval)){
          detP<-get(detpval)
        }else{
          detP <- detectionP(rgSet)
        }
        ctrls <- getProbeInfo(rgSet, type = "Control")
    }
  if (is(rgSet, "RGChannelSetExtended")){
    bc_I <- assays(rgSet)$NBeads[typeI$AddressA, ]
    flag <- bc_I > assays(rgSet)$NBeads[typeI$AddressB, ]
    bc_I[flag] <- assays(rgSet)$NBeads[typeI$AddressB, ][flag]
    nbead <- matrix(NA_real_, ncol = ncol(rgSet), nrow = length(locusNames), 
        dimnames = list(locusNames, colnames(rgSet)))
    nbead[typeI$Name, ] <- bc_I
    nbead[typeII$Name, ] <- assays(rgSet)$NBeads[typeII$AddressA, 
        ]
    rm(list = c("bc_I", "flag"))
    if (!identical(rownames(detP), rownames(nbead))) {
        idx = intersect(rownames(detP), rownames(nbead))
        detP = detP[idx, ]
        nbead = nbead[idx, ]
    }
    if (!identical(colnames(detP), colnames(nbead))) {
        idx = intersect(colnames(detP), colnames(nbead))
        detP = detP[, idx]
        nbead = nbead[, idx]
    }
  } else{
    nbead=NULL
  }
    ctrls = ctrls[ctrls$Address %in% rownames(rgSet), ]
    ctrl_r <- assays(rgSet)$Red[ctrls$Address, ]
    ctrl_g <- assays(rgSet)$Green[ctrls$Address, ]
    cc = ctrls[(ctrls$Type %in% c("BISULFITE CONVERSION I")) & 
        (ctrls$ExtendedType %in% c("BS Conversion I C1", 
            "BS Conversion I-C2", "BS Conversion I-C3")), 
        ]
    I_green = colMeans(ctrl_g[cc$Address, ])
    cc = ctrls[(ctrls$Type %in% c("BISULFITE CONVERSION I")) & 
        (ctrls$ExtendedType %in% c("BS Conversion I-C4", 
            "BS Conversion I-C5", "BS Conversion I-C6")), 
        ]
    I_red = colMeans(ctrl_r[cc$Address, ])
    cc = ctrls[ctrls$Type %in% c("BISULFITE CONVERSION II"), 
        ]
    II_red = colMeans(ctrl_r[cc$Address, ])
    bisul = (I_green + I_red + II_red)/3
    if (is.null(bisulthre)) {
        bisulthre = mean(bisul, na.rm = TRUE) - 3 * sd(bisul, 
            na.rm = TRUE)
    }
    if (is(rgSet, "RGChannelSetExtended")){
    qcmat <- nbead < nbthre | detP > detPthre
    badValuePerSample <- apply(qcmat, 2, sum)/nrow(qcmat)
    flag <- badValuePerSample > samplethre | bisul < bisulthre
    cat(sum(flag), " samples with percentage of low quality CpG value greater than ", 
        samplethre, " or bisulfite intensity less than ", 
        bisulthre, "\n")
    badsample = colnames(qcmat)[flag]
    qcmat <- qcmat[, !flag]
    NbadValuePerCpG <- apply(qcmat, 1, sum)
    badValuePerCpG <- NbadValuePerCpG/ncol(qcmat)
    flag2 <- badValuePerCpG > CpGthre & NbadValuePerCpG > 1
    cat(sum(flag2), " CpGs with percentage of low quality value greater than ", 
        CpGthre, "\n")
    badCpG <- rownames(qcmat)[flag2]
    qcmat = qcmat[!flag2, ]
    } else{
      qcmat <- detP > detPthre
    badValuePerSample <- apply(qcmat, 2, sum)/nrow(qcmat)
    flag <- badValuePerSample > samplethre | bisul < bisulthre
    cat(sum(flag), " samples with percentage of low quality CpG value greater than ", 
        samplethre, " or bisulfite intensity less than ", 
        bisulthre, "\n")
    badsample = colnames(qcmat)[flag]
    qcmat <- qcmat[, !flag]
    NbadValuePerCpG <- apply(qcmat, 1, sum)
    badValuePerCpG <- NbadValuePerCpG/ncol(qcmat)
    flag2 <- badValuePerCpG > CpGthre & NbadValuePerCpG > 1
    cat(sum(flag2), " CpGs with percentage of low quality value greater than ", 
        CpGthre, "\n")
    badCpG <- rownames(qcmat)[flag2]
    qcmat = qcmat[!flag2, ]
    }
    cat("Plotting qc_sample.jpg ...")
    color = rep("black", length(badValuePerSample))
    color[flag] = "red"
    plot(badValuePerSample, bisul, xlab = "Percent of low quality data per sample", 
        ylab = "Average bisulfite conversion intensity", 
        cex = 1.5, col = color, main = paste(length(badsample), 
            " samples were classified as low quality samples"))
    abline(h = bisulthre, lty = 2, col = "red")
    abline(v = samplethre, lty = 2, col = "red")
    dev.off()
    cat("Done\n")
    cat("Plotting qc_CpG.jpg ...")
    par(mfrow = c(2, 1))
    hist(badValuePerCpG, breaks = 1000, xlab = "Percent of low quality data per CpG", 
        main = paste(length(badCpG), " CpGs were classified as low quality CpGs"))
    abline(v = CpGthre, lty = 2, col = "red")
    hist(badValuePerCpG, breaks = 1000, xlim = c(0, 0.1), xlab = "Percent of low quality data per CpG", 
        main = "Zoom in view")
    abline(v = CpGthre, lty = 2, col = "red")
    dev.off()
    cat("Done\n")
    if (outlier) {
        cat("Identifying outlier samples based on beta or total intensity values...\n")
        if (is(rgSet, "rgDataSet")) {
            mdat = getmeth(rgSet)
        }
        else if (is(rgSet, "RGChannelSetExtended") | is(rgSet, "RGChannelSet")) {
            mdat = preprocessRaw(rgSet)
        }
        mdat = mdat[rownames(qcmat), ]
        mdat = mdat[, colnames(qcmat)]
        mu <- assays(mdat)$Meth + assays(mdat)$Unmeth
        mumean = apply(mu, 2, mean, na.rm = TRUE)
        q2575 <- quantile(mumean, probs = c(0.25, 0.75), na.rm = TRUE)
        qr <- q2575["75%"] - q2575["25%"]
        low = q2575["25%"] - 3 * qr
        flag1 = mumean < low
        cat("After excluding low quality samples and CpGs\n")
        cat(sum(flag1), " samples are outliers based on averaged total intensity value", 
            "\n")
        beta = getB(mdat, type = "Illumina")
        qq = apply(beta, 2, function(x) quantile(x, probs = c(0.25, 
            0.5, 0.75), na.rm = TRUE))
        q2575 <- apply(qq, 1, function(x) quantile(x, probs = c(0.25, 
            0.75), na.rm = TRUE))
        qr <- q2575["75%", ] - q2575["25%", ]
        up = q2575["75%", ] + 3 * qr
        low = q2575["25%", ] - 3 * qr
        flag = qq > up | qq < low
        flag = apply(flag, 2, sum) > 0
        cat(sum(flag), " samples are outliers in beta value distribution", 
            "\n")
        flag = flag | flag1
        badsample = c(badsample, colnames(beta)[flag])
        outlier_sample = colnames(beta)[flag]
        cat(sum(flag), " outlier samples were added into badsample list\n")
        if (ncol(qcmat) < 15) {
            cat("WARNING: Sample size may be too small to correctly identify outlier samples!\n")
            cat("RECOMMENDATION: set outlier=FALSE or double check total intensity and beta value \n     distribution plots to confirm\n")
        }
    }
    if (distplot) {
        if (is(rgSet, "rgDataSet")) {
            mdat = getmeth(rgSet)
        }
        else if (is(rgSet, "RGChannelSetExtended") | is(rgSet, "RGChannelSet")) {
            mdat = preprocessRaw(rgSet)
        }
        beta = getB(mdat, type = "Illumina")
        cat("Plotting freqpolygon_beta_beforeQC.jpg ...")
        
        color = rep("black", ncol(beta))
        color[colnames(beta) %in% badsample] = "red"
        multifreqpoly(beta, cex.lab = 1.4, cex.axis = 1.5, col = color, 
            legend = "", cex.main = 1.5, main = "Beta value distribution", 
            xlab = "Methylation beta value")
        #dev.off()
        cat("Done\n")
        beta = beta[!(rownames(beta) %in% badCpG), ]
        beta = beta[, !(colnames(beta) %in% badsample)]
        cat("Plotting freqpolygon_beta_afterQC.jpg ...")
        multifreqpoly(beta, cex.lab = 1.4, cex.axis = 1.5, col = "black", 
            legend = "", cex.main = 1.5, main = "Beta value distribution", 
            xlab = "Methylation beta value")
        #dev.off()
        cat("Done\n")
    }
    if (outlier) {
        list(detP = detP, nbead = nbead, 
             bisul = bisul, badsample = badsample, 
            badCpG = badCpG, outlier_sample = outlier_sample)
    }
    else {
        list(detP = detP, nbead = nbead, 
             bisul = bisul, badsample = badsample, 
            badCpG = badCpG)
    }
}
    

multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
    library(grid)
    
    # Make a list from the ... arguments and plotlist
    plots <- c(list(...), plotlist)
    
    numPlots = length(plots)
    
    # If layout is NULL, then use 'cols' to determine layout
    if (is.null(layout)) {
        # Make the panel
        # ncol: Number of columns of plots
        # nrow: Number of rows needed, calculated from # of cols
        layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                         ncol = cols, nrow = ceiling(numPlots/cols))
    }
    
    if (numPlots==1) {
        print(plots[[1]])
        
    } else {
        # Set up the page
        grid.newpage()
        pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))
        
        # Make each plot, in the correct location
        for (i in 1:numPlots) {
            # Get the i,j matrix positions of the regions that contain this subplot
            matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))
            
            print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                            layout.pos.col = matchidx$col))
        }
    }
}

lmmatrix1<-function (y, cov) {
    p <- matrix(NA, ncol = ncol(cov), nrow = ncol(y))
    for (j in 1:ncol(cov)) {
        x <- cov[, j]
        for (i in 1:ncol(y)) {
            fit <- summary(lm(y[, i] ~ x, na.action = na.omit))
            f <- fit$fstatistic
            p[i, j] <- pf(f["value"], f["numdf"], 
                f["dendf"], lower.tail = FALSE)
        }
    }
    colnames(p) <- names(cov)
    p
}
pcrplot2<-function (beta, cov, npc = 50) 
{
    if (!is.matrix(beta)) {
        stop("beta is not a data matirx")
    }
    if (!is.data.frame(cov)) {
        stop("cov is not a data frame")
    }
    if (ncol(beta) != nrow(cov)) {
        stop("number of columns in beta is not equal to number of rows in cov")
    }
    cat("Analysis is running, please wait...!", "\n")
    npc <- min(ncol(beta), npc)
    svd <- prcomp(t(beta), center = TRUE, scale = TRUE, retx = TRUE)#, 
        #na.action = "na.omit")
    screeplot(svd, npc, type = "barplot")
    eigenvalue <- svd[["sdev"]]^2
    prop <- (sum(eigenvalue[1:npc])/sum(eigenvalue)) * 100
    cat("Top ", npc, " principal components can explain ", 
        prop, "% of data \n    variation", "\n")
    prop <- (sum(eigenvalue[1:1])/sum(eigenvalue)) * 100
    cat("Top ", "1", " principal components can explain ", 
        prop, "% of data \n    variation", "\n")
    prop <- (sum(eigenvalue[1:2])/sum(eigenvalue)) * 100
    cat("Top ", "2", " principal components can explain ", 
        prop, "% of data \n    variation", "\n")
    prop <- (sum(eigenvalue[1:3])/sum(eigenvalue)) * 100
    cat("Top ", "3", " principal components can explain ", 
        prop, "% of data \n    variation", "\n")
    prop <- (sum(eigenvalue[1:4])/sum(eigenvalue)) * 100
    cat("Top ", "4", " principal components can explain ", 
        prop, "% of data \n    variation", "\n")
    
    p <- lmmatrix1(svd$x[, 1:npc], cov)
    yaxis <- colnames(p)
    title = "Principal Component Regression Analysis"
    xmax=npc
    plot(1, xlim = c(0, xmax), ylim = c(0, length(yaxis) + 1), 
        type = "n", bty = "n", axes = FALSE, xlab = "Principal Component", 
        ylab = "", main = title)
    axis(1, at = c(1:xmax), pos = 0.5, las = 1, lwd = 3)
    for (i in 1:length(yaxis)) {
        text(0.3, i, yaxis[i], xpd = TRUE, adj = 1)
    }
    for (i in 1:ncol(p)) {
        for (j in 1:nrow(p)) {
            pp <- p[j, i]
            colcode <- "white"
            if (pp <= 1e-09) {
                colcode = "darkred"
            }
            else if (pp <= 1e-04) {
                colcode = "red"
            }
            else if (pp <= 0.01) {
                colcode = "orange"
            }
            else if (pp <= 0.05) {
                colcode = "pink"
            }
            polygon(c(j - 0.5, j - 0.5, j + 0.5, j + 0.5), c(i - 
                0.5, i + 0.5, i + 0.5, i - 0.5), col = colcode, 
                border = NA)
        }
    }
    legend("topright", c("<0.05", "<0.01", 
        "<10E-5", "<10E-10"), col = c("pink", 
        "orange", "red", "darkred"), pch = 15, 
        pt.cex = 2, bty = "o", horiz = TRUE, xpd = TRUE)
}



```

IDOL baseline functions were optimized using future parameters

```{r IDOL, fig.height=11, fig.width=11, eval=TRUE, include=FALSE}
library(IDOL)
IDOLoptimize2<-function (candDMRFinderObject, trainingBetas, trainingCovariates, 
    libSize = 300, maxIt = 500, numCores = 4) 
{
    require(quadprog)
    require(doFuture)
    expit = function(w) exp(w)/(1 + exp(w))
    logit = function(w) log(w) - log(1 - w)
    R2compute = function(obs, pred) {
        r2 = NULL
        for (i in 1:dim(obs)[2]) {
            y = obs[, i]
            x = pred[, i]
            r2[i] = summary(lm(y ~ x))[[8]]
        }
        sum(r2)/dim(obs)[2]
    }
    polar = function(x, y, scale = 1) {
        r = sqrt(x^2 + y^2)
        theta = atan2(y, x)
        r * cos(theta - (scale * pi/4))
    }
    trainingProbes1 = candDMRFinderObject$candidateSet
    coefEsts = candDMRFinderObject$coefEsts
    P = length(trainingProbes1)
    ProbVector = rep(1/P, P)
    V = libSize
    B = maxIt
    R2.null = 0
    R2Vals = NULL
    RMSE.null = 10000
    RMSEVals = NULL
    cellTypes = colnames(coefEsts)
    K = length(cellTypes)
    if (sum(cellTypes %in% colnames(trainingCovariates)) != K) {
        stop("cell type names in target covariate data are not consistent with cell types to be deconvoluted")
    }
    cl <- makeCluster(numCores, type = "SOCK")
    registerDoFuture()
    plan(multiprocess)#, workers = cl,.cleanup = TRUE)
    for (i in 1:B) {
        Probes = sample(1:P, V, prob = ProbVector)
        CpGNames = trainingProbes1[Probes]
        Beta = coefEsts[CpGNames, ]
        Lwbc = diag(ncol(Beta))
        ctpred = data.frame(minfi:::projectCellType(trainingBetas[CpGNames, 
            ], Beta))
        omega.tilde = 100 * ctpred
        omega.obs = trainingCovariates[, cellTypes]
        diff = as.matrix(omega.tilde - omega.obs)
        RMSE = sqrt(sum(diff^2)/K)
        R2 = R2compute(omega.obs, omega.tilde)
        Perform.q = foreach(j = 1:length(CpGNames)) %dopar% {
            Beta.q = Beta[CpGNames[-j], ]
            ctpred.q = data.frame(minfi:::projectCellType(trainingBetas[CpGNames[-j], 
                ], Beta.q, Lwbc))
            omega.tilde.q = 100 * ctpred.q
            diff.q = as.matrix(omega.tilde.q - omega.obs)
            RMSE.q = sqrt(sum(diff.q^2)/K)
            R2.q = R2compute(omega.obs, omega.tilde.q)
            cbind(R2.q, RMSE.q)
        }
        R2.q = unlist(Perform.q)[seq(from = 1, to = (2 * V - 
            1), by = 2)]
        RMSE.q = unlist(Perform.q)[seq(from = 2, to = 2 * V, 
            by = 2)]
        rmse.dq = (RMSE - RMSE.q) * (-1)
        norm.rmse = (rmse.dq)/sd(rmse.dq)
        r2.dq = (R2 - R2.q)
        norm.r2 = (r2.dq)/sd(r2.dq)
        p1 = polar(norm.rmse, norm.r2)
        for (j in 1:length(Probes)) {
            p0 = ProbVector[[Probes[j]]]
            ProbVector[[Probes[j]]] = expit(p1[j]) * p0 + p0/2
        }
        ProbVector = ProbVector/sum(ProbVector)
        if (RMSE <= RMSE.null & R2 >= R2.null) {
            RMSE.null = RMSE
            R2.null = R2
            print(paste("Iteration: ", i, " RMSE=", 
                round(RMSE, 3), "; R2=", round(R2, 3), 
                sep = ""))
            IDOL.optim.DMRs = CpGNames
            IDOL.optim.coefEsts = coefEsts[CpGNames, ]
            
        }
        RMSEVals[i] = RMSE
        R2Vals[i] = R2
    }
    stopCluster(cl)
    plan(sequential)## Explicitly close multisession workers by switching plan
    gc()
    closeAllConnections()
    IDOLObjects = list(IDOL.optim.DMRs, IDOL.optim.coefEsts, 
        RMSEVals, R2Vals, B, V)
    names(IDOLObjects) = c("IDOL Optimized Library", "IDOL Optimized CoefEsts", 
        "RMSE", "R2", "Number of Iterations", 
        "LibrarySize")
    save(IDOL.optim.DMRs, IDOL.optim.coefEsts, IDOLObjects, file = paste("IDOL optimized DMR library_", 
                V, ".RData", sep = ""))
    print(paste("The Average RMSE = ", round(RMSE.null, 
        3), " and R2 = ", round(R2.null, 3), " for the IDOL Optimized Library", 
        sep = ""))
    return(IDOLObjects)
}

```

\#General QC Are sex and karyotype OK? Are the same subjects clustered together? Are the genetic ancestries corresponding to the broad self-reported ethnicity? "White" subjects are part of the broader Indo-European class. The genetic markers only allow broader categories.

```{r general_QC, fig.height=11, fig.width=11, cache=TRUE, eval=TRUE, cache.lazy = FALSE}
library(FlowSorted.BloodExtended.EPIC)
epicData<-FlowSorted.BloodExtended.EPIC[,FlowSorted.BloodExtended.EPIC$CellType!="MIX"]
table(epicData$CellType)
MsetEx<-preprocessRaw(epicData)#Data has been processed through sesame

GMsetEx <- mapToGenome(MsetEx)
estSex <- getSex(GMsetEx)
GMsetEx <- addSex(GMsetEx, sex = estSex)
plotSex(GMsetEx, id= estSex$predictedSex)
plotSex(GMsetEx, id= GMsetEx$Sex)
table(GMsetEx$Sex,estSex$predictedSex,exclude=NULL)
pheno<-as.data.frame(pData(epicData))
table(pheno$karyotype)


# Define color for each of the 3 iris species
groups<-c( Bas= "#E41A1C", 
           #Bcell= "lightblue", 
           Bnv="#1B9E77", Bmem= "magenta", 
           #CD4T="darkorange", 
           CD4mem="darkblue", CD4nv="#D95F02", 
           #CD8T="pink",
           CD8nv ="#7570B3",CD8mem="maroon", 
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02", 
           Treg="black")#, WBC="gray", MIX="lightgray")
colors <- groups
colors <- colors[epicData$CellType]#if not names use as.numeric to color

# Define shapes
shapes = seq(from=12, to=25, by=1)
shapes <- shapes[as.numeric(as.factor(epicData$CellType))]

# Plot
plot(x = epicData$Age, y = epicData$mage, frame = FALSE,
     xlab = "Age (years)", ylab = "Methylation age",
     col = colors, pch = shapes)
legend("topright", legend = names(groups),
      col =  groups,
      pch = seq(from=12, to=25, by=1) )
abline(0,1)


#Select the subset

betas_sesame<-Ext_deconv_betas[, colnames(Ext_deconv_betas)%in%rownames(pheno)]
identical (colnames(betas_sesame),rownames(pheno))

#All the data
annotation_col=data.frame(Sex= epicData$Sex, mSex= epicData$predictedSex, CellType=epicData$CellType, 
                          Age=epicData$Age, mage=epicData$Horvath_age, Ethnicity= epicData$Ethnicity_wide, 
                          methnicity=epicData$meth_ethnicity,frac_na=epicData$frac_na, purity=epicData$meth_purity,
                          row.names = colnames(epicData))

ann_colors = list(
    CellType = groups, Ethnicity = c("African-American"="black", "East-Asian"="yellow", "Indo-European"="pink", "Mixed"="brown")
) 
pheatmap(getSnpBeta(epicData), annotation_colors = ann_colors,
             #      annotation_row = annotation_row, 
             annotation_col = annotation_col, main = "SNPs", cluster_rows = T, show_rownames = T, labels_col = epicData$CellType)

#Filtered data
SNPcheck(betas_sesame)

probes <- names(sesameDataGet("HM450.probeInfo")$probe2chr.hg19)
snp.probes <- probes[grep("rs", probes)]
snp.probes <- intersect(snp.probes, rownames(betas_sesame))

annotation_col=data.frame(Sex= pheno$Sex, mSex= pheno$predictedSex, CellType=pheno$CellType, 
                          Age=pheno$Age, mage=pheno$Horvath_age, Ethnicity= pheno$Ethnicity_wide, 
                          methnicity=pheno$meth_ethnicity,frac_na=pheno$frac_na, purity=pheno$meth_purity,
                          row.names = rownames(pheno))

pheatmap(betas_sesame[snp.probes, ], annotation_colors = ann_colors,#annotation_row = annotation_row, 
             annotation_col = annotation_col, main = "SNPs", cluster_rows = T, show_rownames = T, labels_col = epicData$CellType)


plot(x = pheno$Age, y = pheno$Horvath_age, frame = FALSE,
     xlab = "Age (years)", ylab = "Methylation age Horvath",
     col = colors, pch = shapes)
legend("topright", legend = names(groups),
      col =  groups,
      pch = seq(from=12, to=25, by=1) )
abline(0,1)

probes <- names(sesameDataGet("HM450.probeInfo")$probe2chr.hg19)
snp.probes <- probes[grep("rs", probes)]
snp.probes <- intersect(snp.probes, rownames(betas_sesame))

allsnps<-rbind(betas_sesame[snp.probes, ], Ext_deconv_SNP[,colnames(betas_sesame)])
pheatmap(allsnps, annotation_colors = ann_colors,#      annotation_row = annotation_row, 
             annotation_col = annotation_col, main = "All potential SNPs", cluster_rows = T, show_rownames = F, labels_col = epicData$CellType)


mdsPlot(allsnps, sampGroups = annotation_col$methnicity)

#SNPs organized
load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/snpall.rda")

allsnps2<-allsnps[rownames(allsnps)%in%rownames(snpall),]
snpall2<-snpall[rownames(snpall)%in%rownames(allsnps),]
snpall2<-snpall2[rownames(allsnps2),]
allsnps2[snpall2$U=="ALT",]<- 1-allsnps2[snpall2$U=="ALT",]
allsnps2<-ifelse(allsnps2<0.2, 0, 
                 ifelse(allsnps2>0.8, 1, 0.5))
allsnps3<-matrix(data=NA, nrow = nrow(allsnps2), ncol = ncol(allsnps2),
                 dimnames = list(rownames(allsnps2), colnames(allsnps2)) )
identical(rownames(allsnps3), rownames(snpall2))
for (i in 1:length(rownames(allsnps3))){
  allsnps3[i,]<-ifelse(allsnps2[i,]==0, paste0(snpall2$REF[i],snpall2$REF[i]),
                       ifelse(allsnps2[i,]==1, paste0(snpall2$ALT[i],snpall2$ALT[i]), 
                              paste0(snpall2$REF[i],snpall2$ALT[i])))
}
annotation_col=data.frame(Sex= pheno$Sex, #mSex= pheno$predictedSex, 
                          CellType=pheno$CellType, 
                          #Age=pheno$Age, mage=pheno$Horvath_age, 
                          Ethnicity= pheno$Ethnicity_wide, 
                          SeSaMe_ethnicity=pheno$meth_ethnicity,#frac_na=pheno$frac_na, purity=pheno$meth_purity,
                          row.names = rownames(pheno))
dput(levels(as.factor(annotation_col$SeSaMe_ethnicity)))

ann_colors = list(
    CellType = groups, Ethnicity = c("African-American"="black", "East-Asian"="yellow", "Indo-European"="pink", "Mixed"="brown"), SeSaMe_ethnicity = c("ASIAN"="yellow", "BLACK OR AFRICAN AMERICAN"="black", "WHITE"="pink")
) 

pheatmap(allsnps2[,epicData$CellType!="MIX"], annotation_colors = ann_colors,#      annotation_row = annotation_row, 
             annotation_col = annotation_col, main = "All potential SNPs", cluster_rows = T, labels_row = snpall2$rs, labels_col = epicData$CellType[epicData$CellType!="MIX"], display_numbers = allsnps3[,epicData$CellType!="MIX"])


#SNPs organized according to genetic context
#load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/snpall.rda")

allsnps2<-getSnpBeta(epicData)
allsnps2<-allsnps2[rownames(allsnps2)%in%rownames(snpall),]
snpall2<-snpall[rownames(snpall)%in%rownames(allsnps2),]
snpall2<-snpall2[rownames(allsnps2),]
allsnps2[snpall2$U=="ALT",]<- 1-allsnps2[snpall2$U=="ALT",]
allsnps2<-ifelse(allsnps2<0.2, 0, 
                 ifelse(allsnps2>0.8, 1, 0.5))
allsnps3<-matrix(data=NA, nrow = nrow(allsnps2), ncol = ncol(allsnps2),
                 dimnames = list(rownames(allsnps2), colnames(allsnps2)) )
identical(rownames(allsnps3), rownames(snpall2))
for (i in 1:length(rownames(allsnps3))){
  allsnps3[i,]<-ifelse(allsnps2[i,]==0, paste0(snpall2$REF[i],snpall2$REF[i]),
                       ifelse(allsnps2[i,]==1, paste0(snpall2$ALT[i],snpall2$ALT[i]), 
                              paste0(snpall2$REF[i],snpall2$ALT[i])))
}

annotation_col=data.frame(Sex= pheno$Sex, #mSex= pheno$predictedSex, 
                          CellType=pheno$CellType, 
                          #Age=pheno$Age, mage=pheno$Horvath_age, 
                          Ethnicity= pheno$Ethnicity_wide, 
                          SeSaMe_ethnicity=pheno$meth_ethnicity,#frac_na=pheno$frac_na, purity=pheno$meth_purity,
                          row.names = rownames(pheno))
dput(levels(as.factor(annotation_col$SeSaMe_ethnicity)))

ann_colors = list(
    CellType = groups, Ethnicity = c("African-American"="black", "East-Asian"="yellow", "Indo-European"="pink", "Mixed"="brown"), SeSaMe_ethnicity = c("ASIAN"="yellow", "BLACK OR AFRICAN AMERICAN"="black", "WHITE"="pink")
) 

pheatmap(allsnps2, annotation_colors = ann_colors,#      annotation_row = annotation_row, 
             annotation_col = annotation_col, main = "Technical SNPs", cluster_rows = T, show_rownames = T, labels_col = epicData$CellType, display_numbers = allsnps3)


#Without filtering only reorganizing the 21 tracking to the minor allele

#SNPs organized
load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/snpall.rda")

allsnps2<-allsnps
snpall2<-snpall
allsnps2[rownames(allsnps)%in%rownames(snpall[snpall$U=="ALT",]),]<- 1-allsnps2[rownames(allsnps)%in%rownames(snpall[snpall$U=="ALT",]),]
allsnps2<-ifelse(allsnps2<0.2, 0, 
                 ifelse(allsnps2>0.8, 1, 0.5))

annotation_col=data.frame(Sex= pheno$Sex,  
                          CellType=pheno$CellType, 
                          Ethnicity= pheno$Ethnicity_wide, 
                          SeSaMe_ethnicity=pheno$meth_ethnicity,
                          row.names = rownames(pheno))
dput(levels(as.factor(annotation_col$SeSaMe_ethnicity)))
annotation_col$SeSaMe_ethnicity<-ifelse(annotation_col$SeSaMe_ethnicity=="ASIAN", "Asian",
                                        ifelse(annotation_col$SeSaMe_ethnicity=="BLACK OR AFRICAN AMERICAN", "Black",
                                               "White"))
ann_colors = list(
    CellType = groups, Ethnicity = c("African-American"="black", "East-Asian"="brown", "Indo-European"="pink", "Mixed"="red"), SeSaMe_ethnicity = c("Asian"="brown", "Black"="black", "White"="pink")
) 

pheatmap(allsnps2[,epicData$CellType!="MIX"], annotation_colors = ann_colors,
         annotation_col = annotation_col, main = "All potential SNPs", 
         cluster_rows = T, show_rownames = F, 
         labels_col = epicData$CellType[epicData$CellType!="MIX"])
```

Are the samples distinct enough to allow separating the cell types?

```{r PCA restricted, fig.height=11, fig.width=11, cache= TRUE}

#All samples masked probes Zhou and poor quality


#load(file = "Ext_lib_sesame.RData")

#Only high quality
#Selected samples based on back projection of the average
#Confirm high purity
filter1<-rownames(pheno[pheno$purity>=85,])

pheno3<-pheno[rownames(pheno)%in%filter1 ,]
table(pheno3$CellType)
beta3<-betas_sesame[, colnames(betas_sesame)%in%rownames(pheno3)]
identical(colnames(beta3), rownames(pheno3))


myeloid<-c( "Bas", "Eos", "Mono", "Neu" )
Tcells<-c("CD4mem","CD4nv" ,"CD8nv" ,"CD8mem","Treg")
Bcells<-c("Bmem","Bnv" )
CD8subtypes<-c( "CD8nv","CD8mem")

for (i in c("myeloid", "Tcells", "Bcells",  "CD8subtypes")){
  mdsPlot(beta3[, pheno3$CellType%in%get(i)], 
        sampGroups = pheno3$CellType[pheno3$CellType%in%get(i)],
        sampNames = pheno3$CellType[pheno3$CellType%in%get(i)], 
        legendPos = "topleft", legendNCol = 9, pal = groups[get(i)])
}

```

Are the control probes OK? Is the variance explained by specific phenotypes?

```{r QC, fig.height=6, fig.width=11, cache=TRUE}
#plot controls

plotCtrl2(epicData,IDorder=NULL)

betasnoob<-getBeta(epicData)
#head(pData(MsetEx))

annotation_col=data.frame(Sex= pheno$Sex, mSex= pheno$predictedSex, CellType=pheno$CellType, 
                          Age=pheno$Age, mage=pheno$Horvath_age, Ethnicity= pheno$Ethnicity_wide, 
                          methnicity=pheno$meth_ethnicity,frac_na=pheno$frac_na, purity=pheno$purity,
                          mpurity=pheno$meth_purity,
                          row.names = rownames(pheno))
#Unfiltered
pcrplot2(betasnoob[, colnames(betasnoob)%in%rownames(pheno)],annotation_col,npc=20)

#Filtered problematic samples

annotation_col=data.frame(Sex= pheno3$Sex, mSex= pheno3$predictedSex, CellType=pheno3$CellType, 
                          Age=pheno3$Age, mage=pheno3$Horvath_age, Ethnicity= pheno3$Ethnicity_wide, 
                          methnicity=pheno3$meth_ethnicity,frac_na=pheno3$frac_na, purity=pheno3$purity,
                          mpurity=pheno3$meth_purity,
                          row.names = rownames(pheno3))
pcrplot2(betasnoob[, colnames(betasnoob)%in%colnames(beta3)],annotation_col,npc=20)

```

Principal component regression analysis

```{r Supplementary Fig 3, fig.height=11, fig.width=11, cache=TRUE}
dput(levels(as.factor(pheno3$CellType)))
pheno4<-pheno3[pheno3$CellType%in%c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg"),]
#Filtered problematic samples

annotation_col=data.frame(CellType=pheno4$CellType, 
                          Sex= pheno4$Sex, 
                          Age=pheno4$Age,
                          mAge=pheno4$Horvath_age,
                          Ethnicity= pheno4$Ethnicity_wide,
                          mEthnicity= pheno4$meth_ethnicity,
                          weight=pheno4$weight_kg, 
                          height=pheno4$height_cm,
                          BMI=pheno4$bmi,
                          Smoker=pheno4$smoker,
                          FACS_purity=pheno4$purity,
                          mPurity=pheno4$meth_purity,
                          row.names = rownames(pheno4))
pcrplot2(betasnoob[, colnames(betasnoob)%in%rownames(pheno4)],annotation_col,npc=20)
```

Are the library samples pure enough? USing the IDOL-6 we can examine how they are related to the six broader categories

```{r FlowSorted.Blood.EPIC1, fig.height=11, fig.width=15, cache=TRUE, eval=FALSE, include=FALSE}

hub <- ExperimentHub()
myfiles <- query(hub, "FlowSorted.Blood.EPIC")
FlowSorted.Blood.EPIC <- myfiles[[1]]
#FlowSorted.Blood.EPIC

data (IDOLOptimizedCpGs)

epicData2<-epicData
sampleNames(epicData2) <- paste("ext",
    sampleNames(epicData2), sep = "_")

estimatesbloodref<-estimateCellCounts2(epicData2, compositeCellType = "Blood", 
                                       processMethod = "preprocessNoob",
                                       probeSelect = "IDOL", 
                                       cellTypes = c("CD8T", "CD4T", "NK", "Bcell", 
                                                     "Mono", "Neu"), 
                                       referencePlatform = 
                                           "IlluminaHumanMethylationEPIC",
                                       referenceset = NULL,
                                       IDOLOptimizedCpGs =IDOLOptimizedCpGs, 
                                       returnAll = TRUE, lessThanOne=TRUE)
cellcount_round<-round(estimatesbloodref$counts*100,2)
boxplot(cellcount_round)
#
pheno<-as.data.frame(pData(epicData))


pheno4<-as.data.frame(cellcount_round)
pheno4$CellType<-pheno$CellType

#save(epicData, file="Raw_Data_new.RData")
#save(pheno3, MsetEx, betasnoob, betas_clean2, qcscore2, file = "Betas_cleaned_new.RData")

#save(estimatesbloodref, pheno4,  file = "Cellcountsprojections.RData")
# colnames(epicData)<-colnames(MsetEx)
# rownames(pheno)<-colnames(MsetEx)
```

Cell projections purity barplots

```{r FlowSorted.Blood.EPIC2, fig.height=11, fig.width=15, cache=TRUE}
stack<-pheno4[,c("Bcell","CD4T",  "CD8T", "Neu",  "Mono", "NK" )]
stack<-round(stack*100/rowSums(stack),2)
try(boxplot(stack))
stack<-t(stack)
dat2 <- melt(stack)

pheno4$label<-paste0(pheno4$CellType,"_", seq(along = pheno4$CellType))
fetaladultspine<- ggplot(dat2, aes(x=X2, y=value, fill=X1)) + 
    geom_bar(stat="identity") +
    xlab("\nSample") +
    ylab("Cell %\n") +
    theme_bw()+
    ggtitle("Cell projection IDOL") + scale_fill_manual(values=c("#1B9E77", "#D95F02", "#7570B3",  
                                                                 "#66A61E","#E7298A", 
                                                                 "#E6AB02"))+
    theme(axis.text.x=element_text(angle = -45, hjust = 0))+
    theme(legend.title=element_blank()) +
    scale_x_discrete(labels=pheno3$label)
#theme(legend.position="none")
#detach(pheno2)
try(fetaladultspine)


for (i in dput(levels(as.factor(pheno4$CellType)))){
    stack<-pheno4[pheno4$CellType==i,c("Bcell","CD4T",  "CD8T", "Neu",  "Mono", "NK" )]
    stack<-round(stack*100/rowSums(stack),2)
    stack<-t(stack)
    dat2 <- melt(stack)
    colors1<-c(Bcell= "#1B9E77", CD4T="#D95F02", CD8T= "#7570B3",  
               Mono= "#E7298A", Neu= "#66A61E",
               NK= "#E6AB02")
    colors2<-colors1[levels(dat2$X1)]
    fetaladultspine<- ggplot(dat2, aes(x=X2, y=value, fill=X1)) + 
        geom_bar(stat="identity") +
        xlab("") +
        ylab("") + 
        geom_hline(yintercept=70, linetype="dashed", color = "red")+ 
        theme_bw()+
        ggtitle(paste("Cell projection IDOL", i)) + scale_fill_manual(values=colors2)+
        theme(axis.text.x=element_blank())+
        theme(legend.title=element_blank(), legend.position = "none") 
    assign(x = paste0(i,"_bar"),value = fetaladultspine)   
}

try(multiplot(  Mono_bar, Bas_bar, Neu_bar,Eos_bar, 
            CD4nv_bar, CD4mem_bar, Treg_bar,  NK_bar,
            CD8nv_bar, CD8mem_bar,
            Bnv_bar, Bmem_bar, 
            cols=4))
```
Examining some top genes in the IDOL-6 and how the most variable probes cluster the cell types
```{r top genes IDOL6 and clustering, cache=TRUE, fig.height=11, fig.width=11}
#beta3

table(pheno3$CellType)

identical(colnames(beta3), rownames(pheno3))
beta3a=Ext_deconv_betas3[, colnames(beta3)]
colnames(beta3a)<-paste0(pheno3$CellType,seq(1:length(colnames(beta3a))))

#Top genes in the IDOL-6 library
visualizeGene("CD8A",beta3a,platform = "EPIC",refversion = "hg38")#, cluster.samples = TRUE)
visualizeGene("BLK",beta3a,platform = "EPIC",refversion = "hg38")
visualizeGene("RPTOR",beta3a,platform = "EPIC",refversion = "hg38")
visualizeGene("CLASP1",beta3a,platform = "EPIC",refversion = "hg38")
visualizeGene("NFIA",beta3a,platform = "EPIC",refversion = "hg38")
visualizeGene("SLFN5",beta3a,platform = "EPIC",refversion = "hg38")

groups<-c( Bas= "#E41A1C", Eos="#3E8E93",  Neu="#66A61E", Mono="#E7298A",
           Bnv="#1B9E77", Bmem= "magenta", 
           CD4nv="#D95F02", CD4mem="darkblue", Treg="black",
           CD8nv ="#7570B3",CD8mem="maroon", NK="#E6AB02",MIX="white")

ann_colors = list(
    CellType = groups
) 
annotation_col=data.frame(CellType=pheno3$CellType, row.names = colnames(beta3))

pheatmap::pheatmap(getBeta(FlowSorted.BloodExtended.EPIC)[IDOLOptimizedCpGs,], color = colorRampPalette(c("yellow", "black", "blue"), space = "Lab")(128), 
                   annotation_colors = ann_colors,
                   #      annotation_row = annotation_row, 
                   annotation_col = annotation_col, 
                   main = "IDOL six unfiltered",show_rownames = FALSE,
                   cluster_cols = T, cluster_rows = T, labels_col = pheno3$CellType)
totalna<-rowSums(is.na(beta3[IDOLOptimizedCpGs,]))




annotation_col=data.frame(CellType=pheno3$CellType, row.names = rownames(pheno3))

pheatmap::pheatmap(beta3a[rownames(beta3a)%in%IDOLOptimizedCpGs,], color = colorRampPalette(c("yellow", "black", "blue"), space = "Lab")(128), 
                   annotation_colors = ann_colors,
                   #      annotation_row = annotation_row, 
                   annotation_col = annotation_col, 
                   main = "IDOL six unfiltered",show_rownames = FALSE,
                   cluster_cols = T, cluster_rows = T,  labels_col = pheno3$CellType)




mdsPlot(beta3, sampNames = pheno3$CellType, sampGroups = pheno3$CellType, pal = groups[levels(as.factor(pheno3$CellType))],legendNCol = 8, legendPos = "topleft")

```

```{r Supplementary Fig. 5a 5b, fig.height=11, fig.width=11, eval=TRUE, include=TRUE}

#tmp3<-getBeta(Ext_deconv_rgset)[,Ext_deconv_pheno$CellType=="WBC"]
load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/Raw_data_mix.RData")
tmp3<-getBeta(MsetEx)[,Ext_deconv_pheno$CellType=="WBC"]
tmppheno<-Ext_deconv_pheno[Ext_deconv_pheno$CellType=="WBC",]
stack<-as.data.frame(tmppheno[tmppheno$CellType=="WBC",c("Basophil","Bcell","CD4T",  "CD8T", "Eos","Neu",  "Mono", "NK", "Treg" )])
sumage<-rowSums(stack)
stack<-round(stack*100/sumage,1)
stack$Bas<-stack$Basophil


data("IDOLOptimizedCpGs.compTable")

stack2<-stack
stack2$CD4T<-stack2$CD4T+stack2$Treg
stack2$Gran<-stack2$Neu+stack2$Eos+stack2$Bas

cellcountsbase<-projectCellType_CP(tmp3[rownames(IDOLOptimizedCpGs.compTable),], as.matrix(IDOLOptimizedCpGs.compTable), lessThanOne = TRUE)
cellcountsbase<-round(100*cellcountsbase,2)

head(stack)
cellcountsval<-projectCellType_CP(tmp3[IDOLOptimizedCpGsBloodExtended,], FlowSorted.BloodExtended.EPIC.compTable, lessThanOne = TRUE)
boxplot(cellcountsval)
cellcountsval<-round(cellcountsval*100,2)
cellcountsval<-as.data.frame(cellcountsval)
cellcountsval$CD8T<-cellcountsval$CD8mem+cellcountsval$CD8nv
cellcountsval$CD4T<-cellcountsval$CD4mem+cellcountsval$CD4nv
cellcountsval$Bcell<-cellcountsval$Bnv+cellcountsval$Bmem


cellTypes<-c("Bcell","CD4T",  "CD8T", "Gran",  "Mono", "NK")
cellcolors<-c(  Bcell="#1B9E77", 
                   CD4T="#D95F02", CD8T ="#7570B3",
           Gran="#66A61E", Mono="#E7298A",NK="#E6AB02",  
           Treg="black")
truevalues<-stack2
cellcount_EPICIDOL<-as.data.frame(cellcountsbase)
cellcount_EPICIDOL$Gran<-cellcount_EPICIDOL$Neu
K = length(cellTypes)  # number of cell types


cellTypes<-c("Bas", "Bcell","CD4T",  "CD8T", "Eos","Neu",  "Mono", "NK", "Treg" )
cellcolors<-c( Bas= "#E41A1C", Bcell="#1B9E77", 
                   CD4T="#D95F02", CD8T ="#7570B3",
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02", 
           Treg="black")
truevalues<-stack
cellcount_EPICIDOL<-cellcountsval

K = length(cellTypes)  # number of cell types

matrixr<-data.frame(Celltypes=cellTypes, R2=rep(NA, length(cellTypes)), RMSE= rep(NA, length(cellTypes)), row.names=cellTypes)

par(mfrow=c(3,3))
for(k in 1:K) {
	pred = as.numeric(cellcount_EPICIDOL[,cellTypes[k]])
	obs = as.numeric(truevalues[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))
	matrixr$R2[k]<-round(R2,2)
	matrixr$RMSE[k]<-round(RMSE,2)
	rangexy = range(c(pred,obs))

}

cellTypes<-c("Bas", "Bcell","CD4T",  "CD8T", "Eos","Neu",  "Mono", "NK", "Treg" )
cellcolors<-c( Bas= "#E41A1C", Bcell="#1B9E77", CD4mem="darkblue", 
                   CD4nv="#D95F02", CD8T ="#7570B3", CD8mem="maroon", 
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02", 
           Treg="black")

truevalues<-stack
cellcount_EPICIDOL<-cellcountsval


# Scatter Pieplots
stack2<-t(stack[,cellTypes])
dat2 <- melt(stack2)#, id.vars = "Sample")
dat2$X1<-as.character(dat2$X1)
dat2<-dat2[order(dat2$X2),]
dat2<-dat2[order(dat2$X1),]

stack3<-t(cellcountsval[,colnames(cellcountsval)%in%cellTypes])
dat3 <- melt(stack3)
dat3$X1<-as.character(dat3$X1)
dat3<-dat3[order(as.character(dat3$X2)),]
dat3<-dat3[order(as.character(dat3$X1)),]

identical(dat2$X2, dat3$X2)
identical(as.character(dat2$X1), as.character(dat3$X1))

dat2$obs<-dat2$value
dat2$pred<-dat3$value

groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
#Bcell = "gold", 
Bnv = "#1B9E77", Bmem = "magenta", #CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", #CD8T = "darkgray", 
CD8nv = "#7570B3", CD8mem = "maroon", #CD8T_CM = "#705C83", 
#NKT = "#7E6E85", 
NK = "#E6AB02")#, WBC = "white")

for (i in names(groups))
dat2[,i]<-0

for(i in c("Bas", "Eos","Neu",  "Mono", "NK", "Treg" ))
  dat2[dat2$X1==i,i]<-cellcountsval[,i]
dat2[dat2$X1=="CD4T",c("CD4nv","CD4mem")]<-cellcountsval[,c("CD4nv","CD4mem")]
dat2[dat2$X1=="CD8T",c("CD8nv","CD8mem")]<-cellcountsval[,c("CD8nv","CD8mem")]
dat2[dat2$X1=="Bcell",c("Bnv","Bmem")]<-cellcountsval[,c("Bnv","Bmem")]


dput(colnames(dat2))

plotval1<-ggplot() + 
  geom_scatterpie(aes(x=obs, y=pred, r=1), data=dat2, cols=c("Bas", 
                                                               "Bnv", "Bmem", 
                                                               "CD4nv", "CD4mem",
                                                               "CD8nv", "CD8mem",
                                                               "Eos", "Mono", "Neu", 
                                                               "NK", "Treg"))+
  xlab("Observed (%)") +#xlab("\nSample") +
        ylab("Predicted (%)") +#ylab("Cell %\n") + 
  geom_smooth(method = "lm", se = FALSE)+
        geom_abline(intercept=0, slope=1, linetype="dashed", color = "black")+ 
        #guides(fill=T) +
        theme_classic(base_size = 16)+ #theme(text = element_text(size=rel(2)))+
        ggtitle("Cell projection extended versus FACS and CBC data") + scale_fill_manual(values=groups)+
        #theme(axis.text.x=element_blank())+#element_text(angle = -90, hjust = 0))+
  annotation_custom(tableGrob(matrixr, rows=NULL), 
                    xmin=40, xmax=60, ymin=0, ymax=18)
        

plotval1



#counts
cell_counts<-read.csv("FCM_counts.csv", row.names = 1, header = TRUE)


cellTypes<-c("Bas", "Bcell","CD4T",  "CD8T", "Eos","Neu",  "Mono", "NK", "Treg" )
cellcolors<-c( Bas= "#E41A1C", Bcell="#1B9E77", #CD4mem="darkblue", 
                   CD4T="#D95F02", CD8T ="#7570B3",#CD8T_mem="maroon", #CD8T_CM="#705C83", 
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02", #NKT= "#7E6E85", 
           Treg="black")
identical(rownames(stack), rownames(cell_counts))
truevalues0<-stack
truevalues<-matrix(NA,nrow=nrow(stack), ncol = ncol(stack), dimnames = list(rownames(stack), colnames(stack)))
for (i in cellTypes){
  truevalues[,i]<-cell_counts[, paste0(i,"_counts")]
}
cellcount_EPICIDOL<-(cellcountsval*cell_counts$WBC_counts)/100

K = length(cellTypes)  # number of cell types

matrixr<-data.frame(Celltypes=cellTypes, R2=rep(NA, length(cellTypes)), 
                    RMSE= rep(NA, length(cellTypes)),wRMSE =rep(NA, length(cellTypes)),
                    normal_count =rep(NA, length(cellTypes)),row.names=cellTypes)

normal_count<-c( Bas= "10-80", Bcell="70-530", 
                   CD4T="300-1500", CD8T ="140-820", 
           Eos="30-300",
           Neu="2k-6k", Mono="200-900",NK="80-430", 
           Treg="7-52*")

par(mfrow=c(3,3))
for(k in 1:K) {
	pred = as.numeric(cellcount_EPICIDOL[,cellTypes[k]])
	obs = as.numeric(truevalues[,cellTypes[k]])
	wgt2 = as.numeric(truevalues0[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))
	wRMSE = MetricsWeighted::rmse(obs,pred, w= wgt2)
	matrixr$R2[k]<-round(R2,2)
	matrixr$RMSE[k]<-round(RMSE,0)
	matrixr$wRMSE[k]<-round(wRMSE,0)
	matrixr$normal_count[k]<-normal_count[cellTypes[k]]
	rangexy = range(c(pred,obs))

}

cellTypes<-c("Bas", "Bcell","CD4T",  "CD8T", "Eos","Neu",  "Mono", "NK", "Treg" )
cellcolors<-c( Bas= "#E41A1C", Bcell="#1B9E77", CD4mem="darkblue", 
                   CD4nv="#D95F02", CD8T ="#7570B3", CD8mem="maroon", 
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02",  
           Treg="black")

truevalues<-truevalues
cellcount_EPICIDOL<-cellcount_EPICIDOL


stack2<-t(truevalues[,cellTypes])
dat2 <- melt(stack2)#, id.vars = "Sample")
dat2$X1<-as.character(dat2$X1)
dat2<-dat2[order(dat2$X2),]
dat2<-dat2[order(dat2$X1),]

stack3<-t(cellcount_EPICIDOL[,colnames(cellcount_EPICIDOL)%in%cellTypes])
dat3 <- melt(stack3)
dat3$X1<-as.character(dat3$X1)
dat3<-dat3[order(as.character(dat3$X2)),]
dat3<-dat3[order(as.character(dat3$X1)),]

identical(dat2$X2, dat3$X2)
identical(as.character(dat2$X1), as.character(dat3$X1))

dat2$obs<-dat2$value
dat2$pred<-dat3$value

groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
Bnv = "#1B9E77", Bmem = "magenta", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", 
CD8nv = "#7570B3", CD8mem = "maroon", 
NK = "#E6AB02")

for (i in names(groups))
dat2[,i]<-0

for(i in c("Bas", "Eos","Neu",  "Mono", "NK", "Treg" ))
  dat2[dat2$X1==i,i]<-cellcountsval[,i]
dat2[dat2$X1=="CD4T",c("CD4nv","CD4mem")]<-cellcountsval[,c("CD4nv","CD4mem")]
dat2[dat2$X1=="CD8T",c("CD8nv","CD8mem")]<-cellcountsval[,c("CD8nv","CD8mem")]
dat2[dat2$X1=="Bcell",c("Bnv","Bmem")]<-cellcountsval[,c("Bnv","Bmem")]


dput(colnames(dat2))

plotval1a<-ggplot() + 
  geom_scatterpie(aes(x=obs, y=pred, r=80), data=dat2, cols=c("Bas", 
                                                               "Bnv", "Bmem", 
                                                               "CD4nv", "CD4mem",
                                                               "CD8nv", "CD8mem",
                                                               "Eos", "Mono", "Neu", 
                                                               "NK", "Treg"))+
  xlab("Observed counts") +#xlab("\nSample") +
        ylab("Predicted counts") +#ylab("Cell %\n") + 
  geom_smooth(method = "lm", se = FALSE)+
        geom_abline(intercept=0, slope=1, linetype="dashed", color = "black")+ 
        #guides(fill=T) +
        theme_classic(base_size = 20)+ #theme(text = element_text(size=rel(2)))+
        ggtitle("Cell projection extended versus FACS and CBC data") + scale_fill_manual(values=groups)+
        #theme(axis.text.x=element_blank())+#element_text(angle = -90, hjust = 0))+
  annotation_custom(tableGrob(matrixr, rows=NULL), 
                    xmin=2500, xmax=3500, ymin=0, ymax=1000)
        

plotval1a

```

CB artificial mixtures Preprocessing was te same minfi+sesame

```{r, eval= FALSE, include=FALSE}
load("I:/Dropbox (Christensen Lab)/References_ext_new/Final/CordbloodEPIC_ref.rda")
CordbloodEPICset<-preprocessNoob(CordbloodEPIC)
    IDATprefixes <- CordbloodEPIC$filenames#searchIDATprefixes(getwd(), recursive=TRUE)
    names(IDATprefixes)<-colnames(CordbloodEPIC)
    register(FutureParam())
    cl <- parallel::makeCluster(detectCores(), type = "SOCK")
    plan(cluster, workers = cl,.cleanup = TRUE)
    p<-FutureParam()
    tmp<-openSesame(IDATprefixes,
                    mask = TRUE,
                    sum.TypeI = TRUE, 
                    platform = "EPIC", 
                    BPPARAM = p)
    parallel::stopCluster(cl)
    plan(sequential)
    gc()
    closeAllConnections()
    register(FutureParam())
    cl <- parallel::makeCluster(detectCores(), type = "SOCK")
    plan(cluster, workers = cl,.cleanup = TRUE)
    p<-FutureParam()
    tmp2 <- openSesame(IDATprefixes,
                       mask = TRUE,
                    sum.TypeI = TRUE, 
                    platform = "EPIC", 
                    BPPARAM = p,
                    what="sigset")
    parallel::stopCluster(cl)
    plan(sequential)
    gc()
    closeAllConnections()
    register(FutureParam())
    cl <- parallel::makeCluster(detectCores(), type = "SOCK")
    plan(cluster, workers = cl,.cleanup = TRUE)
    p<-FutureParam()
    tmp3<-SigSetsToRGChannelSet(tmp2, BPPARAM = p)
    parallel::stopCluster(cl)
    plan(sequential)
    gc()
    closeAllConnections()
    #print(table(sheet$CellType))
    pheno<-as.data.frame(pData(CordbloodEPIC))
    
    qc<-foreach(j = 1:length(names(tmp2))) %do% {#
      sesameQC(tmp2[[j]], betas = tmp[,j])
    }

    names(qc)<-names(tmp2)
    df1<-as.data.frame(data.table::rbindlist(qc))
    rownames(df1)<-names(tmp2)
    df1<-df1[rownames(df1)%in%rownames(pheno),]
    df1<-df1[rownames(pheno),]
    identical(rownames(pheno), rownames(df1))
    pheno<-cbind.data.frame(pheno, df1)
    tmp<-tmp[,colnames(tmp)%in%rownames(pheno)]
    tmp<-tmp[,rownames(pheno)]
    identical(rownames(pheno), colnames(tmp))
    tmp3<-tmp3[,colnames(tmp3)%in%rownames(pheno)]
    tmp3<-tmp3[,rownames(pheno)]
    identical(rownames(pheno), colnames(tmp3))
    pheno$karyotype<-NA
    karyotype<-foreach(k = 1:length(names(tmp2))) %do% {
      inferSexKaryotypes(tmp2[[k]])
    }
    pheno$karyotype<-unlist(karyotype)
    print(table(pheno$karyotype, pheno$Sex))
    boxplot(pheno$frac_na, main= paste("Fraction of non detection"))
    pData(tmp3)<-DataFrame(pheno, row.names = rownames(pheno))
    #extract OOB p-values
    pvalsesame<-matrix(data = NA,nrow = length(tmp2[[1]]@pval$pOOBAH),
                       ncol=length(rownames(pheno)), 
                       dimnames = list(names(tmp2[[1]]@pval$pOOBAH),
                                       rownames(pheno)))
    for ( l in rownames(pheno)){
      pvalsesame[,l]<-tmp2[[l]]@pval$pOOBAH
      }
    pvalsesame[1:5,1:5]
    pvalsesame<-pvalsesame[,rownames(pheno)]
    identical (rownames(pheno), colnames(pvalsesame))
    #calculating CNV segments
    ssets.normal <- sesameDataGet('EPIC.5.normal')
    segmentmeth<-function(m,sset, sset.normal, refversion ="hg38"){
      seg <- cnSegmentation(sset[[m]], ssets.normal, refversion = refversion)
      seg$seg.signals$Sample_ID<-names(sset)[[m]]
      seg2<-list(seg$seg.signals)
      names(seg2)<-m
      return(seg2)
    }
    # registerDoFuture()
    # plan(cluster, workers = cl,.cleanup = TRUE)
    seg0<-foreach(m = 1:length(names(tmp2))) %do% {#
      segmentmeth(m, tmp2, ssets.normal, refversion = "hg38")
    }
    for (i  in 1: length(names(tmp2))){
      if (i==1){
        seg<-as.data.frame(seg0[[i]])
        colnames(seg)<-c("ID", "chrom", "loc.start", "loc.end", "num.mark", 
"seg.mean", "seg.sd", "seg.median", "seg.mad", "pval", 
"lcl", "ucl", "Sample_ID")
      } else{
        seg1<-as.data.frame(seg0[[i]])
        colnames(seg1)<-c("ID", "chrom", "loc.start", "loc.end", "num.mark", 
"seg.mean", "seg.sd", "seg.median", "seg.mad", "pval", 
"lcl", "ucl", "Sample_ID")
        seg<-rbind.data.frame(seg,seg1)
      }
    }
    seg<-as.data.frame(data.table::rbindlist(list(seg0)))
    
    # # parallel::stopCluster(cl)
    # # plan(sequential)
    # # gc()
    # # closeAllConnections()
    # # #ExtraSNPs using Type I probes
    tmp1<-  getAFTypeIbySumAlleles(tmp2[[1]])
    extraSNPAFs <- matrix(data = NA,nrow = length(tmp1),
                   ncol=length(rownames(pheno)), dimnames = list(names(tmp1),
                                                           rownames(pheno)))
    for ( n in rownames(pheno)){
      extraSNPAFs[,n]<-getAFTypeIbySumAlleles(tmp2[[n]])
      }
i="CB_deconv"
    assign(x = paste0(i, "_betas"),value = tmp)
    assign(x = paste0(i, "_rgset"),value = tmp3)
    assign(x = paste0(i, "_seg"),value = seg)
    assign(x = paste0(i, "_pOOBHA"),value = pvalsesame)
    assign(x = paste0(i, "_SNP"),value = extraSNPAFs)
    assign(x = paste0(i, "_pheno"),value = pheno)
    save(list = c(paste0(i, "_betas"),
                  paste0(i, "_rgset"),
                  paste0(i, "_seg"),
                  paste0(i, "_pOOBHA"),
                  paste0(i, "_SNP"),
                  paste0(i, "_pheno")), file= paste0(i,"_sesame.RData"))
```

```{r filter CBmix, eval= TRUE, include=FALSE}
load("I:/Dropbox (Christensen Lab)/References_ext_new/Final/CordbloodEPIC_ref.rda")
CordbloodEPICset<-preprocessNoob(CordbloodEPIC)
load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/CB_deconv_sesame.RData")
annot<-as.data.frame(readRDS("D:/OneDrive/OneDrive - Dartmouth College/Crossreactive/EPIC.hg19.manifest.rds"))
#head(annot)

annot<-annot[rownames(annot)%in% rownames(CB_deconv_betas),]

#filter crossreactive, problematic probes, X, Y and ch
CB_deconv_betas2<-CB_deconv_betas[rownames(CB_deconv_betas)%in%rownames(annot[annot$seqnames!="chrY" & 
                                            annot$seqnames!="chrX" & 
                                            annot$probeType=="cg" &
                                            annot$MASK_general==FALSE,]),]

dim(CB_deconv_betas2)

CB_deconv_betas3<-CB_deconv_betas2[complete.cases(CB_deconv_betas2),]

dim(CB_deconv_betas3)
```

```{r CBprojections, fig.height=11, fig.width=15, cache=TRUE}
#pheno3<-as.data.frame(pData(epicData))
#load(file = "Cellcountsprojections.RData")
#attach(pheno3)
#Only Cord blood
#pheno4<-pheno4[1:86,]
coefEsts = FlowSorted.BloodExtended.EPIC.compTable
annotation_col=data.frame(CellType=colnames(coefEsts), row.names = colnames(coefEsts))
groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", 
  Neu = "#66A61E", Mono = "#E7298A", 
#Bcell = "gold", 
Bnv = "#1B9E77", Bmem = "magenta", 
#CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", 
#CD8T = "darkgray", 
CD8nv = "#7570B3",  CD8mem = "maroon",# 
#nRBC = "#7E6E85", 
NK = "#E6AB02")#, CB_Mix = "white")
ann_colors = list(
    CellType = groups#[names(groups)%in%colnames(coefEsts)]
) 

try(pheatmap(coefEsts,color = colorRampPalette(c("yellow", "black", "blue"), space = "Lab")(128), annotation_colors = ann_colors,
             #      annotation_row = annotation_row, 
             annotation_col = annotation_col, main = "DMRs EPIC filtered cells", cluster_rows = T, show_rownames = F))#, labels_col = combined$CellType))


annotation_col=data.frame(CellType=CB_deconv_pheno$CellType, row.names = rownames(CB_deconv_pheno))

groups<-c(#Bas = "#E41A1C", Eos = "#3E8E93", 
  Neu = "#66A61E", Mono = "#E7298A", 
Bcell = "gold", 
#Bnv = "#1B9E77", Bmem = "magenta", 
CD4T = "lightgray", 
#CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", 
CD8T = "darkgray", 
#CD8nv = "#7570B3",  # 
nRBC = "#7E6E85", 
NK = "#E6AB02", CB_Mix = "white")
ann_colors = list(
    CellType = groups#[names(groups)%in%colnames(coefEsts)]
) 

referencePd2<-CB_deconv_pheno
#referenceBetas2<-CB_deconv_betas3[rownames(CB_deconv_betas3), ]
referenceBetas2<-getBeta(CordbloodEPICset)#[rownames(CB_deconv_betas3), ]

identical(colnames(referenceBetas2), rownames(referencePd2))

ann_colors = list(
    CellType = groups
) 

try(pheatmap(referenceBetas2[rownames(referenceBetas2)%in%rownames(coefEsts),],color = colorRampPalette(c("yellow", "black", "blue"), space = "Lab")(128), annotation_colors = ann_colors,
             #      annotation_row = annotation_row, 
             annotation_col = annotation_col, main = "DMRs EPIC CB filtered cells  ", cluster_rows = T, show_rownames = F, labels_col = referencePd2$CellType))


#All together
groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", 
  Neu = "#66A61E", Mono = "#E7298A", 
Bcell = "gold", 
Bnv = "#1B9E77", Bmem = "magenta", 
CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", 
CD8T = "darkgray", 
CD8nv = "#7570B3",  CD8mem = "maroon",# 
nRBC = "#7E6E85", 
NK = "#E6AB02")#, CB_Mix = "white", MIX="white")
ann_colors = list(
    CellType = groups#[names(groups)%in%colnames(coefEsts)]
) 

referencePd2<-CB_deconv_pheno
referenceBetas3<-cbind(getBeta(CB_deconv_rgset)[rownames(coefEsts),
                                                CB_deconv_rgset$CellType!="CB_Mix"],
                       getBeta(FlowSorted.BloodExtended.EPIC)[rownames(coefEsts),
                                                              FlowSorted.BloodExtended.EPIC$CellType!="MIX"])
# referenceBetas3<-cbind(getBeta(CordbloodEPICset)[rownames(coefEsts),
#                                                 CordbloodEPICset$CellType!="CB_Mix"],
#                        getBeta(FlowSorted.BloodExtended.EPIC2)[rownames(coefEsts),
#                                                               FlowSorted.BloodExtended.EPIC2$CellType!="MIX"])
identical(rownames(referenceBetas3), rownames(coefEsts))

annotation_col=data.frame(CellType=c(CB_deconv_pheno$CellType[CB_deconv_rgset$CellType!="CB_Mix"],
                                     FlowSorted.BloodExtended.EPIC$CellType[FlowSorted.BloodExtended.EPIC$CellType!="MIX"]), 
                          Source=c(rep("Cord Blood", length(CB_deconv_pheno$CellType[CB_deconv_rgset$CellType!="CB_Mix"])),
                                   rep("Adult Blood",
                                       length(FlowSorted.BloodExtended.EPIC$CellType[FlowSorted.BloodExtended.EPIC$CellType!="MIX"]))),
                          row.names = c(rownames(CB_deconv_pheno[CB_deconv_rgset$CellType!="CB_Mix",]),
                                        colnames(FlowSorted.BloodExtended.EPIC[,FlowSorted.BloodExtended.EPIC$CellType!="MIX"])))


ann_colors = list(
    CellType = groups
) 

try(pheatmap(referenceBetas3,color = colorRampPalette(c("yellow", "black", "blue"), space = "Lab")(128), annotation_colors = ann_colors,
             #      annotation_row = annotation_row, 
             annotation_col = annotation_col, main = "IDOL EPIC-Ext purified cord blood and adult cells", cluster_rows = T, show_rownames = F, labels_col = annotation_col$CellType))

#####






# cellcountsCB<-FlowSorted.Blood.EPIC:::projectCellType(CB_deconv_betas[rownames(coefEsts),], coefEsts, lessThanOne = TRUE)
cellcountsCB<-FlowSorted.Blood.EPIC:::projectCellType(getBeta(CordbloodEPICset)[rownames(coefEsts),], coefEsts, lessThanOne = FALSE)
sumage<-rowSums(cellcountsCB)
cellcountsCB2<-round(cellcountsCB*100/sumage,1)
#cellcountsCB2
phenoCB<-cbind.data.frame(cellcountsCB2, CellType=referencePd2$CellType)
#write.csv(phenoCB,file="phenoCB.csv")
#Add the Exclude
stack<-cellcountsCB2
#stack<-round(stack*100/rowSums(stack),2)
boxplot(stack)
stack<-t(stack)
dat2 <- melt(stack)#, id.vars = "Sample")

phenoCB$label<-paste0(phenoCB$CellType,"_", seq(along = phenoCB$CellType))



for (i in dput(levels(as.factor(phenoCB$CellType)))){
    stack<-phenoCB[phenoCB$CellType==i,1:12]
    stack<-round(stack*100/rowSums(stack),2)
    stack<-t(stack)
    dat2 <- melt(stack)#, id.vars = "Sample")
    #dat2$X1 <- relevel(dat2$X1, ref=i)
    #levels(dat2$X1)
    #dat2$X1 <-fct_rev(dat2$X1)
    #levels(dat2$X1)
    colors1<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
#Bcell = "gold", 
Bnv = "#1B9E77", Bmem = "magenta", #CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", #CD8T = "darkgray", 
CD8nv = "#7570B3", CD8mem = "maroon", #CD8T_CM = "#705C83", 
#nRBC = "#7E6E85", 
NK = "#E6AB02")#, CB_Mix = "white")
    colors2<-colors1[levels(dat2$X1)]
    fetaladultspine<- ggplot(dat2, aes(x=X2, y=value, fill=X1)) + 
        geom_bar(stat="identity") +
        xlab("") +#xlab("\nSample") +
        ylab("") +#ylab("Cell %\n") + 
        geom_hline(yintercept=70, linetype="dashed", color = "red")+ 
        #guides(fill=T) +
        theme_bw()+
        ggtitle(paste("Cell projection IDOL", i)) + scale_fill_manual(values=colors2)+
        theme(axis.text.x=element_blank())+#element_text(angle = -90, hjust = 0))+
        theme(legend.title=element_blank(), legend.position = "none") 
    #theme(legend.position="none")
    #detach(pheno2)
    assign(x = paste0(i,"_bar"),value = fetaladultspine)   
}


#multiplot(  WBC_bar, WBC_FACS_bar, cols=2)
multiplot(  Mono_bar, #Basophil_bar, 
            Neu_bar,#Eos_bar, 
            nRBC_bar, CB_Mix_bar,
            Bcell_bar,CD4T_bar, 
            #CD4nv_bar, CD4mem_bar, Treg_bar,  
            CD8T_bar,NK_bar,
            #CD8T_naive_bar, CD8T_EM_bar,CD8T_CM_bar, NKT_bar,
            #Bnv_bar, Bmem_bar, 
             cols=4)

```

\#CB mixture adjusted

```{r,fig.height=11, fig.width=11}
library(FlowSorted.CordBloodCombined.450k)
library(ExperimentHub)
hub <- ExperimentHub()
myfiles <- query(hub, "FlowSorted.CordBloodCombined.450k")
FlowSorted.CordBloodCombined.450k <- myfiles[[1]]
FlowSorted.CordBloodCombined.450k

coefEsts = FlowSorted.BloodExtended.450klegacy.compTable

cellcountsCBref<-FlowSorted.Blood.EPIC:::projectCellType(getBeta(FlowSorted.CordBloodCombined.450k)[rownames(coefEsts),], coefEsts, lessThanOne = TRUE)

sumage<-rowSums(cellcountsCBref)
cellcountsCBref<-round(cellcountsCBref*100/sumage,1)
#cellcountsCB2
phenoCBref<-cbind.data.frame(cellcountsCBref, CellType=FlowSorted.CordBloodCombined.450k$CellType)
#write.csv(phenoCB,file="phenoCB.csv")
#Add the Exclude
stack<-cellcountsCBref
#stack<-round(stack*100/rowSums(stack),2)
boxplot(stack)
stack<-t(stack)
dat2 <- melt(stack)#, id.vars = "Sample")

phenoCBref$label<-paste0(phenoCBref$CellType,"_", seq(along = phenoCBref$CellType))



for (i in dput(levels(as.factor(phenoCBref$CellType)))){
    stack<-phenoCBref[phenoCBref$CellType==i,1:12]
    stack<-round(stack*100/rowSums(stack),2)
    stack<-t(stack)
    dat2 <- melt(stack)#, id.vars = "Sample")
    #dat2$X1 <- relevel(dat2$X1, ref=i)
    #levels(dat2$X1)
    #dat2$X1 <-fct_rev(dat2$X1)
    #levels(dat2$X1)
    colors1<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
#Bcell = "gold", 
Bnv = "#1B9E77", Bmem = "magenta", #CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", #CD8T = "darkgray", 
CD8nv = "#7570B3", CD8mem = "maroon", #CD8T_CM = "#705C83", 
#nRBC = "#7E6E85", 
NK = "#E6AB02")#, CB_Mix = "white")
    colors2<-colors1[levels(dat2$X1)]
    fetaladultspine<- ggplot(dat2, aes(x=X2, y=value, fill=X1)) + 
        geom_bar(stat="identity") +
        xlab("") +#xlab("\nSample") +
        ylab("") +#ylab("Cell %\n") + 
        geom_hline(yintercept=70, linetype="dashed", color = "red")+ 
        #guides(fill=T) +
        theme_bw()+
        ggtitle(paste("Cell projection IDOL \n ref UCB", i)) + scale_fill_manual(values=colors2)+
        theme(axis.text.x=element_blank())+#element_text(angle = -90, hjust = 0))+
        theme(legend.title=element_blank(), legend.position = "none") 
    #theme(legend.position="none")
    #detach(pheno2)
    assign(x = paste0(i,"_bar"),value = fetaladultspine)   
}


#multiplot(  WBC_bar, WBC_FACS_bar, cols=2)
multiplot(  #Neu_bar,
            Gran_bar, Mono_bar, #Basophil_bar, 
            #Eos_bar, 
            nRBC_bar, #CB_Mix_bar, 
            WBC_bar,
            Bcell_bar,CD4T_bar, 
            #CD4nv_bar, CD4mem_bar, Treg_bar,  
            CD8T_bar,NK_bar,
            #CD8T_naive_bar, CD8T_EM_bar,CD8T_CM_bar, NKT_bar,
            #Bnv_bar, Bmem_bar, 
             cols=3)

```

```{r UCSF IDOL, fig.height=11, fig.width=11}
load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/GBM UCSF/UCSF_sesame.RData")
load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/GBM UCSF/Unfiltered.rda")
UCSF_set<-preprocessNoob(epicData)
colnames(UCSF_set)<-colnames(UCSF_betas)
# cellcountsval3<-projectCellType_CP(getBeta(UCSF_sesame)[rownames(matFDR),], as.matrix(matFDR), lessThanOne = TRUE)
# cellcountsval3<-round(cellcountsval3*100/rowSums(cellcountsval3),2)
# boxplot(cellcountsval3)
cellTypes = c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", #"CD8T_CM", 
              "CD8mem", 
"CD8nv", "Eos", #"MIX", 
"Mono", "Neu", "NK", #"NKT", 
"Treg")#,"WBC")
 Tcellmatrix <- read.metharray.sheet(base = "I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/GBM UCSF", pattern = "phenotype_extended2.csv")
 rownames(Tcellmatrix)<-Tcellmatrix$array_pos
#dput(colnames(sheet2))
#Tcellmatrix<-UCSF_pheno
#  c("Sample_Name", "Sample_Well", "Sample_Plate", "Sample_Group", 
# "Pool_ID", "array_pos", "Sampling.ID", "Idno", "Histological.Group", 
# "Time.point", "White_Cell_count", "Tcell_count", "CD4T_count_tube1", 
# "DP_count", "CD8T_count", "DN_count", "Tcell", "CD4T1", "DP", 
# "CD8T1", "DN", "CD4T_count_tube2", "CD4Tnv_count", "CD4T_RARO_DN_count", 
# "CD4Tmem_count", "CD4T_RARO_DN_count.1", "CD4T", "CD4Tnv", "CD4T_RARO_DN", 
# "CD4Tmem", "CD4T_RARO_DN.1", "CD8T_count_tube2", "CD8Tnv_count_tube2", 
# "CD8T_RARO_DN_count", "CD8Tmem_count", "CD8T_RARO_DP_count", 
# "CD8T", "CD8T_naive", "CD8T_RARO_DN", "CD8T_mem", "CD8T_RARO_DP", 
# "grade", "dxgroup", "timet", "Array", "Slide", "Basename")]

#head(estimatesbloodref)
#estimatesbloodref1<- round(estimatesbloodref$counts*100,2)
#head(estimatesbloodref1)
head(Tcellmatrix)


rmse <- function(error)
{
    sqrt(mean(error^2))
}
Tcellmatrix$CD8mem<-Tcellmatrix$CD8Tmem
Tcellmatrix$CD8nv<-Tcellmatrix$CD8T_naive
par(mfrow=c(2,3))
cellTypes<-c("CD4T", "CD4nv", "CD4mem",  "CD8T", "CD8mem","CD8nv")
celltypes<-cellTypes
colorcelltype<-c(  CD4T="#1B9E77", CD4mem="darkblue", 
                   CD4nv="#D95F02", CD8T ="#7570B3",CD8mem="maroon", CD8nv="#705C83")#, Basophil= "#E41A1C",
           # Eos="#3E8E93",
           # Neu="#66A61E", Mono="#E7298A",NK="#E6AB02", #NKT= "#7E6E85", 
           # Treg="black")
groups<-colorcelltype
truevalues<-Tcellmatrix[,c("CD4T", "CD4nv", "CD4mem",  "CD8T", "CD8mem","CD8nv", "Time.point", "Histological.Group")]
truevalues<-truevalues[complete.cases(truevalues),]
#The new estimates are extrapolated from the RARO experiment using the info from the Tcell experiment


# 3b: extract the coefEsts object from the Step 2 results
coefEsts = FlowSorted.BloodExtended.EPIC.compTable#IDOLopt_fix_1200[["IDOL Optimized CoefEsts"]]
#tmp<-getBeta(UCSF_rgset)
tmp<-getBeta(UCSF_set)
tmp<-tmp[, rownames(truevalues)]
# 3c: predict cell composition

predictedCellFraction = round(projectWBCnew(tmp[rownames(coefEsts),], coefEsts)*100,1)
predictedCellFraction<-as.data.frame(predictedCellFraction)
predictedCellFraction$CD4Tmem2<-predictedCellFraction$CD4mem
predictedCellFraction$CD4T<-rowSums(predictedCellFraction[,c("CD4mem", "CD4nv", "Treg")])
predictedCellFraction$CD4mem<-rowSums(predictedCellFraction[,c("CD4mem", "Treg")])
predictedCellFraction$CD8T<-rowSums(predictedCellFraction[,c("CD8mem", #"CD8T_CM",
                                               "CD8nv")])
#predictedCellFraction$CD8T_mem<-predictedCellFraction$CD8T_EM#rowSums(cellcountsval3[,c("CD8T_EM", "CD8T_CM")])

head(predictedCellFraction)
boxplot(predictedCellFraction)
cellcount_EPICIDOL<-predictedCellFraction
testingCovariates<-truevalues
# 3d: do a visual depiction of the results (run all of the code below at once)
cellcolors = groups#brewer.pal(6,"Set2")
K = length(cellTypes)  # number of cell types

par(mfrow = c(3,3))
for(k in 1:K) {
	pred = as.numeric(predictedCellFraction[,cellTypes[k]])
	obs = as.numeric(testingCovariates[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))

	rangexy = range(c(pred,obs), na.rm =TRUE)

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
	      cex.main = 2)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}



#Per time.point
for (i in levels(as.factor(truevalues$Time.point))){
  par(mfrow = c(3,3))
for(k in 1:K) {
	pred = as.numeric(predictedCellFraction[truevalues$Time.point==i,cellTypes[k]])
	obs = as.numeric(testingCovariates[truevalues$Time.point==i,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))

	rangexy = range(c(pred,obs), na.rm =TRUE)

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 1.5, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), "\n",i, sep = ""), 
	      cex.main = 1.5)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}
}

#Per histology
for (i in levels(as.factor(truevalues$Histological.Group))){
  par(mfrow = c(3,3))
for(k in 1:K) {
	pred = as.numeric(predictedCellFraction[truevalues$Histological.Group==i,cellTypes[k]])
	obs = as.numeric(testingCovariates[truevalues$Histological.Group==i,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))

	rangexy = range(c(pred,obs), na.rm =TRUE)

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 1.5, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), "\n",i, sep = ""), 
	      cex.main = 1.5)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}
}

#Inverted CD8Tnaive and mem
# k="CD8T_naive"
# k1="CD8T_mem"
# pred = as.numeric(predictedCellFraction[,k1])
# 	obs = as.numeric(testingCovariates[,k])
# 	R2 = cor(obs, pred, method = "pearson")^2
# 	RMSE = sqrt(sum((pred - obs)^2/length(pred)))
# 
# 	rangexy = range(c(pred,obs), na.rm =TRUE)
# 
# 	par(mar = c(5,5,4,2))
# 	plot(obs, pred, cex = 2, col = cellcolors[k], pch = 19, 
# 		xlim = rangexy, ylim = rangexy, main = "",
# 		xlab = paste0("Observed (%) ", k), ylab = paste0("Predicted (%) ", k1),
# 		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
# 	title(main = paste(cellTypes[k], "\nRMSE = ", 
# 		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
# 	      cex.main = 2)
# 	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
# 
# k1="CD8T_naive"
# k="CD8T_mem"
# pred = as.numeric(predictedCellFraction[,k1])
# 	obs = as.numeric(testingCovariates[,k])
# 	R2 = cor(obs, pred, method = "pearson")^2
# 	RMSE = sqrt(sum((pred - obs)^2/length(pred)))
# 
# 	rangexy = range(c(pred,obs), na.rm =TRUE)
# 
# 	par(mar = c(5,5,4,2))
# 	plot(obs, pred, cex = 2, col = cellcolors[k], pch = 19, 
# 		xlim = rangexy, ylim = rangexy, main = "",
# 		xlab = paste0("Observed (%) ", k), ylab = paste0("Predicted (%) ", k1),
# 		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
# 	title(main = paste(cellTypes[k], "\nRMSE = ", 
# 		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
# 	      cex.main = 2)
# 	abline(a = 0, b = 1, lwd = 2, lty = "dotted")




# par(mfrow = c(3,3))
# 
# for (i in celltypes) {
#     
#     
#     if (i=="Neu"){
#         plot(truevalues[,i],
#              cellcount_EPICIDOL[,i], main=paste(i),
#              xlab="",ylab="",#paste("True", i, "%"), ylab=paste("Estimated", i, "%"),
#              xlim=c(0,80), ylim=c(0,80),
#              pch = 21,  col="black", bg=colorcelltype[i], #pch = 20,#as.numeric(mix),
#              cex=2.5, cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5 )
#         lm.D9 <- lm(cellcount_EPICIDOL[,i]~truevalues[,i] )
#         rmse(lm.D9$residuals)
#         summary(lm.D9)$r.squared
#         abline(lm.D9, col=colorcelltype[i])
#         abline(a = 0, b=1, col="black", lwd=1, lty=2)
#         text(10, 65, substitute(paste(R^2,m), list(m=round(summary(lm.D9)$r.squared[1]*100,1))), cex = 1.5)
#         text(10, 55, substitute(paste("RMSE",m),list(m=round(rmse(lm.D9$residuals)[1],2))), cex = 1.5)
#         
#     }
#     else if (i=="CD4T"){
#         plot(truevalues[,i],
#              cellcount_EPICIDOL[,i], main=paste(i),
#              xlab="",ylab="",#paste("True", i, "%"), ylab=paste("Estimated", i, "%"),
#              xlim=c(0,25), ylim=c(0,25),
#              pch = 21,  col="black", bg=colorcelltype[i], #pch = 20,#as.numeric(mix),
#              cex=2.5, cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5 )
#         lm.D9 <- lm(cellcount_EPICIDOL[,i]~truevalues[,i] )
#         rmse(lm.D9$residuals)
#         summary(lm.D9)$r.squared
#         abline(lm.D9, col=colorcelltype[i])
#         abline(a = 0, b=1, col="black", lwd=1, lty=2)
#         text(5, 23, substitute(paste(R^2,m), list(m=round(summary(lm.D9)$r.squared[1]*100,1))), cex = 1.5)
#         text(5, 21, substitute(paste("RMSE",m),list(m=round(rmse(lm.D9$residuals)[1],2))), cex = 1.5)
#         
#     } else {
#         plot(truevalues[,i],
#              cellcount_EPICIDOL[,i], main=paste(i),
#              xlab="",ylab="",#paste("True", i, "%"), ylab=paste("Estimated", i, "%"),
#              xlim=c(0,20), ylim=c(0,20),
#              pch = 21,  col="black", bg=colorcelltype[i], #pch = 20,#as.numeric(mix),
#              cex=2.5, cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5 )
#         
#         lm.D9 <- lm(cellcount_EPICIDOL[,i]~truevalues[,i] )
#         rmse(lm.D9$residuals)
#         summary(lm.D9)$r.squared
#         abline(lm.D9, col=colorcelltype[i])
#         abline(a = 0, b=1, col="black", lwd=1, lty=2)
#         text(5, 18, substitute(paste(R^2,m), list(m=round(summary(lm.D9)$r.squared[1]*100,1))), cex = 1.5)
#         text(5, 16, substitute(paste("RMSE",m),list(m=round(rmse(lm.D9$residuals)[1],2))), cex = 1.5)
#         
#     }
# }
# 

```

```{r validation, fig.height=8, fig.width=11, eval=TRUE, include=TRUE}
#mat150<-matcells2
#matFDR<-matFDR
#validation
library(ExperimentHub)
hub <- ExperimentHub()
query(hub, "FlowSorted.Blood.EPIC")
FlowSorted.Blood.EPIC <- hub[["EH1136"]]
FlowSorted.Blood.EPIC

# Step 2 separate the reference from the testing dataset if you want to run 
# examples for estimations for this function example

RGsetTargets <- FlowSorted.Blood.EPIC[,
             FlowSorted.Blood.EPIC$CellType == "MIX"]
             
sampleNames(RGsetTargets) <- paste(RGsetTargets$CellType,
                            seq_len(dim(RGsetTargets)[2]), sep = "_")
RGsetTargets



#tmp3<-getBeta(preprocessNoob(RGsetTargets))#[,Ext_deconv_pheno$CellType=="WBC"]
#tmp3a<-getBeta(sesamize(RGsetTargets))#[,Ext_deconv_pheno$CellType=="WBC"]
tmp3a<-getBeta(preprocessNoob(RGsetTargets))#[,Ext_deconv_pheno$CellType=="WBC"]
phenoIDOLpre<-pData(RGsetTargets)[,c("Bcell","CD4T",  "CD8T", "Neu",  "Mono", "NK" )]
stack<-as.data.frame(phenoIDOLpre)#[phenoIDOLpre$CellType=="MIX",c("Basophil","Bcell","CD4T",  "CD8T", "Eos","Neu",  "Mono", "NK", "Treg" )])
#stack<-stack[1:6,]
sumage<-rowSums(stack)
stack<-round(stack*100/sumage,1)
stack$Gran<-stack$Neu
head(stack)
cellcountsbase<-projectCellType_CP(tmp3a[IDOLOptimizedCpGs,], IDOLOptimizedCpGs.compTable, lessThanOne = TRUE)
boxplot(cellcountsbase)
cellcountsbase<-round(cellcountsbase*100,2)
cellcountsbase<-as.data.frame(cellcountsbase)
cellcountsbase$Gran<-cellcountsbase$Neu
cellcountsval<-projectCellType_CP(tmp3a[IDOLOptimizedCpGsBloodExtended,], FlowSorted.BloodExtended.EPIC.compTable, lessThanOne = TRUE)

boxplot(cellcountsval)
cellcountsval<-round(cellcountsval*100,2)
cellcountsval<-as.data.frame(cellcountsval)
cellcountsval$CD8T<-cellcountsval$CD8mem+ cellcountsval$CD8nv
cellcountsval$CD4T<-cellcountsval$CD4mem+cellcountsval$CD4nv+cellcountsval$Treg
cellcountsval$Bcell<-cellcountsval$Bnv+cellcountsval$Bmem
cellcountsval$Gran<-cellcountsval$Neu+cellcountsval$Eos+cellcountsval$Bas



cellTypes<-c("Bcell","CD4T",  "CD8T", "Gran",  "Mono", "NK")
cellcolors<-c(  Bcell="#1B9E77", #CD4mem="darkblue", 
                   CD4T="#D95F02", CD8T ="#7570B3",#CD8T_mem="maroon", #CD8T_CM="#705C83", Basophil= "#E41A1C",
           Eos="#3E8E93",
           Gran="#66A61E", Mono="#E7298A",NK="#E6AB02")#, #NKT= "#7E6E85", 
           #Treg="black")
truevalues<-stack
cellcount_EPICIDOL<-cellcountsval

K = length(cellTypes)  # number of cell types

par(mfrow = c(2,3))
for(k in 1:K) {
	pred = as.numeric(cellcount_EPICIDOL[,cellTypes[k]])
	obs = as.numeric(truevalues[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))

	rangexy = range(c(pred,obs))

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
	      cex.main = 2)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}



cellTypes<-c("Bcell","CD4T",  "CD8T", "Gran",  "Mono", "NK")
cellcolors<-c(  Bcell="#1B9E77", #CD4mem="darkblue", 
                   CD4T="#D95F02", CD8T ="#7570B3",#CD8T_mem="maroon", #CD8T_CM="#705C83", Basophil= "#E41A1C",
           Eos="#3E8E93",
           Gran="#66A61E", Mono="#E7298A",NK="#E6AB02")#, #NKT= "#7E6E85", 
           #Treg="black")
truevalues<-stack


cellcount_EPICIDOL<-cellcountsbase

K = length(cellTypes)  # number of cell types

par(mfrow = c(2,3))
for(k in 1:K) {
	pred = as.numeric(cellcount_EPICIDOL[,cellTypes[k]])
	obs = as.numeric(truevalues[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))

	rangexy = range(c(pred,obs))

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
	      cex.main = 2)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}




```

```{r Supplementary Figure 1, fig.height=8, fig.width=12}


celltype<-c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv",  "CD8mem", 
                                              "CD8nv", "Eos", 
                                              "Mono", "Neu", "NK", 
                                              "Treg")

epicData2<-FlowSorted.BloodExtended.EPIC[,FlowSorted.BloodExtended.EPIC$CellType%in%celltype]

print("all purity")
  print(summary(epicData2$purity))
  print("all meth purity")
 print(summary(epicData2$meth_purity))
 print(summary(epicData2$meth_purity[epicData2$CellType!="CD8mem"]))
for (i in levels(as.factor(epicData2$CellType))){
 
  print(i)
 print(summary(epicData2$purity[epicData2$CellType==i]))
 
 print(summary(epicData2$meth_purity[epicData2$CellType==i]))
}

boxplot(FlowSorted.BloodExtended.EPIC$purity[FlowSorted.BloodExtended.EPIC$CellType%in%celltype]~FlowSorted.BloodExtended.EPIC$CellType[FlowSorted.BloodExtended.EPIC$CellType%in%celltype])

library(ggsignif)
library(ggpubr)
phenobox<-pData(epicData2)
boxdat<-as.data.frame(phenobox[phenobox$CellType!="MIX",c("purity", "CellType", "meth_purity")])
boxdat<-boxdat[order(boxdat$CellType),]

#boxdat$purity<-as.numeric(boxdat$purity)
my_comparisons<- list(c("Bas", "CD8mem"),
                      c("Bmem", "CD8mem"),
                      c("Bnv", "CD8mem"),
                      c("CD4mem", "CD8mem"),
                      c("CD4nv", "CD8mem"),
                      c("CD8nv", "CD8mem"),
                      c("Eos", "CD8mem"),
                      c("Mono", "CD8mem"),
                      c("Neu", "CD8mem"),
                      c("NK", "CD8mem"),
                      c("Treg", "CD8mem"))#,
#c("CD8T", "NK"),
#c("Neu", "Bcell"))
#c("Bcell",  "CD4T", "CD8T", "Mono", "Neu", "NK")
colors1<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
Bnv = "#1B9E77", Bmem = "magenta",  
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black",  
CD8nv = "#7570B3", CD8mem = "maroon", 
 NK = "#E6AB02")
p<-ggboxplot(boxdat, x = "CellType", y = "purity",
          color = "CellType", palette = colors1,
          xlab = "Cell type", ylab = "Cell sorting purity %") +
    #stat_compare_means(comparisons = my_comparisons, label.y = c(105, 103, 101, 102, 104,
     #                                                            105, 103, 101, 102, 104, 105))+
    stat_compare_means(label.y = 101) + theme(legend.position="none")
q<-ggboxplot(boxdat, x = "CellType", y = "meth_purity",
          color = "CellType", palette = colors1,
          xlab = "Cell type", ylab = "Estimated DNA methylation purity %") +
    #stat_compare_means(comparisons = my_comparisons, label.y = c(105, 103, 101, 102, 104,
     #                                                            105, 103, 101, 102, 104, 105))+
    stat_compare_means(label.y = 101) + theme(legend.position="none")
multiplot(p,q,cols = 1)

```

```{r Supplementary figures 2 and 3, fig.height=8, fig.width=12}
pheno2<-as.data.frame(pData(Ext_deconv_rgset)) 

#General QC and supplementary figures
#Genetic heatmap control SNPs
library(pheatmap)


pheno2$CellType<-ifelse(pheno2$CellType=="CD8T_naive", "CD8nv",
                         ifelse(pheno2$CellType=="Basophil", "Bas",
                                ifelse(pheno2$CellType=="CD8T_EM", "CD8mem",pheno2$CellType)))

colors1<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
Bnv = "#1B9E77", Bmem = "magenta",  
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black",  
CD8nv = "#7570B3", CD8mem = "maroon", 
 NK = "#E6AB02")
pheno2$meth_ethnicity2<-ifelse(pheno2$meth_ethnicity=="ASIAN", "Asian", 
                               ifelse(pheno2$meth_ethnicity=="WHITE", "White", 
                                      "Black"))
pheno2<-pheno2[pheno2$CellType%in%names(colors1) & pheno2$include!="No",]
ann_colors = list(CellType=colors1,
                  Sex = c(F="red", M="blue"),
                  Ethnicity=c("African-American"='black', "East-Asian"="firebrick", "Indo-European"="lightgray", Mixed="green"), SeSaMe_ethnicity=c("Black"='black', "Asian"="firebrick", "White"="lightgray"))

#levels(pheno2$CellType)<-c( "CD4T" , "CD8T" , "Bcell", "NK", "Neu" , "Mono" , "MIX"   )
annotation_row<-data.frame(CellType=pheno2$CellType, Sex=pheno2$Sex, Ethnicity= pheno2$Ethnicity_wide, 
                           SeSaMe_ethnicity=pheno2$meth_ethnicity2,
                           #factor(pheno2$CellType, levels = c( "CD4T" , "CD8T" , "Bcell", "NK", "Neu" , "Mono" , "MIX"   ))
                           row.names = rownames(pheno2))
pheatmap(rbind(getSnpBeta(Ext_deconv_rgset[,colnames(Ext_deconv_rgset)%in%rownames(pheno2)]),
               Ext_deconv_SNP[,colnames(Ext_deconv_SNP)%in%rownames(pheno2)]), annotation_col = annotation_row, show_colnames = F, show_rownames = F, annotation_colors = ann_colors)

#Table one
library(tableone)
factorvars2<-c("CellType","Sex", "bmi_clas", "Ethnicity_wide", "Ethnic_self", "smoker")
design5<-as.data.frame(pheno2)
design5[factorvars2] <- lapply(design5[factorvars2], factor)

contvars<-c("Age","weight_kg", "height_cm", "bmi", "purity")
design5[contvars] <- lapply(design5[contvars], as.numeric)

tableOne <- CreateTableOne(vars = c(factorvars2, contvars), data = design5)

print(tableOne,  quote = TRUE)


#Principal component regression analysis
cov<-data.frame(CellType=factor(pheno2$CellType),
    sex=factor(pheno2$Sex),
    slide=factor(pheno2$Slide),
    #array=factor(pheno2$Array),
    Ethnicity=pheno2$Ethnicity_wide,
    Age=pheno2$Age,
    weight=pheno2$weight_kg,
    height=pheno2$height_cm,
    BMI=pheno2$bmi,
    Smoker=factor(pheno2$smoker),
    purity=pheno2$purity)
#Only those analyzed
betas2<-Ext_deconv_betas3[,colnames(Ext_deconv_betas3)%in%rownames(pheno2)]
betas3<-betas2[complete.cases(betas2),]
#Eliminate 0 variance
which(apply(betas3, 1, var)==0)#2 probes 0 variance
betas3<-betas3[apply(betas3, 1, var) != 0,]
dim(betas3)#[1] 866089     49 2 probes deleted
pcrplot(betas3, cov, npc=20)

npc=50
npc <- min(ncol(betas3), npc)
svd <- prcomp(t(betas3), center = TRUE, scale = TRUE, retx = TRUE,
              na.action = "na.omit")

screeplot(svd, npc, type = "barplot")

eigenvalue <- svd[["sdev"]]^2
prop <- (sum(eigenvalue[1:npc])/sum(eigenvalue)) * 100
cat("Top ", npc, " principal components can explain ", prop,
    "% of data \n    variation", "\n")

for(i in 1:20){
  prop <- (sum(eigenvalue[i])/sum(eigenvalue)) * 100
prop1 <- (sum(eigenvalue[1:i])/sum(eigenvalue)) * 100
print(cat("Top ", i, " principal components can explain ", prop1,
    "% of data \n    variation", "\n the",i, "PC explains", prop , 
    "% of data \n    variation"))
}

p <- ENmix:::lmmatrix(svd$x[, 1:npc], cov)
yaxis <- colnames(p)
plotp2<-function ( p, yaxis, xmax, title)
{
    par(mar = c(5, 4.2, 4, 2) + 0.1)
    plot(1, xlim = c(0, xmax), ylim = c(0, length(yaxis) + 1),
         type = "n", bty = "n", axes = FALSE, xlab = "Principal Component",
         ylab = "", main = title)
    axis(1, at = c(1:xmax), pos = 0.5, las = 1, lwd = 3)
    for (i in 1:length(yaxis)) {
        text(0.3, i, yaxis[i], xpd = TRUE, adj = 1)
    }
    for (i in 1:ncol(p)) {
        for (j in 1:nrow(p)) {
            pp <- p[j, i]
            colcode <- "white"
            if (pp <= 1e-09) {
                colcode = "darkred"
            }
            else if (pp <= 1e-04) {
                colcode = "red"
            }
            else if (pp <= 0.01) {
                colcode = "orange"
            }
            else if (pp <= 0.05) {
                colcode = "pink"
            }
            polygon(c(j - 0.5, j - 0.5, j + 0.5, j + 0.5), c(i -
                                                                 0.5, i + 0.5, i + 0.5, i - 0.5), col = colcode,
                    border = NA)
        }
    }
    legend("topright", c("<0.05", "<0.01", "<10E-5", "<10E-10"),
           col = c("pink", "orange", "red", "darkred"), pch = 15,
           pt.cex = 2, bty = "o", horiz = TRUE, xpd = TRUE)
}
plotp2(p, yaxis, 20, title = "Principal Component Regression Analysis")

```

```{r IDOL fix Figure 1, fig.height=9, fig.width=12}
cellTypes = c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg")
MIXbetas<-FlowSorted.BloodExtended.EPIC[, FlowSorted.BloodExtended.EPIC$CellType=="MIX"]
trainingBetas<-getBeta(MIXbetas)[,c(1:6)]
trainingCovariates<-as.data.frame(pData(MIXbetas))[c(1:6),cellTypes]*100
testingBetas<-getBeta(MIXbetas)[,c(7:12)]
testingCovariates<-as.data.frame(pData(MIXbetas))[c(7:12),cellTypes]*100

groups<-c(Bas = "#E41A1C",  
          Bmem = "magenta", Bnv = "#1B9E77", 
CD4mem = "darkblue", CD4nv = "#D95F02", 
CD8mem = "maroon", 
CD8nv = "#7570B3", Eos = "#3E8E93",  
Mono = "#E7298A", 
Neu = "#66A61E", NK = "#E6AB02", 
Treg = "black")#, 
#WBC = "lightblue")

# for (i in c(seq(from = 250, to = 1050, by =50), seq(from = 1100, to = 2000, by =100),
#             seq(from = 2000, to = 3000, by =500))){
#   set.seed(1234)
#   assign(paste0("IDOLopt_fix_", i), IDOLoptimize2(candObject, trainingBetas, trainingCovariates,
# 	                    libSize = i, maxIt = 500, numCores = 5))
# }
# 
# 
# save(IDOLopt_fix_250, IDOLopt_fix_300, IDOLopt_fix_400, IDOLopt_fix_450, IDOLopt_fix_500, IDOLopt_fix_550, 
#      IDOLopt_fix_600, IDOLopt_fix_650, IDOLopt_fix_700, IDOLopt_fix_750, IDOLopt_fix_800, IDOLopt_fix_850,
#      IDOLopt_fix_900, IDOLopt_fix_950, 
#      IDOLopt_fix_1000, IDOLopt_fix_1050, IDOLopt_fix_1100, IDOLopt_fix_1200, IDOLopt_fix_1300, IDOLopt_fix_1400,
#      IDOLopt_fix_1500, IDOLopt_fix_1600, IDOLopt_fix_1700, IDOLopt_fix_1800, IDOLopt_fix_1900, IDOLopt_fix_2000,
#      IDOLopt_fix_2500, IDOLopt_fix_3000, file="IDOLopt_fix_EPIC.rda")


#IDOLOptimizedCpGsBloodExtended<-IDOLOptimizedCpGsBloodExtended

#Raw selection without filtering is unfair as the default universe is different
minfipickcomp<- minfi:::pickCompProbes(preprocessRaw(FlowSorted.BloodExtended.EPIC),cellTypes = c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg"),numProbes = 50, probeSelect = "auto")


table(rownames(minfipickcomp$coefEsts)%in%IDOLOptimizedCpGsBloodExtended)

#USe the same universe
minfipickcomp2<- minfi:::pickCompProbes(preprocessRaw(FlowSorted.BloodExtended.EPIC)[rownames(Ext_deconv_betas3),],cellTypes = c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg"),numProbes = 50, probeSelect = "auto")


table(rownames(minfipickcomp2$coefEsts)%in%IDOLOptimizedCpGsBloodExtended)

# for (i in c(seq(from = 250, to = 1050, by =50), seq(from = 1100, to = 2000, by =100),
#             seq(from = 2000, to = 3000, by =500))){
#   load(paste0("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/Fix/IDOL optimized DMR library_", i, ".RData"))
#   assign(paste0("IDOLopt_fix_", i), get("IDOL.optim.coefEsts"))
# }

# load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/IDOLopt_fix_EPIC.rda")
# 
# matrixEPICIDOL<-matrix(data = NA, nrow=28, ncol = 2, 
#                       dimnames = list(c(paste0("CpG_",
#                                                c(250,300,seq(from = 400, to = 1100, by =50),
#                                                  seq(from = 1200, to = 2000, by =100),
#                                                  2500,3000))), 
#                                         c("RMSE", "R2")))
# for (i in  c(seq(from = 400, to = 1100, by =50),
#                                                  seq(from = 1200, to = 2000, by =100),
#                                                  2500,3000)){
#   IDOLtmp<-get(paste0("IDOLopt_fix_",i))
#   for (j in 1:100){
#     if (j==1){
#       RMSE=RMSE.null=IDOLtmp$RMSE[j]
#       R2=R2.null=IDOLtmp$R2[j]
#     }else{
#       if (IDOLtmp$RMSE[j] <= RMSE.null & IDOLtmp$R2[j] >= R2.null) {
#             RMSE.null = IDOLtmp$RMSE[j]
#             R2.null = IDOLtmp$R2[j]
#             #print(paste("Iteration: ", j, " RMSE=", 
#             #    round(IDOLtmp$RMSE[j], 3), "; R2=", round(IDOLtmp$R2[j], 3), 
#             #    sep = ""))
#         }
#     }
#   }
#   print(paste("The Average RMSE = ", round(RMSE.null, 
#         3), " and R2 = ", round(R2.null, 3), " for the IDOL Optimized Library ",i, 
#         sep = ""))
#   matrixEPICIDOL[paste0("CpG_",i), 1]<-RMSE.null
#   matrixEPICIDOL[paste0("CpG_",i), 2]<-R2.null
# }

#####################################################################################
# Step 3: Using the IDOL optimized library identified in Step 2, deconvolute the 
# training data set and compare the observed versus predicted cell proportions 
#####################################################################################

# 3a: subset the training data to consist of the IDOL optimized library identified
# from Step 2
# trainingBetas.sub = trainingBetas[IDOLopt_fix_1200[["IDOL Optimized Library"]],]
# testingBetas.sub=testingBetas[IDOLopt_fix_1200[["IDOL Optimized Library"]],]
# # 3b: extract the coefEsts object from the Step 2 results
# coefEsts = IDOLopt_fix_1200[["IDOL Optimized CoefEsts"]]


trainingBetas.sub = trainingBetas[IDOLOptimizedCpGsBloodExtended,]
testingBetas.sub=testingBetas[IDOLOptimizedCpGsBloodExtended,]
# 3b: extract the coefEsts object from the Step 2 results
coefEsts = FlowSorted.BloodExtended.EPIC.compTable

colnames(coefEsts)<-c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg")
# 3c: predict cell composistion
predictedCellFraction = round(projectWBCnew(trainingBetas.sub, coefEsts)*100,1)
predictedCellFraction1 = round(projectWBCnew(testingBetas.sub, coefEsts)*100,1)
# 3d: do a visual depiction of the results (run all of the code below at once)
cellcolors = groups#brewer.pal(6,"Set2")
cellTypes<-c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg")
K = length(cellTypes)  # number of cell types

par(mfrow = c(3,4))
for(k in 1:K) {
	pred = c(as.numeric(predictedCellFraction[,cellTypes[k]]),
	         as.numeric(predictedCellFraction1[,cellTypes[k]]))
	obs = c(as.numeric(trainingCovariates[,cellTypes[k]]), 
	        as.numeric(testingCovariates[,cellTypes[k]]))
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))

	rangexy = c(0,25)#max(c(pred,obs)))#range(c(pred,obs))

	#par(mar = c(5,5,4,2))
	plot(obs[1:6], pred[1:6], cex = 2, 
	     pch = 21,  col="black", bg= cellcolors[cellTypes[k]],  
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "", ylab = "",
		cex.main = 2, cex.lab = 1.2, cex.axis = 1.3, frame.plot = FALSE, xaxt='n', yaxt='n')
		par(new=T)
	plot(obs[6:12], pred[6:12], cex = 2, 
	     pch = 22,  col="black", bg= cellcolors[cellTypes[k]], 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.2, cex.axis = 1.3, frame.plot = FALSE)
	#axis(side = 1)
	title(main = substitute(paste(o, " RMSE = ", 
		  m, "; ",R^2,"= ", n, sep = ""), 
		  list(o=cellTypes[k], m=round(RMSE, 2), n=round(R2, 2))), 
	      cex.main = 1.5)
	lm.D9 <- lm(pred~obs)
	abline(lm.D9, col=cellcolors[cellTypes[k]])
	abline(a = 0, b=1, col="black", lwd=1, lty=2)
	
  
}

#Comparison with pickcompprobes (minfi)
# 3a: subset the training data 
trainingBetas.sub = trainingBetas[rownames(minfipickcomp$coefEsts),]
testingBetas.sub=testingBetas[rownames(minfipickcomp$coefEsts),]
# 3b: extract the coefEsts object from the Step 2 results
coefEsts = minfipickcomp$coefEsts
colnames(coefEsts)<-c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg")
# 3c: predict cell composistion
predictedCellFraction = round(projectWBCnew(trainingBetas.sub, coefEsts)*100,1)
predictedCellFraction1 = round(projectWBCnew(testingBetas.sub, coefEsts)*100,1)
# 3d: do a visual depiction of the results (run all of the code below at once)
cellcolors = groups#brewer.pal(6,"Set2")
cellTypes<-c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg")
K = length(cellTypes)  # number of cell types

par(mfrow = c(3,4))
for(k in 1:K) {
	pred = c(as.numeric(predictedCellFraction[,cellTypes[k]]),
	         as.numeric(predictedCellFraction1[,cellTypes[k]]))
	obs = c(as.numeric(trainingCovariates[,cellTypes[k]]), 
	        as.numeric(testingCovariates[,cellTypes[k]]))
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))

	rangexy = c(0,25)#max(c(pred,obs)))#range(c(pred,obs))

	#par(mar = c(5,5,4,2))
	plot(obs[1:6], pred[1:6], cex = 2, 
	     pch = 21,  col="black", bg= cellcolors[cellTypes[k]],  
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "", ylab = "",
		cex.main = 2, cex.lab = 1.2, cex.axis = 1.3, frame.plot = FALSE, xaxt='n', yaxt='n')
		par(new=T)
	plot(obs[6:12], pred[6:12], cex = 2, 
	     pch = 22,  col="black", bg= cellcolors[cellTypes[k]], 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.2, cex.axis = 1.3, frame.plot = FALSE)
	#axis(side = 1)
	title(main = substitute(paste(o, " RMSE = ", 
		  m, "; ",R^2,"= ", n, sep = ""), 
		  list(o=cellTypes[k], m=round(RMSE, 2), n=round(R2, 2))), 
	      cex.main = 1.5)
	lm.D9 <- lm(pred~obs)
	abline(lm.D9, col=cellcolors[cellTypes[k]])
	abline(a = 0, b=1, col="black", lwd=1, lty=2)
	
  
}


#minfi subset
#Comparison with pickcompprobes (minfi)
# 3a: subset the training data 
trainingBetas.sub = trainingBetas[rownames(minfipickcomp2$coefEsts),]
testingBetas.sub=testingBetas[rownames(minfipickcomp2$coefEsts),]
# 3b: extract the coefEsts object from the Step 2 results
coefEsts = minfipickcomp$coefEsts
colnames(coefEsts)<-c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg")
# 3c: predict cell composistion
predictedCellFraction = round(projectWBCnew(trainingBetas.sub, coefEsts)*100,1)
predictedCellFraction1 = round(projectWBCnew(testingBetas.sub, coefEsts)*100,1)
# 3d: do a visual depiction of the results (run all of the code below at once)
cellcolors = groups#brewer.pal(6,"Set2")
cellTypes<-c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg")
K = length(cellTypes)  # number of cell types

par(mfrow = c(3,4))
for(k in 1:K) {
	pred = c(as.numeric(predictedCellFraction[,cellTypes[k]]),
	         as.numeric(predictedCellFraction1[,cellTypes[k]]))
	obs = c(as.numeric(trainingCovariates[,cellTypes[k]]), 
	        as.numeric(testingCovariates[,cellTypes[k]]))
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))

	rangexy = c(0,25)#max(c(pred,obs)))#range(c(pred,obs))

	#par(mar = c(5,5,4,2))
	plot(obs[1:6], pred[1:6], cex = 2, 
	     pch = 21,  col="black", bg= cellcolors[cellTypes[k]],  
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "", ylab = "",
		cex.main = 2, cex.lab = 1.2, cex.axis = 1.3, frame.plot = FALSE, xaxt='n', yaxt='n')
		par(new=T)
	plot(obs[6:12], pred[6:12], cex = 2, 
	     pch = 22,  col="black", bg= cellcolors[cellTypes[k]], 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.2, cex.axis = 1.3, frame.plot = FALSE)
	#axis(side = 1)
	title(main = substitute(paste(o, " RMSE = ", 
		  m, "; ",R^2,"= ", n, sep = ""), 
		  list(o=cellTypes[k], m=round(RMSE, 2), n=round(R2, 2))), 
	      cex.main = 1.5)
	lm.D9 <- lm(pred~obs)
	abline(lm.D9, col=cellcolors[cellTypes[k]])
	abline(a = 0, b=1, col="black", lwd=1, lty=2)
	
  
}



#450k
#load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/IDOL optimized DMR library_1500.RData")


# 3a: subset the training data 
# trainingBetas.sub = trainingBetas[IDOLOptimizedCpGsBloodExtended450k,]
# testingBetas.sub=testingBetas[IDOLOptimizedCpGsBloodExtended450k,]
# # 3b: extract the coefEsts object from the Step 2 results
# coefEsts = FlowSorted.BloodExtended.450klegacy.compTable


trainingBetas.sub = trainingBetas[IDOLOptimizedCpGsBloodExtended450k,]
testingBetas.sub=testingBetas[IDOLOptimizedCpGsBloodExtended450k,]
# 3b: extract the coefEsts object from the Step 2 results
coefEsts = FlowSorted.BloodExtended.450klegacy.compTable


colnames(coefEsts)<-c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg")
# 3c: predict cell composistion
predictedCellFraction = round(projectWBCnew(trainingBetas.sub, coefEsts)*100,1)
predictedCellFraction1 = round(projectWBCnew(testingBetas.sub, coefEsts)*100,1)
# 3d: do a visual depiction of the results (run all of the code below at once)
cellcolors = groups#brewer.pal(6,"Set2")
cellTypes<-c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg")
K = length(cellTypes)  # number of cell types

par(mfrow = c(3,4))
for(k in 1:K) {
	pred = c(as.numeric(predictedCellFraction[,cellTypes[k]]),
	         as.numeric(predictedCellFraction1[,cellTypes[k]]))
	obs = c(as.numeric(trainingCovariates[,cellTypes[k]]), 
	        as.numeric(testingCovariates[,cellTypes[k]]))
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))

	rangexy = c(0,25)#max(c(pred,obs)))#range(c(pred,obs))

	#par(mar = c(5,5,4,2))
	plot(obs[1:6], pred[1:6], cex = 2, 
	     pch = 21,  col="black", bg= cellcolors[cellTypes[k]],  
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "", ylab = "",
		cex.main = 2, cex.lab = 1.2, cex.axis = 1.3, frame.plot = FALSE, xaxt='n', yaxt='n')
		par(new=T)
	plot(obs[6:12], pred[6:12], cex = 2, 
	     pch = 22,  col="black", bg= cellcolors[cellTypes[k]], 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.2, cex.axis = 1.3, frame.plot = FALSE)
	#axis(side = 1)
	title(main = substitute(paste(o, " RMSE = ", 
		  m, "; ",R^2,"= ", n, sep = ""), 
		  list(o=cellTypes[k], m=round(RMSE, 2), n=round(R2, 2))), 
	      cex.main = 1.5)
	lm.D9 <- lm(pred~obs)
	abline(lm.D9, col=cellcolors[cellTypes[k]])
	abline(a = 0, b=1, col="black", lwd=1, lty=2)
	
  
}




#Comparison vs true values
#pheno<-as.data.frame(pData(RGsetTargets)) 
#pheno$clas<-paste("Method", substr(pheno$Subject.ID, 1,1))


#Remember the order for the colors look for the order if the results do not match the figure in the paper.
#Figure 1

par(mfrow=c(1,2))
barplot(t(trainingCovariates), col= cellcolors[cellTypes], names.arg = seq(1:6), xlab = "Training")
barplot(t(testingCovariates), col= cellcolors[cellTypes], names.arg = seq(1:6), #legend.text = cellTypes, 
        xlab = "Testing")


#Only training
####################################################################################
# Step 3: Using the IDOL optimized library identified in Step 2, deconvolute the 
# training data set and compare the observed versus predicted cell proportions 
#####################################################################################

# 3a: subset the training data to consist of the IDOL optimized library identified
# from Step 2
# trainingBetas.sub = trainingBetas[IDOLopt_fix_1200[["IDOL Optimized Library"]],]
# testingBetas.sub=testingBetas[IDOLopt_fix_1200[["IDOL Optimized Library"]],]
# # 3b: extract the coefEsts object from the Step 2 results
# coefEsts = IDOLopt_fix_1200[["IDOL Optimized CoefEsts"]]


trainingBetas.sub = trainingBetas[IDOLOptimizedCpGsBloodExtended,]
#testingBetas.sub=testingBetas[IDOLOptimizedCpGsBloodExtended,]
# 3b: extract the coefEsts object from the Step 2 results
coefEsts = FlowSorted.BloodExtended.EPIC.compTable

colnames(coefEsts)<-c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg")
# 3c: predict cell composistion
predictedCellFraction = round(projectWBCnew(trainingBetas.sub, coefEsts)*100,1)
#predictedCellFraction1 = round(projectWBCnew(testingBetas.sub, coefEsts)*100,1)
# 3d: do a visual depiction of the results (run all of the code below at once)
cellcolors = groups#brewer.pal(6,"Set2")
cellTypes<-c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg")
K = length(cellTypes)  # number of cell types

par(mfrow = c(3,4))
for(k in 1:K) {
	pred = as.numeric(predictedCellFraction[,cellTypes[k]])
	         #as.numeric(predictedCellFraction1[,cellTypes[k]]))
	obs = as.numeric(trainingCovariates[,cellTypes[k]]) 
	        #as.numeric(testingCovariates[,cellTypes[k]]))
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))

	rangexy = c(0,25)#max(c(pred,obs)))#range(c(pred,obs))

	#par(mar = c(5,5,4,2))
	# plot(obs[1:6], pred[1:6], cex = 2, 
	#      pch = 21,  col="black", bg= cellcolors[cellTypes[k]],  
	# 	xlim = rangexy, ylim = rangexy, main = "",
	# 	xlab = "", ylab = "",
	# 	cex.main = 2, cex.lab = 1.2, cex.axis = 1.3, frame.plot = FALSE, xaxt='n', yaxt='n')
	# 	par(new=T)
	plot(obs[1:6], pred[1:6], cex = 2, 
	     pch = 21,  col="black", bg= cellcolors[cellTypes[k]], 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.2, cex.axis = 1.3, frame.plot = FALSE)
	title(main = substitute(paste(o, " RMSE = ", 
		  m, "; ",R^2,"= ", n, sep = ""), 
		  list(o=cellTypes[k], m=round(RMSE, 2), n=round(R2, 2))), 
	      cex.main = 1.5)
	lm.D9 <- lm(pred~obs)
	abline(lm.D9, col=cellcolors[cellTypes[k]])
	abline(a = 0, b=1, col="black", lwd=1, lty=2)
	
  
}


#Only testing

testingBetas.sub=testingBetas[IDOLOptimizedCpGsBloodExtended,]

coefEsts = FlowSorted.BloodExtended.EPIC.compTable

colnames(coefEsts)<-c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg")
# 3c: predict cell composistion
predictedCellFraction = round(projectWBCnew(trainingBetas.sub, coefEsts)*100,1)

# 3d: do a visual depiction of the results (run all of the code below at once)
cellcolors = groups#brewer.pal(6,"Set2")
cellTypes<-c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
"Eos", "Mono", "Neu", "NK", "Treg")
K = length(cellTypes)  # number of cell types

par(mfrow = c(3,4))
for(k in 1:K) {
	pred = as.numeric(predictedCellFraction1[,cellTypes[k]])
	obs = as.numeric(testingCovariates[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))

	rangexy = c(0,25)#max(c(pred,obs)))#range(c(pred,obs))

	plot(obs[1:6], pred[1:6], cex = 2, 
	     pch = 22,  col="black", bg= cellcolors[cellTypes[k]], 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.2, cex.axis = 1.3, frame.plot = FALSE)
	title(main = substitute(paste(o, " RMSE = ", 
		  m, "; ",R^2,"= ", n, sep = ""), 
		  list(o=cellTypes[k], m=round(RMSE, 2), n=round(R2, 2))), 
	      cex.main = 1.5)
	lm.D9 <- lm(pred~obs)
	abline(lm.D9, col=cellcolors[cellTypes[k]])
	abline(a = 0, b=1, col="black", lwd=1, lty=2)
}


#Validation
load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/Validation/Raw_data_mixval.RData")
load(paste0(workdir,"/Validation/Validation_deconv_sesame.RData"))

cellcolors<-c(Bas = "#E41A1C", 
          Bmem = "magenta", Bnv = "#1B9E77", 
CD4mem = "darkblue", CD4nv = "#D95F02", 
CD8mem = "maroon", 
CD8nv = "#7570B3", Eos = "#3E8E93", 
Mono = "#E7298A", 
Neu = "#66A61E", NK = "#E6AB02", 
Treg = "black")

cellTypes<-names(cellcolors)


trainingCovariates<-as.data.frame(pData(Validation_deconv_rgset))[Validation_deconv_rgset$CellType=="MIX",cellTypes]*100
trainingCovariates<-round(trainingCovariates,1)




#noob minfi
trainingBetas.sub = getBeta(MsetEx)[IDOLOptimizedCpGsBloodExtended,
                                                     Validation_deconv_rgset$CellType=="MIX"]


# 3b: extract the coefEsts object from the Step 2 results
coefEsts = FlowSorted.BloodExtended.EPIC.compTable

#colnames(coefEsts)<-c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", "CD8mem", "CD8nv", 
#"Eos", "Mono", "Neu", "NK", "Treg")
# 3c: predict cell composistion
predictedCellFraction = round(projectWBCnew(trainingBetas.sub, coefEsts)*100,1)

# 3d: do a visual depiction of the results (run all of the code below at once)

K = length(cellTypes)  # number of cell types



#Artificial mixtures
par(mfrow = c(3,4))
for(k in 1:K) {
	pred = as.numeric(predictedCellFraction[,cellTypes[k]])
	obs = as.numeric(trainingCovariates[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))

	rangexy = c(0,25)#max(c(pred,obs)))#range(c(pred,obs))

	#par(mar = c(5,5,4,2))
	
	plot(obs, pred, cex = 2, 
	     pch = 21,  col="black", bg= cellcolors[cellTypes[k]], 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.2, cex.axis = 1.3, frame.plot = FALSE)
	#axis(side = 1)
	title(main = substitute(paste(o, " RMSE = ", 
		  m, "; ",R^2,"= ", n, sep = ""), 
		  list(o=cellTypes[k], m=round(RMSE, 2), n=round(R2, 2))), 
	      cex.main = 1.5)
	lm.D9 <- lm(pred~obs)
	abline(lm.D9, col=cellcolors[cellTypes[k]])
	abline(a = 0, b=1, col="black", lwd=1, lty=2)
	
  
}


par(mfrow=c(1,1))
barplot(t(trainingCovariates), col= cellcolors[cellTypes], names.arg = seq(1:12), xlab = "Replication")




```

```{r Figure 3a Suppl table 5, fig.height=11, fig.width=11}
#IDOLopt_EM_2500$`IDOL Optimized CoefEsts`
cellTypes = c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", #"CD8T_CM", 
              "CD8mem", 
"CD8nv", "Eos", #"MIX", 
"Mono", "Neu", "NK", #"NKT", 
"Treg")#,"WBC")
coefEsts = FlowSorted.BloodExtended.EPIC.compTable

#load(file = "I:/Dropbox (Christensen Lab)/manifest EPIC/annotationEPICb5.rda")
load("D:/OneDrive/OneDrive - Dartmouth College/Crossreactive/annotationEPICb5updated.rda")
annot<-annot[rownames(coefEsts),]
dim(annot)
matcells5<-cbind.data.frame(coefEsts, annot)
write.csv(matcells5, file="Suppl_table_5.csv")

#Extract the first reference group
anno4<-strsplit(matcells5$UCSC_RefGene_Group, ";")
anno5a<-lapply(anno4, function(x) if(identical(x, NA_character_)) "Intergenic" else x)
anno5<-lapply(anno5a, `[[`, 1)

matcells5$Function = factor(unlist(anno5), levels= c("TSS1500", "TSS200", "1stExon","ExonBnd", "5'UTR","Body", "3'UTR", "Intergenic"))

matcells5$CGI<-ifelse(matcells5$Relation_to_UCSC_CpG_Island=="","OpenSea",matcells5$Relation_to_UCSC_CpG_Island)
matcells5$CGI2<-ifelse(matcells5$CGI=="N_Shelf"| matcells5$CGI=="S_Shelf","Shelves",
                       ifelse(matcells5$CGI=="N_Shore"| matcells5$CGI=="S_Shore","Shores",
                              matcells5$CGI))

matcells5$Enhancer<-ifelse(matcells5$Phantom5_Enhancers=="", "No", "Yes")
matcells5$DHS<-ifelse(matcells5$DNase_Hypersensitivity_NAME=="", "No", "Yes")
matcells5$OpenChromatin<-ifelse(matcells5$OpenChromatin_NAME=="", "No", "Yes")
matcells5$TFBS<-ifelse(matcells5$TFBS_NAME=="", "No", "Yes")
table(matcells5$CGI)
table(matcells5$Function)
table(matcells5$Enhancer)
table(matcells5$DHS)
table(matcells5$OpenChromatin)
table(matcells5$TFBS)
table(matcells5$Methyl450_Loci)
matcells5$Methyl450_Loci2<-ifelse(matcells5$Methyl450_Loci==TRUE, "Yes", "No")
annotation_row= data.frame(CGI=matcells5$CGI2, Function= matcells5$Function, Enhancer=matcells5$Enhancer, DHS=matcells5$DHS,
                           OpenChromatin=matcells5$OpenChromatin, TFBS=matcells5$TFBS, HM450k=matcells5$Methyl450_Loci2,
                           
                           row.names = rownames(coefEsts))
annotation_col=data.frame(CellType=colnames(coefEsts), row.names = colnames(coefEsts))
groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
#Bcell = "gold", 
Bnv = "#1B9E77", Bmem = "magenta", #CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", #CD8T = "darkgray", 
CD8nv = "#7570B3", CD8mem = "maroon", #CD8T_CM = "#705C83", 
#NKT = "#7E6E85", 
NK = "#E6AB02")#, WBC = "white")
ann_colors = list(
    CellType = groups, #CGI=c(No="white", Yes="darkgray"), 
    Enhancer=c(No="white", Yes="darkgray"), 
    DHS=c(No="white", Yes="darkgray"),
                           OpenChromatin=c(No="white", Yes="darkgray"), TFBS=c(No="white", Yes="darkgray"),
    HM450k=c(No="white", Yes="darkgray"), 
    Function= c("TSS1500"="#D53E4F", "TSS200"="#F46D43",
            "1stExon"= "#FDAE61","ExonBnd"="#FEE08B",
            "5'UTR"="#E6F598","Body"="#ABDDA4",
            "3'UTR"= "#66C2A5", "Intergenic"="#3288BD"),
    CGI<-c("Island"= "#E41A1C", "Shores"="#377EB8",
       "Shelves"="#4DAF4A", "OpenSea"="#984EA3")
) 


try(pheatmap(coefEsts,color = colorRampPalette(c("yellow", "black", "blue"), space = "Lab")(128), annotation_colors = ann_colors,
                  annotation_row = annotation_row, 
             annotation_col = annotation_col, main = "L-DMR IDOL EPIC library", cluster_rows = T, show_rownames = F))#, labels_col = combined$CellType))


referencePd<-Ext_deconv_pheno[Ext_deconv_pheno$CellType%in%cellTypes,]
table(referencePd$CellType)
referencePd$CellType<-ifelse(referencePd$CellType=="Basophil", "Bas",
                             ifelse(referencePd$CellType=="CD8T_EM", "CD8mem",
                                    ifelse(referencePd$CellType=="CD8T_naive", "CD8nv", referencePd$CellType)))
referenceBetas<-Ext_deconv_betas3[rownames(Ext_deconv_betas3), colnames(Ext_deconv_betas3)%in%rownames(referencePd)]
identical(colnames(referenceBetas), rownames(referencePd))
annotation_col=data.frame(CellType=referencePd$CellType, row.names = rownames(referencePd))
try(pheatmap(referenceBetas[rownames(referenceBetas)%in%rownames(coefEsts),],color = colorRampPalette(c("yellow", "black", "blue"), space = "Lab")(128), annotation_colors = ann_colors,
                  annotation_row = annotation_row, 
             annotation_col = annotation_col, main = "L-DMR IDOL EPIC library", cluster_rows = T, show_rownames = F, labels_col = referencePd$CellType))

```

```{r Figure 3c Supplementary table 7, fig.height=11, fig.width=11}
#IDOLopt_EM_2500$`IDOL Optimized CoefEsts`
cellTypes = c("Bas", "Bmem", "Bnv", "CD4mem", "CD4nv", #"CD8T_CM", 
              "CD8mem", 
"CD8nv", "Eos", #"MIX", 
"Mono", "Neu", "NK", #"NKT", 
"Treg")#,"WBC")
coefEsts = minfipickcomp$coefEsts


#load(file = "I:/Dropbox (Christensen Lab)/manifest EPIC/annotationEPICb5.rda")
load("D:/OneDrive/OneDrive - Dartmouth College/Crossreactive/annotationEPICb5updated.rda")
annot<-annot[rownames(coefEsts),]
dim(annot)
matcells5<-cbind.data.frame(coefEsts, annot)
#write.csv(matcells5, file="Suppl table 7.csv")

#Extract the first reference group
anno4<-strsplit(matcells5$UCSC_RefGene_Group, ";")
#character 0 should be transformed to NA or the lapply will fail
#anno5a<-lapply(anno4, function(x) if(identical(x, character(0))) NA_character_ else x)
#Replace for "Intergenic"
anno5a<-lapply(anno4, function(x) if(identical(x, NA_character_)) "Intergenic" else x)
anno5<-lapply(anno5a, `[[`, 1)

matcells5$Function = factor(unlist(anno5), levels= c("TSS1500", "TSS200", "1stExon","ExonBnd", "5'UTR","Body", "3'UTR", "Intergenic"))

matcells5$CGI<-ifelse(matcells5$Relation_to_UCSC_CpG_Island=="","OpenSea",matcells5$Relation_to_UCSC_CpG_Island)
matcells5$CGI2<-ifelse(matcells5$CGI=="N_Shelf"| matcells5$CGI=="S_Shelf","Shelves",
                       ifelse(matcells5$CGI=="N_Shore"| matcells5$CGI=="S_Shore","Shores",
                              matcells5$CGI))

matcells5$Enhancer<-ifelse(matcells5$Phantom5_Enhancers=="", "No", "Yes")
matcells5$DHS<-ifelse(matcells5$DNase_Hypersensitivity_NAME=="", "No", "Yes")
matcells5$OpenChromatin<-ifelse(matcells5$OpenChromatin_NAME=="", "No", "Yes")
matcells5$TFBS<-ifelse(matcells5$TFBS_NAME=="", "No", "Yes")
table(matcells5$CGI)
table(matcells5$Function)
table(matcells5$Enhancer)
table(matcells5$DHS)
table(matcells5$OpenChromatin)
table(matcells5$TFBS)
table(matcells5$Methyl450_Loci)
matcells5$Methyl450_Loci2<-ifelse(matcells5$Methyl450_Loci==TRUE, "Yes", "No")
annotation_row= data.frame(CGI=matcells5$CGI2, Function= matcells5$Function, Enhancer=matcells5$Enhancer, DHS=matcells5$DHS,
                           OpenChromatin=matcells5$OpenChromatin, TFBS=matcells5$TFBS, HM450k=matcells5$Methyl450_Loci2,
                           
                           row.names = rownames(coefEsts))
annotation_col=data.frame(CellType=colnames(coefEsts), row.names = colnames(coefEsts))
groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
#Bcell = "gold", 
Bnv = "#1B9E77", Bmem = "magenta", #CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", #CD8T = "darkgray", 
CD8nv = "#7570B3", CD8mem = "maroon", #CD8T_CM = "#705C83", 
#NKT = "#7E6E85", 
NK = "#E6AB02")#, WBC = "white")
ann_colors = list(
    CellType = groups, #CGI=c(No="white", Yes="darkgray"), 
    Enhancer=c(No="white", Yes="darkgray"), 
    DHS=c(No="white", Yes="darkgray"),
                           OpenChromatin=c(No="white", Yes="darkgray"), TFBS=c(No="white", Yes="darkgray"),
    HM450k=c(No="white", Yes="darkgray"), 
    Function= c("TSS1500"="#D53E4F", "TSS200"="#F46D43",
            "1stExon"= "#FDAE61","ExonBnd"="#FEE08B",
            "5'UTR"="#E6F598","Body"="#ABDDA4",
            "3'UTR"= "#66C2A5", "Intergenic"="#3288BD"),
    CGI<-c("Island"= "#E41A1C", "Shores"="#377EB8",
       "Shelves"="#4DAF4A", "OpenSea"="#984EA3")
) 



try(pheatmap(coefEsts,color = colorRampPalette(c("yellow", "black", "blue"), space = "Lab")(128), annotation_colors = ann_colors,
                  annotation_row = annotation_row, 
             annotation_col = annotation_col, main = "L-DMR pickcompprobes EPIC library", cluster_rows = T, show_rownames = F))#, labels_col = combined$CellType))


referencePd<-Ext_deconv_pheno[Ext_deconv_pheno$CellType%in%cellTypes,]
table(referencePd$CellType)
referencePd$CellType<-ifelse(referencePd$CellType=="Basophil", "Bas",
                             ifelse(referencePd$CellType=="CD8T_EM", "CD8mem",
                                    ifelse(referencePd$CellType=="CD8T_naive", "CD8nv", referencePd$CellType)))
referenceBetas<-Ext_deconv_betas3[rownames(Ext_deconv_betas3), colnames(Ext_deconv_betas3)%in%rownames(referencePd)]
identical(colnames(referenceBetas), rownames(referencePd))
annotation_col=data.frame(CellType=referencePd$CellType, row.names = rownames(referencePd))
try(pheatmap(referenceBetas[rownames(referenceBetas)%in%rownames(coefEsts),],color = colorRampPalette(c("yellow", "black", "blue"), space = "Lab")(128), annotation_colors = ann_colors,
                  annotation_row = annotation_row, 
             annotation_col = annotation_col, main = "L-DMR pickcompprobes EPIC library", cluster_rows = T, show_rownames = F, labels_col = referencePd$CellType))

```

```{r Table 1, fig.width=11, fig.height=11 }
#load(file = "I:/Dropbox (Christensen Lab)/manifest EPIC/annotationEPICb5.rda")
load("D:/OneDrive/OneDrive - Dartmouth College/Crossreactive/annotationEPICb5updated.rda")
annot<-annot[IDOLOptimizedCpGs,]
dim(annot)
matcells5<-cbind.data.frame(IDOLOptimizedCpGs.compTable, annot)
#write.csv(matcells5, file="Minfi_library_1200_EPIC.csv")

#Extract the first reference group
anno4<-strsplit(matcells5$UCSC_RefGene_Group, ";")
#character 0 should be transformed to NA or the lapply will fail
#anno5a<-lapply(anno4, function(x) if(identical(x, character(0))) NA_character_ else x)
#Replace for "Intergenic"
anno5a<-lapply(anno4, function(x) if(identical(x, NA_character_)) "Intergenic" else x)
anno5<-lapply(anno5a, `[[`, 1)

matcells5$Function = factor(unlist(anno5), levels= c("TSS1500", "TSS200", "1stExon","ExonBnd", "5'UTR","Body", "3'UTR", "Intergenic"))

matcells5$CGI<-ifelse(matcells5$Relation_to_UCSC_CpG_Island=="","OpenSea",matcells5$Relation_to_UCSC_CpG_Island)
matcells5$Enhancer<-ifelse(matcells5$Phantom5_Enhancers=="", "No", "Yes")
matcells5$DHS<-ifelse(matcells5$DNase_Hypersensitivity_NAME=="", "No", "Yes")
matcells5$OpenChromatin<-ifelse(matcells5$OpenChromatin_NAME=="", "No", "Yes")
matcells5$TFBS<-ifelse(matcells5$TFBS_NAME=="", "No", "Yes")
table(matcells5$CGI)
table(matcells5$Function)
table(matcells5$Enhancer)
table(matcells5$DHS)
table(matcells5$OpenChromatin)
table(matcells5$TFBS)
table(matcells5$Methyl450_Loci)


#OVerlap
table(rownames(minfipickcomp$coefEsts)%in%IDOLOptimizedCpGsBloodExtended)
table(rownames(minfipickcomp$coefEsts)%in%IDOLOptimizedCpGsBloodExtended450k)
table(rownames(minfipickcomp$coefEsts)%in%IDOLOptimizedCpGs)

table(IDOLOptimizedCpGs %in% IDOLOptimizedCpGsBloodExtended)
table(IDOLOptimizedCpGs %in% IDOLOptimizedCpGsBloodExtended450k)
#table(rownames(minfipickcomp$coefEsts)%in%IDOLOptimizedCpGs)

#table(IDOLOptimizedCpGs %in% IDOLOptimizedCpGsBloodExtended)
table(IDOLOptimizedCpGsBloodExtended %in% IDOLOptimizedCpGsBloodExtended450k)
#table(rownames(minfipickcomp$coefEsts)%in%IDOLOptimizedCpGs)
```

```{r Suppl table 6 aging dataset, fig.height=11, fig.width=11}
load("D:/OneDrive/OneDrive - Dartmouth College/Fetal adult/Aging dataset2.RData")

coefEsts = FlowSorted.BloodExtended.450klegacy.compTable


#load(file = "I:/Dropbox (Christensen Lab)/manifest EPIC/annotationEPICb5.rda")
load("D:/OneDrive/OneDrive - Dartmouth College/Crossreactive/annotationEPICb5updated.rda")
annot<-annot[rownames(coefEsts),]
dim(annot)
matcells5<-cbind.data.frame(coefEsts, annot)
#write.csv(matcells5, file="Suppl table 6.csv")

#Extract the first reference group
anno4<-strsplit(matcells5$UCSC_RefGene_Group, ";")
#character 0 should be transformed to NA or the lapply will fail
#anno5a<-lapply(anno4, function(x) if(identical(x, character(0))) NA_character_ else x)
#Replace for "Intergenic"
anno5a<-lapply(anno4, function(x) if(identical(x, NA_character_)) "Intergenic" else x)
anno5<-lapply(anno5a, `[[`, 1)

matcells5$Function = factor(unlist(anno5), levels= c("TSS1500", "TSS200", "1stExon","ExonBnd", "5'UTR","Body", "3'UTR", "Intergenic"))

matcells5$CGI<-ifelse(matcells5$Relation_to_UCSC_CpG_Island=="","OpenSea",matcells5$Relation_to_UCSC_CpG_Island)
matcells5$CGI2<-ifelse(matcells5$CGI=="N_Shelf"| matcells5$CGI=="S_Shelf","Shelves",
                       ifelse(matcells5$CGI=="N_Shore"| matcells5$CGI=="S_Shore","Shores",
                              matcells5$CGI))

matcells5$Enhancer<-ifelse(matcells5$Phantom5_Enhancers=="", "No", "Yes")
matcells5$DHS<-ifelse(matcells5$DNase_Hypersensitivity_NAME=="", "No", "Yes")
matcells5$OpenChromatin<-ifelse(matcells5$OpenChromatin_NAME=="", "No", "Yes")
matcells5$TFBS<-ifelse(matcells5$TFBS_NAME=="", "No", "Yes")
table(matcells5$CGI)
table(matcells5$Function)
table(matcells5$Enhancer)
table(matcells5$DHS)
table(matcells5$OpenChromatin)
table(matcells5$TFBS)
table(matcells5$Methyl450_Loci)
matcells5$Methyl450_Loci2<-ifelse(matcells5$Methyl450_Loci==TRUE, "Yes", "No")
annotation_row= data.frame(CGI=matcells5$CGI2, Function= matcells5$Function, Enhancer=matcells5$Enhancer, DHS=matcells5$DHS,
                           OpenChromatin=matcells5$OpenChromatin, TFBS=matcells5$TFBS, HM450k=matcells5$Methyl450_Loci2,
                           
                           row.names = rownames(coefEsts))
annotation_col=data.frame(CellType=colnames(coefEsts), row.names = colnames(coefEsts))
groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
#Bcell = "gold", 
Bnv = "#1B9E77", Bmem = "magenta", #CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", #CD8T = "darkgray", 
CD8nv = "#7570B3", CD8mem = "maroon", #CD8T_CM = "#705C83", 
#NKT = "#7E6E85", 
NK = "#E6AB02")#, WBC = "white")
ann_colors = list(
    CellType = groups, #CGI=c(No="white", Yes="darkgray"), 
    Enhancer=c(No="white", Yes="darkgray"), 
    DHS=c(No="white", Yes="darkgray"),
                           OpenChromatin=c(No="white", Yes="darkgray"), TFBS=c(No="white", Yes="darkgray"),
    HM450k=c(No="white", Yes="darkgray"), 
    Function= c("TSS1500"="#D53E4F", "TSS200"="#F46D43",
            "1stExon"= "#FDAE61","ExonBnd"="#FEE08B",
            "5'UTR"="#E6F598","Body"="#ABDDA4",
            "3'UTR"= "#66C2A5", "Intergenic"="#3288BD"),
    CGI<-c("Island"= "#E41A1C", "Shores"="#377EB8",
       "Shelves"="#4DAF4A", "OpenSea"="#984EA3")
) 


try(pheatmap(coefEsts,color = colorRampPalette(c("yellow", "black", "blue"), space = "Lab")(128), annotation_colors = ann_colors,
                  annotation_row = annotation_row, 
             annotation_col = annotation_col, main = "L-DMRs IDOL legacy  450K", cluster_rows = T, show_rownames = F))#, labels_col = combined$CellType))
annotation_col=data.frame(CellType=referencePd$CellType, row.names = rownames(referencePd))


Ext_deconv_pheno$CellType<-ifelse(Ext_deconv_pheno$CellType=="Basophil", "Bas",
                             ifelse(Ext_deconv_pheno$CellType=="CD8T_EM", "CD8mem",
                                    ifelse(Ext_deconv_pheno$CellType=="CD8T_naive", "CD8nv", Ext_deconv_pheno$CellType)))

referencePd<-Ext_deconv_pheno[Ext_deconv_pheno$CellType%in%cellTypes,]
# referencePd$CellType<-ifelse(referencePd$CellType=="Basophil", "Bas",
#                              ifelse(referencePd$CellType=="CD8T_EM", "CD8mem",
#                                     ifelse(referencePd$CellType=="CD8T_naive", "CD8nv", referencePd$CellType)))
referenceBetas<-Ext_deconv_betas3[rownames(Ext_deconv_betas3)%in%rownames(annot[annot$Methyl450_Loci==TRUE,]), colnames(Ext_deconv_betas3)%in%rownames(referencePd)]
identical(colnames(referenceBetas), rownames(referencePd))

try(pheatmap(referenceBetas[rownames(referenceBetas)%in%rownames(coefEsts),],color = colorRampPalette(c("yellow", "black", "blue"), space = "Lab")(128), annotation_colors = ann_colors,
             #      annotation_row = annotation_row, 
             annotation_col = annotation_col, main = "IDOL legacy DMRs 450K", cluster_rows = T, show_rownames = F, labels_col = referencePd$CellType))




cellcountsage<-FlowSorted.Blood.EPIC:::projectCellType(betas_aging[rownames(coefEsts),], coefEsts, lessThanOne = TRUE)

sumage<-rowSums(cellcountsage)
cellcountsage2<-round(cellcountsage*100/sumage,1)
head(cellcountsage2)
pheno5<-cbind.data.frame(cellcountsage2, Age_groups=pheno_aging$Age_groups)
pheno5<-pheno5[order(pheno5$CD4mem),]
pheno5<-pheno5[order(pheno5$CD4nv),]
pheno5<-pheno5[order(pheno5$Neu),]

for (i in dput(levels(as.factor(pheno_aging$Age_groups)))){
    stack<-pheno5[pheno5$Age_groups==i,dput(colnames(coefEsts))]
    stack<-t(stack)
    dat2 <- melt(stack)
    colors1<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
Bcell = "gold", Bnv = "#1B9E77", Bmem = "magenta", CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", CD8T = "darkgray", 
CD8nv = "#7570B3", CD8mem = "maroon", 
NKT = "#7E6E85", NK = "#E6AB02")
    colors2<-colors1[levels(dat2$X1)]
    fetaladultspine<- ggplot(dat2, aes(x=X2, y=value, fill=X1)) + 
        geom_bar(stat="identity") +
        xlab("") +#xlab("\nSample") +
        ylab("") +#ylab("Cell %\n") + 
        geom_hline(yintercept=70, linetype="dashed", color = "red")+ 
        #guides(fill=T) +
        theme_bw()+
        ggtitle(paste("Cell projection extended in ", i)) + scale_fill_manual(values=colors2)+
        theme(axis.text.x=element_blank())+#element_text(angle = -90, hjust = 0))+
        theme(legend.title=element_blank(), legend.position = "none") 
    #theme(legend.position="none")
    #detach(pheno2)
    assign(x = paste0(i,"_bar"),value = fetaladultspine)   
}

paste0(c("Newborn", "<12mo", "12-18mo", "18-24mo", "2-5yr", "5-18yr", 
"18-65yr", ">65yr"), "_bar")
multiplot(  Newborn_bar , `<12mo_bar` ,  `12-18mo_bar` ,`18-24mo_bar`,`2-5yr_bar` ,  `5-18yr_bar` , `18-65yr_bar`, `>65yr_bar`,  cols=3)


boxplot(pheno5[pheno_aging$CellType=="WBC",dput(colnames(coefEsts))], main="All ages")
par(mfrow=c(4,4))
for (i in dput(colnames(coefEsts))){
    boxplot(pheno5[pheno_aging$CellType=="WBC",i]~pheno5$Age_group[pheno_aging$CellType=="WBC"], main= i, xlab="Age group", ylab="Estimated cell %")
}

par(mfrow=c(4,4))
for (i in dput(colnames(coefEsts))){
    plot(pheno5[pheno_aging$CellType=="WBC",i]~pheno_aging$Age[pheno_aging$CellType=="WBC"], main= i, xlab="Age yrs", ylab="Estimated cell %",col="grey")
    fit2<-smooth.spline(pheno5[pheno_aging$CellType=="WBC",i]~pheno_aging$Age[pheno_aging$CellType=="WBC"],cv = TRUE)
#Plotting Regression Line
lines(fit2,lwd=2,col="purple")
}


#validation
tmp2<-Ext_deconv_betas3[, Ext_deconv_pheno$CellType=="WBC"]
#tmp2<-tmp2[,1:6]


cellcountsval<-FlowSorted.Blood.EPIC:::projectCellType(tmp2[rownames(coefEsts),], as.matrix(coefEsts), lessThanOne = TRUE)
sumage<-rowSums(cellcountsval)
cellcountsval2<-round(cellcountsval*100/sumage,1)
head(cellcountsval2)
pheno6<-cbind.data.frame(cellcountsval2, CellType=rep("WBC",6))


stack<-pheno6[pheno6$CellType=="WBC",dput(colnames(coefEsts))]
    stack<-t(stack)
    dat2 <- melt(stack)#, id.vars = "Sample")
    #dat2$X1 <- levels(dat2$X1, ref=i)
    #levels(dat2$X1)
    #dat2$X1 <-fct_rev(dat2$X1)
    #levels(dat2$X1)
    colors1<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
Bcell = "gold", Bnv = "#1B9E77", Bmem = "magenta", CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", CD8T = "darkgray", 
CD8Tnv = "#7570B3", CD8mem = "maroon", 
NKT = "#7E6E85", NK = "#E6AB02")
    colors2<-colors1[levels(dat2$X1)]
    fetaladultspine<- ggplot(dat2, aes(x=X2, y=value, fill=X1)) + 
        geom_bar(stat="identity") +
        xlab("") +#xlab("\nSample") +
        ylab("") +#ylab("Cell %\n") + 
        geom_hline(yintercept=70, linetype="dashed", color = "red")+ 
        #guides(fill=T) +
        theme_bw()+
        ggtitle(paste("Extended projection ", "WBC")) + scale_fill_manual(values=colors2)+
        theme(axis.text.x=element_blank())+#element_text(angle = -90, hjust = 0))+
        theme(legend.title=element_blank(), legend.position = "right") 
    #theme(legend.position="none")
    #detach(pheno2)
    assign(x = paste0("WBC_extended","_bar"),value = fetaladultspine)

#multiplot(  WBC_bar, WBC_FACS_bar, WBC_extended_bar, cols=3)


```

```{r Figure 5, fig.height=12, fig.width=12}
#Figure 5

cellcpgs<-c("cg19486047",#SPTBN1 Bas"
            "cg01884715", #GRB2 Bmem #"cg17291906" #SWAP70
            "cg06243638", #P4HA2 Bnv
            "cg16008150", #LIMS1 CD4mem
            "cg23788058", #CHD7 CD4nv
            "cg07616471",#CCR5 CD8mem
            "cg05496363",#CD248 CD8nv
            "cg17374802", #EPX Eos
            "cg00771778", #LYST Mono
            "cg05501357", #HIPK3 Neu
            "cg08326410", #KIR2DL4 NK
            "cg22572158")#CTLA4 Treg

betascpgs<-getBeta(FlowSorted.BloodExtended.EPIC)

betascpgs<-betascpgs[rownames(betascpgs)%in%cellcpgs,]
phenoref<-as.data.frame(pData(FlowSorted.BloodExtended.EPIC))
identical(rownames(phenoref), colnames(betascpgs))


par(mfrow=c(1,1))
boxplot(t(betascpgs)~phenoref$CellType)

groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
Bnv = "#1B9E77", Bmem = "magenta",  
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", 
CD8nv = "#7570B3", CD8mem = "maroon", NK = "#E6AB02")
cellTypes2 = names(groups)

titlecpgs<-c("cg19486047"="SPTBN1",#TESC Bas
            "cg01884715"="GRB2",# Bmem
            "cg06243638"= "P4HA2",# Bnv
            "cg16008150"= "LIMS1",#CD4mem #"cg05770620"="TRRAP",# CD4mem
            "cg23788058"="CHD7",# CD4nv
            "cg07616471"="CCR5",# CD8mem
            "cg05496363"="CD248",#CD8nv
            "cg17374802"="EPX",# Eos
            "cg00771778"="LYST",# Mono
            "cg05501357"="HIPK3",# Neu
            "cg08326410"="KIR2DL4",# NK
            "cg22572158"="CTLA4")# Treg
colorcelltype<-groups
colorcelltype<-colorcelltype[cellTypes2]
phenoref$CellType<-as.character(phenoref$CellType)
phenoref$CellType<-factor(phenoref$CellType, levels = cellTypes2)
par(mfrow=c(3,4))
for(i in cellcpgs){
    boxplot(betascpgs[i,]~phenoref$CellType, ylim=c(0,1), xlab="", ylab=expression(beta-value), main=paste(titlecpgs[i],i),
            #cex=2, cex.lab=1.3, cex.axis=1.3, cex.main=1.3, cex.sub=1.3,
            boxfill= colorcelltype, las=2, frame.plot = FALSE)
  box(bty="l")
}
```

```{r Figure 4a, fig.height=11, fig.width=11}
load("I:/Dropbox (Christensen Lab)/Longitudinal data/Longitudinal data.RData")
longData<-updateObject(longData)
longpheno<-as.data.frame(pData(longData))
longsesame<-preprocessNoob(longData)#sesamize(longData)
library(readr)
FACSlong <- read_csv("I:/Dropbox (Christensen Lab)/Longitudinal data/FACSlong.csv")
FACSlong<-as.data.frame(FACSlong)
rownames(FACSlong)<-FACSlong$`Sample name`


tmp3<-getBeta(Ext_deconv_rgset)[,Ext_deconv_pheno$CellType=="WBC"]

stack<-as.data.frame(FACSlong[,c("Lymph",   "Mono",  "Gran",   "CD4T",   "CD8T", "nonTcell")])

cellcountsval<-projectCellType_CP(getBeta(longsesame)[IDOLOptimizedCpGsBloodExtended,], FlowSorted.BloodExtended.EPIC.compTable, lessThanOne = TRUE)
boxplot(cellcountsval)
cellcountsval<-round(cellcountsval*100,2)
cellcountsval<-as.data.frame(cellcountsval)
cellcountsval$CD8T<-cellcountsval$CD8mem+cellcountsval$CD8nv
cellcountsval$CD4T<-cellcountsval$CD4mem+cellcountsval$CD4nv+cellcountsval$Treg
cellcountsval$Bcell<-cellcountsval$Bnv+cellcountsval$Bmem
cellcountsval$nonTcell<-cellcountsval$Bcell+cellcountsval$NK
cellcountsval$Gran<-cellcountsval$Neu+cellcountsval$Eos+cellcountsval$Bas





cellTypes<-c("Mono",  "Gran",   "CD4T",   "CD8T", "nonTcell" )
cellcolors<-c( nonTcell="#1B9E77", #CD4mem="darkblue", Bas= "#E41A1C", 
                   CD4T="#D95F02", CD8T ="#7570B3",#CD8T_mem="maroon", #CD8T_CM="#705C83", 
           #Eos="#3E8E93",
           Gran="#66A61E", Mono="#E7298A")#,NK="#E6AB02", #NKT= "#7E6E85", 
           #Treg="black")
truevalues<-stack[complete.cases(stack),]
truevalues<-round(truevalues*100,1)
cellcount_EPICIDOL<-cellcountsval[rownames(truevalues),]

K = length(cellTypes)  # number of cell types

matrixr<-data.frame(Celltypes=cellTypes, R2=rep(NA, length(cellTypes)), RMSE= rep(NA, length(cellTypes)), row.names=cellTypes)

par(mfrow=c(3,3))
for(k in 1:K) {
	pred = as.numeric(cellcount_EPICIDOL[,cellTypes[k]])
	obs = as.numeric(truevalues[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))
	matrixr$R2[k]<-round(R2,2)
	matrixr$RMSE[k]<-round(RMSE,2)

	rangexy = range(c(pred,obs))

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
	      cex.main = 2)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}

cellTypes<-c("Mono",  "Gran",   "CD4T",   "CD8T", "nonTcell" )
cellcolors<-c( Bas= "#E41A1C", Bcell="#1B9E77", CD4mem="darkblue", 
                   CD4nv="#D95F02", CD8T ="#7570B3", CD8mem="maroon", #CD8T_CM="#705C83", 
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02", #NKT= "#7E6E85", 
           Treg="black")

truevalues<-truevalues
cellcount_EPICIDOL<-cellcount_EPICIDOL



library(ggplot2)
library(scatterpie)
library(reshape2)
library(gridExtra)


stack2<-t(round(stack[complete.cases(stack),cellTypes]*100,1))
dat2 <- melt(stack2)#, id.vars = "Sample")
dat2$X1<-as.character(dat2$X1)
dat2<-dat2[order(as.character(dat2$X2)),]
dat2<-dat2[order(as.character(dat2$X1)),]

stack3<-t(cellcount_EPICIDOL[,colnames(cellcount_EPICIDOL)%in%cellTypes])
dat3 <- melt(stack3)
dat3$X1<-as.character(dat3$X1)
dat3<-dat3[order(as.character(dat3$X2)),]
dat3<-dat3[order(as.character(dat3$X1)),]

identical(dat2$X2, dat3$X2)
identical(as.character(dat2$X1), as.character(dat3$X1))

dat2$obs<-dat2$value
dat2$pred<-dat3$value

groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
#Bcell = "gold", 
Bnv = "#1B9E77", Bmem = "magenta", #CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", #CD8T = "darkgray", 
CD8nv = "#7570B3", CD8mem = "maroon", #CD8T_CM = "#705C83", 
#NKT = "#7E6E85", 
NK = "#E6AB02")#, WBC = "white")

for (i in names(groups))
dat2[,i]<-0

for(i in c( "Mono" ))
  dat2[dat2$X1==i,i]<-cellcount_EPICIDOL[,i]
dat2[dat2$X1=="CD4T",c("CD4nv","CD4mem", "Treg")]<-cellcount_EPICIDOL[,c("CD4nv","CD4mem", "Treg")]
dat2[dat2$X1=="CD8T",c("CD8nv","CD8mem")]<-cellcount_EPICIDOL[,c("CD8nv","CD8mem")]
dat2[dat2$X1=="nonTcell",c("Bnv","Bmem", "NK")]<-cellcount_EPICIDOL[,c("Bnv","Bmem", "NK")]
dat2[dat2$X1=="Gran",c("Bas", "Eos","Neu")]<-cellcount_EPICIDOL[,c("Bas", "Eos","Neu")]

dput(colnames(dat2))

plotval<-ggplot() + 
  geom_scatterpie(aes(x=obs, y=pred, r=2), data=dat2, cols=c("Bas", 
                                                               "Bnv", "Bmem", 
                                                               "CD4nv", "CD4mem",
                                                               "CD8nv", "CD8mem",
                                                               "Eos", "Mono", "Neu", 
                                                               "NK", "Treg"))+
  xlab("Observed (%)") +#xlab("\nSample") +
        ylab("Predicted (%)") +#ylab("Cell %\n") + 
  geom_smooth(method = "lm", se = FALSE)+
        geom_abline(intercept=0, slope=1, linetype="dashed", color = "black")+ 
        #guides(fill=T) +
        theme_classic(base_size = 20)+ theme(text = element_text(size=18))+
        ggtitle("Peripheral adult samples with FCM+CBC data (EPIC IDOL-ext)") + scale_fill_manual(values=groups)+ theme(legend.position = "none")+
        #theme(axis.text.x=element_blank())+#element_text(angle = -90, hjust = 0))+
  annotation_custom(tableGrob(matrixr, rows=NULL), 
                    xmin=40, xmax=60, ymin=0, ymax=18)
        

plotval

```

```{r Figure 4b, fig.height=11, fig.width=11}

FlowSorted.Blood.EPIC
EPICmix<-FlowSorted.Blood.EPIC[,FlowSorted.Blood.EPIC$CellType=="MIX"]
EPICpheno<-as.data.frame(pData(EPICmix))
EPICmixsesame<-preprocessNoob(EPICmix)


stack<-as.data.frame(EPICpheno[,c("CD8T", "CD4T", "NK", "Bcell", "Mono", "Neu")])
stack$Gran<-stack$Neu

cellcountsval<-projectCellType_CP(getBeta(EPICmixsesame)[IDOLOptimizedCpGsBloodExtended,], FlowSorted.BloodExtended.EPIC.compTable, lessThanOne = TRUE)
boxplot(cellcountsval)
cellcountsval<-round(cellcountsval*100,2)
cellcountsval<-as.data.frame(cellcountsval)
cellcountsval$CD8T<-cellcountsval$CD8mem+cellcountsval$CD8nv
cellcountsval$CD4T<-cellcountsval$CD4mem+cellcountsval$CD4nv+cellcountsval$Treg
cellcountsval$Bcell<-cellcountsval$Bnv+cellcountsval$Bmem
cellcountsval$nonTcell<-cellcountsval$Bcell+cellcountsval$NK
cellcountsval$Gran<-cellcountsval$Neu+cellcountsval$Eos+cellcountsval$Bas





cellTypes<-c("CD8T", "CD4T", "NK", "Bcell", "Mono", "Gran" )
cellcolors<-c( Bcell="#1B9E77", #CD4mem="darkblue", Bas= "#E41A1C", 
                   CD4T="#D95F02", CD8T ="#7570B3",#CD8T_mem="maroon", #CD8T_CM="#705C83", 
           #Eos="#3E8E93",
           Gran="#66A61E", Mono="#E7298A",NK="#E6AB02")#, #NKT= "#7E6E85", 
           #Treg="black")
truevalues<-stack[complete.cases(stack),]
truevalues<-round(truevalues,1)
cellcount_EPICIDOL<-cellcountsval[rownames(truevalues),]

K = length(cellTypes)  # number of cell types

matrixr<-data.frame(Celltypes=cellTypes, R2=rep(NA, length(cellTypes)), RMSE= rep(NA, length(cellTypes)), row.names=cellTypes)

par(mfrow=c(3,3))
for(k in 1:K) {
	pred = as.numeric(cellcount_EPICIDOL[,cellTypes[k]])
	obs = as.numeric(truevalues[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))
	matrixr$R2[k]<-round(R2,2)
	matrixr$RMSE[k]<-round(RMSE,2)

	rangexy = range(c(pred,obs))

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
	      cex.main = 2)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}

cellTypes<-c("CD8T", "CD4T", "NK", "Bcell", "Mono", "Gran")
cellcolors<-c( Bas= "#E41A1C", Bcell="#1B9E77", CD4mem="darkblue", 
                   CD4nv="#D95F02", CD8T ="#7570B3", CD8mem="maroon", #CD8T_CM="#705C83", 
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02", #NKT= "#7E6E85", 
           Treg="black")

truevalues<-truevalues
cellcount_EPICIDOL<-cellcount_EPICIDOL



library(ggplot2)
library(scatterpie)
library(reshape2)
library(gridExtra)


stack2<-t(round(stack[complete.cases(stack),cellTypes],1))
dat2 <- melt(stack2)#, id.vars = "Sample")
dat2$X1<-as.character(dat2$X1)
dat2<-dat2[order(as.character(dat2$X2)),]
dat2<-dat2[order(as.character(dat2$X1)),]

stack3<-t(cellcount_EPICIDOL[,colnames(cellcount_EPICIDOL)%in%cellTypes])
dat3 <- melt(stack3)
dat3$X1<-as.character(dat3$X1)
dat3<-dat3[order(as.character(dat3$X2)),]
dat3<-dat3[order(as.character(dat3$X1)),]

identical(dat2$X2, dat3$X2)
identical(as.character(dat2$X1), as.character(dat3$X1))

dat2$obs<-dat2$value
dat2$pred<-dat3$value

groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
#Bcell = "gold", 
Bnv = "#1B9E77", Bmem = "magenta", #CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", #CD8T = "darkgray", 
CD8nv = "#7570B3", CD8mem = "maroon", #CD8T_CM = "#705C83", 
#NKT = "#7E6E85", 
NK = "#E6AB02")#, WBC = "white")

for (i in names(groups))
dat2[,i]<-0

for(i in c( "Mono", "NK" ))
  dat2[dat2$X1==i,i]<-cellcount_EPICIDOL[,i]
dat2[dat2$X1=="CD4T",c("CD4nv","CD4mem", "Treg")]<-cellcount_EPICIDOL[,c("CD4nv","CD4mem", "Treg")]
dat2[dat2$X1=="CD8T",c("CD8nv","CD8mem")]<-cellcount_EPICIDOL[,c("CD8nv","CD8mem")]
dat2[dat2$X1=="Bcell",c("Bnv","Bmem")]<-cellcount_EPICIDOL[,c("Bnv","Bmem")]
dat2[dat2$X1=="Gran",c("Bas", "Eos","Neu")]<-cellcount_EPICIDOL[,c("Bas", "Eos","Neu")]

dput(colnames(dat2))

plotval2<-ggplot() + 
  geom_scatterpie(aes(x=obs, y=pred, r=2), data=dat2, cols=c("Bas", 
                                                               "Bnv", "Bmem", 
                                                               "CD4nv", "CD4mem",
                                                               "CD8nv", "CD8mem",
                                                               "Eos", "Mono", "Neu", 
                                                               "NK", "Treg"))+
  xlab("Observed (%)") +
        ylab("Predicted (%)") + 
  geom_smooth(method = "lm", se = FALSE)+
        geom_abline(intercept=0, slope=1, linetype="dashed", color = "black")+ 
        theme_classic(base_size = 20)+ theme(text = element_text(size=18))+
        ggtitle("Adult peripheral blood isolated cells artificial mixtures (EPIC IDOL-ext)") + scale_fill_manual(values=groups)+ theme(legend.position = "none")+
  annotation_custom(tableGrob(matrixr, rows=NULL), 
                    xmin=40, xmax=60, ymin=0, ymax=18)
        

plotval2

```

CB_mixtures

```{r cord blood mix, fig.height=11, fig.width=11}
CB_mixpheno<- read.metharray.sheet(base = "I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data", pattern = "CB_mixinfo.csv")
rownames(CB_mixpheno)<-CB_mixpheno$Array_well
CB_mixpheno$CD8nv<-CB_mixpheno$CD8T_naive


cellTypes = c(#"Basophil", "Bmem", 
  "Bnv", #"CD4mem", 
  "CD4nv", "CD4T",#"CD8T_CM", 
  #            "CD8T_EM", 
"CD8nv", #"Eos", #"MIX", 
"Mono", "Neu", "NK")#, #"NKT", 
#"Treg")#,"WBC")

truevaluesCB<-CB_mixpheno[CB_mixpheno$CellType=="CB_Mix", cellTypes]



```

```{r suppl fig 7 c, fig.height=11, fig.width=11}
# library(ExperimentHub)
# hub <- ExperimentHub()
# query(hub, "FlowSorted.Blood.EPIC")
# FlowSorted.Blood.EPIC <- hub[["EH1136"]]
# FlowSorted.Blood.EPIC
load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/CordbloodEPICsesame.rda")
CordbloodEPICsesame
CordbloodEPICset
EPICmix<-CordbloodEPICset[,CordbloodEPICset$CellType=="CB_Mix"]
EPICpheno<-as.data.frame(CB_mixpheno[CB_mixpheno$CellType=="CB_Mix",])
EPICmixsesame<-EPICmix#sesamize(EPICmix)


stack<-as.data.frame(EPICpheno[,c("CD8T", "CD4T", "NK", "Bcell", "Mono", "Neu")])
stack$Gran<-stack$Neu

cellcountsval<-projectCellType_CP(getBeta(EPICmixsesame)[IDOLOptimizedCpGsBloodExtended,], FlowSorted.BloodExtended.EPIC.compTable, lessThanOne = TRUE)
boxplot(cellcountsval)
cellcountsval<-round(cellcountsval*100,2)
cellcountsval<-as.data.frame(cellcountsval)
cellcountsval$CD8T<-cellcountsval$CD8mem+cellcountsval$CD8nv
cellcountsval$CD4T<-cellcountsval$CD4mem+cellcountsval$CD4nv+cellcountsval$Treg
cellcountsval$Bcell<-cellcountsval$Bnv+cellcountsval$Bmem
cellcountsval$nonTcell<-cellcountsval$Bcell+cellcountsval$NK
cellcountsval$Gran<-cellcountsval$Neu+cellcountsval$Eos+cellcountsval$Bas





cellTypes<-c("CD8T", "CD4T", "NK", "Bcell", "Mono", "Gran" )
cellcolors<-c( Bcell="#1B9E77", #CD4mem="darkblue", Bas= "#E41A1C", 
                   CD4T="#D95F02", CD8T ="#7570B3",#CD8T_mem="maroon", #CD8T_CM="#705C83", 
           #Eos="#3E8E93",
           Gran="#66A61E", Mono="#E7298A",NK="#E6AB02")#, #NKT= "#7E6E85", 
           #Treg="black")
truevalues<-stack[complete.cases(stack),]
truevalues<-round(truevalues,1)
cellcount_EPICIDOL<-cellcountsval[rownames(truevalues),]

K = length(cellTypes)  # number of cell types

matrixr<-data.frame(Celltypes=cellTypes, R2=rep(NA, length(cellTypes)), RMSE= rep(NA, length(cellTypes)), row.names=cellTypes)

par(mfrow=c(3,3))
for(k in 1:K) {
	pred = as.numeric(cellcount_EPICIDOL[,cellTypes[k]])
	obs = as.numeric(truevalues[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))
	matrixr$R2[k]<-round(R2,2)
	matrixr$RMSE[k]<-round(RMSE,2)

	rangexy = range(c(pred,obs))

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
	      cex.main = 2)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}

cellTypes<-c("CD8T", "CD4T", "NK", "Bcell", "Mono", "Gran")
cellcolors<-c( Bas= "#E41A1C", Bcell="#1B9E77", CD4mem="darkblue", 
                   CD4nv="#D95F02", CD8T ="#7570B3", CD8mem="maroon", #CD8T_CM="#705C83", 
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02", #NKT= "#7E6E85", 
           Treg="black")

truevalues<-truevalues
cellcount_EPICIDOL<-cellcount_EPICIDOL



library(ggplot2)
library(scatterpie)
library(reshape2)
library(gridExtra)


stack2<-t(round(stack[complete.cases(stack),cellTypes],1))
dat2 <- melt(stack2)#, id.vars = "Sample")
dat2$X1<-as.character(dat2$X1)
dat2<-dat2[order(as.character(dat2$X2)),]
dat2<-dat2[order(as.character(dat2$X1)),]

stack3<-t(cellcount_EPICIDOL[,colnames(cellcount_EPICIDOL)%in%cellTypes])
dat3 <- melt(stack3)
dat3$X1<-as.character(dat3$X1)
dat3<-dat3[order(as.character(dat3$X2)),]
dat3<-dat3[order(as.character(dat3$X1)),]

identical(dat2$X2, dat3$X2)
identical(as.character(dat2$X1), as.character(dat3$X1))

dat2$obs<-dat2$value
dat2$pred<-dat3$value

groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
#Bcell = "gold", 
Bnv = "#1B9E77", Bmem = "magenta", #CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", #CD8T = "darkgray", 
CD8nv = "#7570B3", CD8mem = "maroon", #CD8T_CM = "#705C83", 
#NKT = "#7E6E85", 
NK = "#E6AB02")#, WBC = "white")

for (i in names(groups))
dat2[,i]<-0

for(i in c( "Mono", "NK" ))
  dat2[dat2$X1==i,i]<-cellcount_EPICIDOL[,i]
dat2[dat2$X1=="CD4T",c("CD4nv","CD4mem", "Treg")]<-cellcount_EPICIDOL[,c("CD4nv","CD4mem", "Treg")]
dat2[dat2$X1=="CD8T",c("CD8nv","CD8mem")]<-cellcount_EPICIDOL[,c("CD8nv","CD8mem")]
dat2[dat2$X1=="Bcell",c("Bnv","Bmem")]<-cellcount_EPICIDOL[,c("Bnv","Bmem")]
dat2[dat2$X1=="Gran",c("Bas", "Eos","Neu")]<-cellcount_EPICIDOL[,c("Bas", "Eos","Neu")]

dput(colnames(dat2))

plotval2a<-ggplot() + 
  geom_scatterpie(aes(x=obs, y=pred, r=2), data=dat2, cols=c("Bas", 
                                                               "Bnv", "Bmem", 
                                                               "CD4nv", "CD4mem",
                                                               "CD8nv", "CD8mem",
                                                               "Eos", "Mono", "Neu", 
                                                               "NK", "Treg"))+
  xlab("Observed (%)") +#xlab("\nSample") +
        ylab("Predicted (%)") +#ylab("Cell %\n") + 
  geom_smooth(method = "lm", se = FALSE)+
        geom_abline(intercept=0, slope=1, linetype="dashed", color = "black")+ 
        #guides(fill=T) +
        theme_classic(base_size = 20)+ theme(text = element_text(size=18))+
        ggtitle("Cell projection in umbilical cord blood artificial mixtures") + scale_fill_manual(values=groups)+ theme(legend.position = "none")+
  annotation_custom(tableGrob(matrixr, rows=NULL), 
                    xmin=40, xmax=60, ymin=0, ymax=18)
        

plotval2a

```

```{r suppl fig 7b, fig.width=12, fig.height=12}
# sheet <- read.metharray.sheet(base = "D:/OneDrive/OneDrive - Dartmouth College/Cord blood letter to the editor/archive", pattern = "WBasmplesheetMay2016.csv")
# #workdir<-"I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data"
# JonesepicData <- read.metharray.exp(targets = sheet,extended = TRUE, force = TRUE)
# Jonessesame<-sesamize(JonesepicData)
# save(JonesepicData, Jonessesame, file= "Raw_data_Jones.RData")

load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/Raw_data_Jones.RData")

stack<-as.data.frame(pData(Jonessesame))[,c("nRBC", "Bcell","CD4T",  "CD8T", "Gran",  "Mono", "NK" )]
stack$myeloid_Bcell<-stack$Gran+stack$nRBC+stack$Mono+stack$Bcell
stack$T_NK_lymphoid<-stack$CD8T+stack$CD4T+stack$NK
stack$myeloid<-stack$Gran+stack$nRBC
#stack<-as.data.frame(FACSlong[,c("Lymph",   "Mono",  "Gran",   "CD4T",   "CD8T", "nonTcell")])
Jonessesame<-preprocessNoob(JonesepicData)
cellcountsval<-projectCellType_CP(getBeta(Jonessesame)[IDOLOptimizedCpGsBloodExtended450k,], FlowSorted.BloodExtended.450klegacy.compTable, lessThanOne = TRUE)
boxplot(cellcountsval)
cellcountsval<-round(cellcountsval*100,2)
cellcountsval<-as.data.frame(cellcountsval)
cellcountsval$CD8T<-cellcountsval$CD8mem+cellcountsval$CD8nv

cellcountsval$CD4T<-cellcountsval$CD4mem+cellcountsval$CD4nv+cellcountsval$Treg
cellcountsval$Bcell<-cellcountsval$Bnv+cellcountsval$Bmem
cellcountsval$nRBC<-cellcountsval$Bas
cellcountsval$Gran<-cellcountsval$Neu+cellcountsval$Eos
cellcountsval$myeloid<-cellcountsval$Gran+cellcountsval$nRBC
cellcountsval$myeloid_Bcell<-cellcountsval$Gran+cellcountsval$nRBC+cellcountsval$Mono+cellcountsval$Bcell
cellcountsval$T_NK_lymphoid<-cellcountsval$CD8T+cellcountsval$CD4T+cellcountsval$NK




cellTypes<-c("nRBC", "Bcell","CD4T",  "CD8T", "Gran",  "Mono", "NK" )
cellcolors<-c( nRBC= "#E41A1C", Bcell="#1B9E77", #CD4mem="darkblue", 
                   CD4T="#D95F02", CD8T ="#7570B3",#CD8T_mem="maroon", #CD8T_CM="#705C83", 
           #Eos="#3E8E93",
           Gran="#66A61E", Mono="#E7298A",NK="#E6AB02")#, #NKT= "#7E6E85", 
           #Treg="black")
truevalues<-stack#[complete.cases(stack),]
#truevalues<-round(truevalues*100,1)
cellcount_EPICIDOL<-cellcountsval#[rownames(truevalues),]

K = length(cellTypes)  # number of cell types

matrixr<-data.frame(Celltypes=cellTypes, R2=rep(NA, length(cellTypes)), RMSE= rep(NA, length(cellTypes)), row.names=cellTypes)

par(mfrow=c(3,3))
for(k in 1:K) {
	pred = as.numeric(cellcount_EPICIDOL[,cellTypes[k]])
	obs = as.numeric(truevalues[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))
	matrixr$R2[k]<-round(R2,2)
	matrixr$RMSE[k]<-round(RMSE,2)

	rangexy = range(c(pred,obs))

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
	      cex.main = 2)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}

cellTypes<-c("nRBC", "Bcell","CD4T",  "CD8T", "Gran",  "Mono", "NK" )
cellcolors<-c( nRBC= "#E41A1C", Bcell="#1B9E77", CD4mem="darkblue", 
                   CD4nv="#D95F02", CD8T ="#7570B3", CD8mem="maroon", #CD8T_CM="#705C83", 
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02", #NKT= "#7E6E85", 
           Treg="black")

truevalues<-truevalues
cellcount_EPICIDOL<-cellcount_EPICIDOL



library(ggplot2)
library(scatterpie)
library(reshape2)
library(gridExtra)


stack2<-t(stack[,cellTypes])
dat2 <- melt(stack2)#, id.vars = "Sample")
dat2$X1<-as.character(dat2$X1)
dat2<-dat2[order(as.character(dat2$X2)),]
dat2<-dat2[order(as.character(dat2$X1)),]

stack3<-t(cellcount_EPICIDOL[,colnames(cellcount_EPICIDOL)%in%cellTypes])
dat3 <- melt(stack3)
dat3$X1<-as.character(dat3$X1)
dat3<-dat3[order(as.character(dat3$X2)),]
dat3<-dat3[order(as.character(dat3$X1)),]

identical(dat2$X2, dat3$X2)
identical(as.character(dat2$X1), as.character(dat3$X1))

dat2$obs<-dat2$value
dat2$pred<-dat3$value

groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
#Bcell = "gold", 
Bnv = "#1B9E77", Bmem = "magenta", #CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", #CD8T = "darkgray", 
CD8nv = "#7570B3", CD8mem = "maroon", #CD8T_CM = "#705C83", 
#NKT = "#7E6E85", 
NK = "#E6AB02")#, WBC = "white")

for (i in names(groups))
dat2[,i]<-0

for(i in c("nRBC", "Mono", "NK" ))
  dat2[dat2$X1==i,i]<-cellcount_EPICIDOL[,i]
dat2[dat2$X1=="CD4T",c("CD4nv","CD4mem", "Treg")]<-cellcount_EPICIDOL[,c("CD4nv","CD4mem", "Treg")]
dat2[dat2$X1=="CD8T",c("CD8nv","CD8mem")]<-cellcount_EPICIDOL[,c("CD8nv","CD8mem")]
dat2[dat2$X1=="Bcell",c("Bnv","Bmem")]<-cellcount_EPICIDOL[,c("Bnv","Bmem")]
dat2[dat2$X1=="Gran",c( "Eos","Neu")]<-cellcount_EPICIDOL[,c( "Eos","Neu")]

dput(colnames(dat2))

plotval3<-ggplot() + 
  geom_scatterpie(aes(x=obs, y=pred, r=1), data=dat2, cols=c("nRBC", 
                                                               "Bnv", "Bmem", 
                                                               "CD4nv", "CD4mem",
                                                               "CD8nv", "CD8mem",
                                                               "Eos", "Mono", "Neu", 
                                                               "NK", "Treg"))+
  xlab("Observed (%)") +#xlab("\nSample") +
        ylab("Predicted (%)") +#ylab("Cell %\n") + 
  geom_smooth(method = "lm", se = FALSE)+
        geom_abline(intercept=0, slope=1, linetype="dashed", color = "black")+ 
        #guides(fill=T) +
        theme_classic(base_size = 16)+ #theme(text = element_text(size=rel(2)))+
        ggtitle("Cell projection extended versus FACS data") + scale_fill_manual(values=groups)+
        #theme(axis.text.x=element_blank())+#element_text(angle = -90, hjust = 0))+
  annotation_custom(tableGrob(matrixr, rows=NULL), 
                    xmin=40, xmax=60, ymin=0, ymax=18)
        

plotval3

#Collapsing all the myeloid

cellTypes<-c( "Bcell","CD4T",  "CD8T", "myeloid",  "Mono", "NK" )
cellcolors<-c( #nRBC= "#E41A1C", 
  Bcell="#1B9E77", #CD4mem="darkblue", 
                   CD4T="#D95F02", CD8T ="#7570B3",#CD8T_mem="maroon", #CD8T_CM="#705C83", 
           #Eos="#3E8E93",
           myeloid="#66A61E", Mono="#E7298A",NK="#E6AB02")#, #NKT= "#7E6E85", 
           #Treg="black")
truevalues<-stack#[complete.cases(stack),]
#truevalues<-round(truevalues*100,1)
cellcount_EPICIDOL<-cellcountsval#[rownames(truevalues),]

K = length(cellTypes)  # number of cell types

matrixr<-data.frame(Celltypes=cellTypes, R2=rep(NA, length(cellTypes)), RMSE= rep(NA, length(cellTypes)), row.names=cellTypes)

par(mfrow=c(3,3))
for(k in 1:K) {
	pred = as.numeric(cellcount_EPICIDOL[,cellTypes[k]])
	obs = as.numeric(truevalues[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))
	matrixr$R2[k]<-round(R2,2)
	matrixr$RMSE[k]<-round(RMSE,2)

	rangexy = range(c(pred,obs))

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
	      cex.main = 2)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}

cellTypes<-c("myeloid", "Bcell","CD4T",  "CD8T", "Gran",  "Mono", "NK" )
cellcolors<-c( Bas= "#E41A1C", Bcell="#1B9E77", CD4mem="darkblue", 
                   CD4nv="#D95F02", CD8T ="#7570B3", CD8mem="maroon", #CD8T_CM="#705C83", 
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02", #NKT= "#7E6E85", 
           Treg="black")

truevalues<-truevalues
cellcount_EPICIDOL<-cellcount_EPICIDOL



library(ggplot2)
library(scatterpie)
library(reshape2)
library(gridExtra)


stack2<-t(stack[,cellTypes])
dat2 <- melt(stack2)#, id.vars = "Sample")
dat2$X1<-as.character(dat2$X1)
dat2<-dat2[order(as.character(dat2$X2)),]
dat2<-dat2[order(as.character(dat2$X1)),]

stack3<-t(cellcount_EPICIDOL[,colnames(cellcount_EPICIDOL)%in%cellTypes])
dat3 <- melt(stack3)
dat3$X1<-as.character(dat3$X1)
dat3<-dat3[order(as.character(dat3$X2)),]
dat3<-dat3[order(as.character(dat3$X1)),]

identical(dat2$X2, dat3$X2)
identical(as.character(dat2$X1), as.character(dat3$X1))

dat2$obs<-dat2$value
dat2$pred<-dat3$value

groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
#Bcell = "gold", 
Bnv = "#1B9E77", Bmem = "magenta", #CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", #CD8T = "darkgray", 
CD8nv = "#7570B3", CD8mem = "maroon", #CD8T_CM = "#705C83", 
#NKT = "#7E6E85", 
NK = "#E6AB02")#, WBC = "white")

for (i in names(groups))
dat2[,i]<-0

for(i in c("Mono", "NK" ))
  dat2[dat2$X1==i,i]<-cellcount_EPICIDOL[,i]
dat2[dat2$X1=="CD4T",c("CD4nv","CD4mem", "Treg")]<-cellcount_EPICIDOL[,c("CD4nv","CD4mem", "Treg")]
dat2[dat2$X1=="CD8T",c("CD8nv","CD8mem")]<-cellcount_EPICIDOL[,c("CD8nv","CD8mem")]
dat2[dat2$X1=="Bcell",c("Bnv","Bmem")]<-cellcount_EPICIDOL[,c("Bnv","Bmem")]
dat2[dat2$X1=="myeloid",c( "Bas","Eos","Neu")]<-cellcount_EPICIDOL[,c( "Bas","Eos","Neu")]

dput(colnames(dat2))

plotval3a<-ggplot() + 
  geom_scatterpie(aes(x=obs, y=pred, r=2), data=dat2, cols=c("Bas", 
                                                               "Bnv", "Bmem", 
                                                               "CD4nv", "CD4mem",
                                                               "CD8nv", "CD8mem",
                                                               "Eos", "Mono", "Neu", 
                                                               "NK", "Treg"))+
  xlab("Observed (%)") +#xlab("\nSample") +
        ylab("Predicted (%)") +#ylab("Cell %\n") + 
  geom_smooth(method = "lm", se = FALSE)+
        geom_abline(intercept=0, slope=1, linetype="dashed", color = "black")+ 
        #guides(fill=T) +
        theme_classic(base_size = 20)+ theme(text = element_text(size=18))+
        ggtitle("Cell projection extended versus FACS data") + scale_fill_manual(values=groups)+
        theme(legend.position = "none")+#element_text(angle = -90, hjust = 0))+
  annotation_custom(tableGrob(matrixr, rows=NULL), 
                    xmin=40, xmax=60, ymin=0, ymax=18)
        

plotval3a



#Myeloid/lymphoid only


cellTypes<-c("myeloid_Bcell", "T_NK_lymphoid" )
cellcolors<-c( myeloid_Bcell= "black", T_NK_lymphoid="lightgray")
# cellcolors<-c( nRBC= "#E41A1C", Bcell="#1B9E77", #CD4mem="darkblue", 
#                    CD4T="#D95F02", CD8T ="#7570B3",#CD8T_mem="maroon", #CD8T_CM="#705C83", 
#            #Eos="#3E8E93",
#            Gran="#66A61E", Mono="#E7298A",NK="#E6AB02")#, #NKT= "#7E6E85", 
#            #Treg="black")
truevalues<-stack#[complete.cases(stack),]
#truevalues<-round(truevalues*100,1)
cellcount_EPICIDOL<-cellcountsval#[rownames(truevalues),]

K = length(cellTypes)  # number of cell types

matrixr<-data.frame(Celltypes=cellTypes, R2=rep(NA, length(cellTypes)), RMSE= rep(NA, length(cellTypes)), row.names=cellTypes)

par(mfrow=c(3,3))
for(k in 1:K) {
	pred = as.numeric(cellcount_EPICIDOL[,cellTypes[k]])
	obs = as.numeric(truevalues[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))
	matrixr$R2[k]<-round(R2,2)
	matrixr$RMSE[k]<-round(RMSE,2)

	rangexy = range(c(pred,obs))

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
	      cex.main = 2)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}

cellTypes<-c("myeloid_Bcell", "T_NK_lymphoid" )
cellcolors<-c( nRBC= "#E41A1C", Bcell="#1B9E77", CD4mem="darkblue", 
                   CD4nv="#D95F02", CD8T ="#7570B3", CD8mem="maroon", #CD8T_CM="#705C83", 
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02", #NKT= "#7E6E85", 
           Treg="black")

truevalues<-truevalues
cellcount_EPICIDOL<-cellcount_EPICIDOL



library(ggplot2)
library(scatterpie)
library(reshape2)
library(gridExtra)


stack2<-t(stack[,cellTypes])
dat2 <- melt(stack2)#, id.vars = "Sample")
dat2$X1<-as.character(dat2$X1)
dat2<-dat2[order(as.character(dat2$X2)),]
dat2<-dat2[order(as.character(dat2$X1)),]

stack3<-t(cellcount_EPICIDOL[,colnames(cellcount_EPICIDOL)%in%cellTypes])
dat3 <- melt(stack3)
dat3$X1<-as.character(dat3$X1)
dat3<-dat3[order(as.character(dat3$X2)),]
dat3<-dat3[order(as.character(dat3$X1)),]

identical(dat2$X2, dat3$X2)
identical(as.character(dat2$X1), as.character(dat3$X1))

dat2$obs<-dat2$value
dat2$pred<-dat3$value

groups<-c(nRBC = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
#Bcell = "gold", 
Bnv = "#1B9E77", Bmem = "magenta", #CD4T = "lightgray", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", #CD8T = "darkgray", 
CD8nv = "#7570B3", CD8mem = "maroon", #CD8T_CM = "#705C83", 
#NKT = "#7E6E85", 
NK = "#E6AB02")#, WBC = "white")

for (i in names(groups))
dat2[,i]<-0


dat2[dat2$X1=="T_NK_lymphoid",c("CD4nv","CD4mem", "Treg", "CD8nv","CD8mem",  "NK")]<-cellcount_EPICIDOL[,c("CD4nv","CD4mem", "Treg", "CD8nv","CD8mem",  "NK")]
dat2[dat2$X1=="myeloid_Bcell",c( "Eos","Neu", "Mono", "Bnv","Bmem","nRBC")]<-cellcount_EPICIDOL[,c( "Eos","Neu", "Mono", "Bnv","Bmem","nRBC")]

dput(colnames(dat2))

plotval4<-ggplot() + 
  geom_scatterpie(aes(x=obs, y=pred, r=1), data=dat2, cols=c("nRBC", 
                                                               "Bnv", "Bmem", 
                                                               "CD4nv", "CD4mem",
                                                               "CD8nv", "CD8mem",
                                                               "Eos", "Mono", "Neu", 
                                                               "NK", "Treg"))+
  xlab("Observed (%)") +#xlab("\nSample") +
        ylab("Predicted (%)") +#ylab("Cell %\n") + 
  geom_smooth(method = "lm", se = FALSE)+
        geom_abline(intercept=0, slope=1, linetype="dashed", color = "black")+ 
        #guides(fill=T) +
        theme_classic(base_size = 16)+ #theme(text = element_text(size=rel(2)))+
        ggtitle("Cell projection extended versus FACS data") + scale_fill_manual(values=groups)+
        #theme(axis.text.x=element_blank())+#element_text(angle = -90, hjust = 0))+
  annotation_custom(tableGrob(matrixr, rows=NULL), 
                    xmin=40, xmax=60, ymin=0, ymax=18)
        

plotval4



# countsextnRBCref2<-projectCellType_CP(getBeta(JonesMsetEx)[IDOLopt_CB_adult_2000$`IDOL Optimized Library`,], IDOLopt_CB_adult_2000$`IDOL Optimized CoefEsts`, lessThanOne = TRUE)
# countsextnRBCref<-round(countsextnRBCref*100,1)
# boxplot(countsextnRBCref)

```

```{r suppl fig 6a, fig.height=11, fig.width=11}

load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/UCSF_sesame.rda")
load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/GBM UCSF/Unfiltered.rda")
#UCSF_sesame
EPICmix<-preprocessNoob(epicData)#UCSF_sesame
colnames(EPICmix)<-colnames(UCSF_betas)
UCSF_pheno$CD8nv<-UCSF_pheno$CD8T_naive
UCSF_pheno$CD8mem<-UCSF_pheno$CD8Tmem
UCSF_pheno$nonTcell<-100-(UCSF_pheno$CD4mem+UCSF_pheno$CD4nv+UCSF_pheno$CD8nv+UCSF_pheno$CD8mem)
EPICpheno<-as.data.frame(UCSF_pheno)
EPICmixsesame<-EPICmix#sesamize(EPICmix)
colnames(EPICmixsesame)<-rownames((UCSF_pheno))

stack<-as.data.frame(EPICpheno[,c("CD8nv", "CD8mem", "CD4mem", "CD4nv", "nonTcell")])


cellcountsval<-projectCellType_CP(getBeta(EPICmixsesame)[IDOLOptimizedCpGsBloodExtended,], FlowSorted.BloodExtended.EPIC.compTable, lessThanOne = TRUE)
boxplot(cellcountsval)
cellcountsval<-round(cellcountsval*100,2)
cellcountsval<-as.data.frame(cellcountsval)
cellcountsval$CD8T<-cellcountsval$CD8mem+cellcountsval$CD8nv
cellcountsval$CD4mem2<-cellcountsval$CD4mem
cellcountsval$CD4mem<-cellcountsval$CD4mem+cellcountsval$Treg
cellcountsval$CD4T<-cellcountsval$CD4mem+cellcountsval$CD4nv
cellcountsval$Bcell<-cellcountsval$Bnv+cellcountsval$Bmem
cellcountsval$nonTcell<-cellcountsval$Bcell+cellcountsval$NK+cellcountsval$Neu+cellcountsval$Eos+cellcountsval$Bas

cellTypes<-c("CD8nv", "CD4nv", "CD8mem", "CD4mem", "nonTcell" )
cellcolors<-c( CD4mem="darkblue", 
                   CD4nv="#D95F02", CD8nv ="#7570B3",CD8mem="maroon", nonTcell="#66A61E")
truevalues<-stack[complete.cases(stack),]
truevalues<-round(truevalues,1)
cellcount_EPICIDOL<-cellcountsval[rownames(truevalues),]

K = length(cellTypes)  # number of cell types

matrixr<-data.frame(Celltypes=cellTypes, R2=rep(NA, length(cellTypes)), RMSE= rep(NA, length(cellTypes)), row.names=cellTypes)

par(mfrow=c(3,3))
for(k in 1:K) {
	pred = as.numeric(cellcount_EPICIDOL[,cellTypes[k]])
	obs = as.numeric(truevalues[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))
	matrixr$R2[k]<-round(R2,2)
	matrixr$RMSE[k]<-round(RMSE,2)

	rangexy = range(c(pred,obs))

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
	      cex.main = 2)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}

cellTypes<-c("CD8nv", "CD4nv", "CD8mem", "CD4mem", "nonTcell" )
cellcolors<-c( Bas= "#E41A1C", Bcell="#1B9E77", CD4mem="darkblue", 
                   CD4nv="#D95F02", CD8nv ="#7570B3", CD8mem="maroon", 
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02",  
           Treg="black")

truevalues<-truevalues
cellcount_EPICIDOL<-cellcount_EPICIDOL



library(ggplot2)
library(scatterpie)
library(reshape2)
library(gridExtra)


stack2<-t(round(stack[complete.cases(stack),cellTypes],1))
dat2 <- melt(stack2)#, id.vars = "Sample")
dat2$X1<-as.character(dat2$X1)
dat2<-dat2[order(as.character(dat2$X2)),]
dat2<-dat2[order(as.character(dat2$X1)),]

stack3<-t(cellcount_EPICIDOL[,colnames(cellcount_EPICIDOL)%in%cellTypes])
dat3 <- melt(stack3)
dat3$X1<-as.character(dat3$X1)
dat3<-dat3[order(as.character(dat3$X2)),]
dat3<-dat3[order(as.character(dat3$X1)),]

identical(dat2$X2, dat3$X2)
identical(as.character(dat2$X1), as.character(dat3$X1))

dat2$obs<-dat2$value
dat2$pred<-dat3$value

groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
Bnv = "#1B9E77", Bmem = "magenta", 
CD4nv = "#D95F02", CD4mem2 = "darkblue", Treg = "black", 
CD8nv = "#7570B3", CD8mem = "maroon", NK = "#E6AB02")

for (i in names(groups))
dat2[,i]<-0

for(i in c( "CD4nv", "CD8nv", "CD8mem" ))
  dat2[dat2$X1==i,i]<-cellcount_EPICIDOL[,i]
dat2[dat2$X1=="CD4mem",c("CD4mem2", "Treg")]<-cellcount_EPICIDOL[,c("CD4mem2", "Treg")]
dat2[dat2$X1=="nonTcell",c("Bnv","Bmem", "Bas", "Eos","Neu")]<-cellcount_EPICIDOL[,c("Bnv","Bmem", "Bas", "Eos","Neu")]

dput(colnames(dat2))

plotval5<-ggplot() + 
  geom_scatterpie(aes(x=obs, y=pred, r=2), data=dat2, cols=c("Bas", 
                                                               "Bnv", "Bmem", 
                                                               "CD4nv", "CD4mem2",
                                                               "CD8nv", "CD8mem",
                                                               "Eos", "Mono", "Neu", 
                                                               "NK", "Treg"))+
  xlab("Observed (%)") +#xlab("\nSample") +
        ylab("Predicted (%)") +#ylab("Cell %\n") + 
  geom_smooth(method = "lm", se = FALSE)+
        geom_abline(intercept=0, slope=1, linetype="dashed", color = "black")+ 
        theme_classic(base_size = 20)+ theme(text = element_text(size=18))+
        ggtitle("Adult peripheral samples from glioma patients with FCM data \n(EPIC IDOL-ext)") + scale_fill_manual(values=groups)+
        theme(legend.position = "none")+
  annotation_custom(tableGrob(matrixr, rows=NULL), 
                    xmin=40, xmax=60, ymin=0, ymax=18)
        

plotval5


```

```{r suppl fig 6b, fig.height=11, fig.width=11}

library(FlowSorted.Blood.450k)
FlowSorted.Blood.450k
EPICmix<-FlowSorted.Blood.450k[,FlowSorted.Blood.450k$CellType%in%c("WBC", "PBMC", "Gran")]
table(EPICmix$CellType)

EPICpheno<-read.csv("Reinius.csv", row.names = 2)
EPICpheno<-EPICpheno[colnames(EPICmix),]


identical(colnames(EPICmix), rownames(EPICpheno))
EPICmixsesame<-preprocessNoob(EPICmix)#sesamize(EPICmix)
EPICmixsesame<-EPICmixsesame[,order(colnames(EPICmixsesame))]
EPICpheno<-EPICpheno[order(rownames(EPICpheno)),]
identical(colnames(EPICmixsesame), rownames(EPICpheno))



stack<-as.data.frame(EPICpheno[,c("CD4T", "CD8T", "Mono", "Bcell", "NK", "Neu", "Eos")])
stack$Gran<-stack$Neu+stack$Eos

cellcountsval<-projectCellType_CP(getBeta(EPICmixsesame)[IDOLOptimizedCpGsBloodExtended450k,], FlowSorted.BloodExtended.450klegacy.compTable, lessThanOne = TRUE)
boxplot(cellcountsval)
cellcountsval<-round(cellcountsval*100,2)
cellcountsval<-as.data.frame(cellcountsval)
cellcountsval$CD8T<-cellcountsval$CD8mem+cellcountsval$CD8nv
cellcountsval$CD4T<-cellcountsval$CD4mem+cellcountsval$CD4nv+cellcountsval$Treg
cellcountsval$Bcell<-cellcountsval$Bnv+cellcountsval$Bmem
cellcountsval$Gran<-cellcountsval$Neu+cellcountsval$Eos+cellcountsval$Bas

cellTypes<-c("CD4T", "CD8T", "Mono", "Bcell", "NK", "Gran" )
cellcolors<-c( Bcell="#1B9E77", 
                   CD4T="#D95F02", CD8T ="#7570B3",
               Gran="#66A61E", Mono="#E7298A",NK="#E6AB02")
truevalues<-stack[complete.cases(stack),]
truevalues<-round(truevalues,1)
cellcount_EPICIDOL<-cellcountsval[rownames(truevalues),]

K = length(cellTypes)  # number of cell types

matrixr<-data.frame(Celltypes=cellTypes, R2=rep(NA, length(cellTypes)), RMSE= rep(NA, length(cellTypes)), row.names=cellTypes)

par(mfrow=c(3,3))
for(k in 1:K) {
	pred = as.numeric(cellcount_EPICIDOL[,cellTypes[k]])
	obs = as.numeric(truevalues[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))
	matrixr$R2[k]<-round(R2,2)
	matrixr$RMSE[k]<-round(RMSE,2)

	rangexy = range(c(pred,obs))

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
	      cex.main = 2)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}

cellTypes<-c("CD4T", "CD8T", "Mono", "Bcell", "NK", "Gran" )
cellcolors<-c( Bas= "#E41A1C", Bcell="#1B9E77", CD4mem="darkblue", 
                   CD4nv="#D95F02", CD8nv ="#7570B3", CD8mem="maroon", 
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02",  
           Treg="black")

truevalues<-truevalues
cellcount_EPICIDOL<-cellcount_EPICIDOL
identical(rownames(truevalues), rownames(cellcount_EPICIDOL))



library(ggplot2)
library(scatterpie)
library(reshape2)
library(gridExtra)


stack2<-t(round(stack[complete.cases(stack),cellTypes],1))
dat2 <- melt(stack2)#, id.vars = "Sample")
dat2$X1<-as.character(dat2$X1)
dat2<-dat2[order(as.character(dat2$X2)),]
dat2<-dat2[order(as.character(dat2$X1)),]

stack3<-t(cellcount_EPICIDOL[,colnames(cellcount_EPICIDOL)%in%cellTypes])
dat3 <- melt(stack3)
dat3$X1<-as.character(dat3$X1)
dat3<-dat3[order(as.character(dat3$X2)),]
dat3<-dat3[order(as.character(dat3$X1)),]

identical(dat2$X2, dat3$X2)
identical(as.character(dat2$X1), as.character(dat3$X1))

dat2$obs<-dat2$value
dat2$pred<-dat3$value

groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
Bnv = "#1B9E77", Bmem = "magenta", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", 
CD8nv = "#7570B3", CD8mem = "maroon", NK = "#E6AB02")

for (i in names(groups))
dat2[,i]<-0

for(i in c( "Mono",  "NK" ))
  dat2[dat2$X1==i,i]<-cellcount_EPICIDOL[,i]
dat2[dat2$X1=="CD4T",c("CD4nv","CD4mem", "Treg")]<-cellcount_EPICIDOL[,c("CD4nv","CD4mem", "Treg")]
dat2[dat2$X1=="CD8T",c("CD8nv","CD8mem")]<-cellcount_EPICIDOL[,c("CD8nv","CD8mem")]
dat2[dat2$X1=="Bcell",c("Bnv","Bmem")]<-cellcount_EPICIDOL[,c("Bnv","Bmem")]
dat2[dat2$X1=="Gran",c( "Bas","Eos","Neu")]<-cellcount_EPICIDOL[,c( "Bas","Eos","Neu")]

dput(colnames(dat2))

plotval6<-ggplot() + 
  geom_scatterpie(aes(x=obs, y=pred, r=2), data=dat2, cols=c("Bas", 
                                                               "Bnv", "Bmem", 
                                                               "CD4nv", "CD4mem",
                                                               "CD8nv", "CD8mem",
                                                               "Eos", "Mono", "Neu", 
                                                               "NK", "Treg"))+
  xlab("Observed (%)") +#xlab("\nSample") +
        ylab("Predicted (%)") +#ylab("Cell %\n") + 
  geom_smooth(method = "lm", se = FALSE)+
        geom_abline(intercept=0, slope=1, linetype="dashed", color = "black")+ 
        #guides(fill=T) +
        theme_classic(base_size = 20)+ #theme(text = element_text(size=rel(2)))+
        ggtitle("Reinius reported FACS") + scale_fill_manual(values=groups)+
        theme(legend.position = "none")+#element_text(angle = -90, hjust = 0))+
  annotation_custom(tableGrob(matrixr, rows=NULL), 
                    xmin=40, xmax=60, ymin=0, ymax=18)
        

plotval6

#Only WBC
stack<-as.data.frame(EPICpheno[EPICpheno$CellType=="WBC",c("CD4T", "CD8T", "Mono", "Bcell", "NK", "Neu", "Eos")])
stack$Gran<-stack$Neu+stack$Eos

cellcountsval<-projectCellType_CP(getBeta(EPICmixsesame)[IDOLOptimizedCpGsBloodExtended450k,EPICpheno$CellType=="WBC"], FlowSorted.BloodExtended.450klegacy.compTable, lessThanOne = TRUE)
boxplot(cellcountsval)
cellcountsval<-round(cellcountsval*100,2)
cellcountsval<-as.data.frame(cellcountsval)
cellcountsval$CD8T<-cellcountsval$CD8mem+cellcountsval$CD8nv
cellcountsval$CD4T<-cellcountsval$CD4mem+cellcountsval$CD4nv+cellcountsval$Treg
cellcountsval$Bcell<-cellcountsval$Bnv+cellcountsval$Bmem
cellcountsval$Gran<-cellcountsval$Neu+cellcountsval$Eos+cellcountsval$Bas

cellTypes<-c("CD4T", "CD8T", "Mono", "Bcell", "NK", "Gran" )
cellcolors<-c( Bcell="#1B9E77", 
               CD4T="#D95F02", CD8T ="#7570B3",Gran="#66A61E", Mono="#E7298A",NK="#E6AB02")
truevalues<-stack[complete.cases(stack),]
truevalues<-round(truevalues,1)
cellcount_EPICIDOL<-cellcountsval[rownames(truevalues),]

K = length(cellTypes)  # number of cell types

matrixr<-data.frame(Celltypes=cellTypes, R2=rep(NA, length(cellTypes)), RMSE= rep(NA, length(cellTypes)), row.names=cellTypes)

par(mfrow=c(3,3))
for(k in 1:K) {
	pred = as.numeric(cellcount_EPICIDOL[,cellTypes[k]])
	obs = as.numeric(truevalues[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))
	matrixr$R2[k]<-round(R2,2)
	matrixr$RMSE[k]<-round(RMSE,2)

	rangexy = range(c(pred,obs))

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
	      cex.main = 2)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}

cellTypes<-c("CD4T", "CD8T", "Mono", "Bcell", "NK", "Gran" )
cellcolors<-c( Bas= "#E41A1C", Bcell="#1B9E77", CD4mem="darkblue", 
                   CD4nv="#D95F02", CD8nv ="#7570B3", CD8mem="maroon",  
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02", 
           Treg="black")

truevalues<-truevalues
cellcount_EPICIDOL<-cellcount_EPICIDOL
identical(rownames(truevalues), rownames(cellcount_EPICIDOL))



library(ggplot2)
library(scatterpie)
library(reshape2)
library(gridExtra)


stack2<-t(round(stack[complete.cases(stack),cellTypes],1))
dat2 <- melt(stack2)#, id.vars = "Sample")
dat2$X1<-as.character(dat2$X1)
dat2<-dat2[order(as.character(dat2$X2)),]
dat2<-dat2[order(as.character(dat2$X1)),]

stack3<-t(cellcount_EPICIDOL[,colnames(cellcount_EPICIDOL)%in%cellTypes])
dat3 <- melt(stack3)
dat3$X1<-as.character(dat3$X1)
dat3<-dat3[order(as.character(dat3$X2)),]
dat3<-dat3[order(as.character(dat3$X1)),]

identical(dat2$X2, dat3$X2)
identical(as.character(dat2$X1), as.character(dat3$X1))

dat2$obs<-dat2$value
dat2$pred<-dat3$value

groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
Bnv = "#1B9E77", Bmem = "magenta", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black",  
CD8nv = "#7570B3", CD8mem = "maroon", NK = "#E6AB02")

for (i in names(groups))
dat2[,i]<-0

for(i in c( "Mono",  "NK" ))
  dat2[dat2$X1==i,i]<-cellcount_EPICIDOL[,i]
dat2[dat2$X1=="CD4T",c("CD4nv","CD4mem", "Treg")]<-cellcount_EPICIDOL[,c("CD4nv","CD4mem", "Treg")]
dat2[dat2$X1=="CD8T",c("CD8nv","CD8mem")]<-cellcount_EPICIDOL[,c("CD8nv","CD8mem")]
dat2[dat2$X1=="Bcell",c("Bnv","Bmem")]<-cellcount_EPICIDOL[,c("Bnv","Bmem")]
dat2[dat2$X1=="Gran",c( "Bas","Eos","Neu")]<-cellcount_EPICIDOL[,c( "Bas","Eos","Neu")]

dput(colnames(dat2))

plotval6a<-ggplot() + 
  geom_scatterpie(aes(x=obs, y=pred, r=2), data=dat2, cols=c("Bas", 
                                                               "Bnv", "Bmem", 
                                                               "CD4nv", "CD4mem",
                                                               "CD8nv", "CD8mem",
                                                               "Eos", "Mono", "Neu", 
                                                               "NK", "Treg"))+
  xlab("Observed (%)") +#xlab("\nSample") +
        ylab("Predicted (%)") +#ylab("Cell %\n") + 
  geom_smooth(method = "lm", se = FALSE)+
        geom_abline(intercept=0, slope=1, linetype="dashed", color = "black")+ 
         theme_classic(base_size = 20)+ #theme(text = element_text(size=rel(2)))+
        ggtitle("Reinius reported FACS") + scale_fill_manual(values=groups)+
        theme(legend.position = "none")+#element_text(angle = -90, hjust = 0))+
  annotation_custom(tableGrob(matrixr, rows=NULL), 
                    xmin=40, xmax=60, ymin=0, ymax=18)
        

plotval6a


```

```{r Figures 4c 4d, fig.height=11, fig.width=11}
#450k artificial mixtures and FACS
# library(minfi)
# library(GEOquery)
# getGEOSuppFiles("GSE77797")
# untar("GSE77797/GSE77797_RAW.tar", exdir = "GSE77797/idat")
# head(list.files("GSE77797/idat", pattern = "idat"))
# 
# idatFiles <- list.files("GSE77797/idat", pattern = "idat.gz$", full = TRUE)
# sapply(idatFiles, gunzip, overwrite = TRUE)
# rgSet <- read.metharray.exp("GSE77797/idat")
# rgSet
# 
# 
# geoMat <- getGEO("GSE77797")
# pD.all <- pData(geoMat[[1]])
# pD <- pD.all#[, c("title", "geo_accession", "characteristics_ch1.1", "characteristics_ch1.2")]
# head(pD)
# 
# 
# #names(pD)[c(3,4)] <- c("group", "sex")
# pD$CD4T <- pD$`cd4+ t cell (%):ch1`
# pD$CD8T <- pD$`cd8+ t cell (%):ch1`
# pD$NK<-pD$`natural killer cell (%):ch1`
# pD$Gran<-pD$`granulocyte (%):ch1`
# pD$Bcell<-pD$`b cell (%):ch1`
# pD$Mono<-pD$`monocyte (%):ch1`
# table(pD$source_name_ch1)
# 
# pD$CellType<-ifelse(pD$source_name_ch1=="Adult human whole blood", "WBC", "MIX")
# 
# #sampleNames(rgSet) <- sub(".*_5", "5", sampleNames(rgSet))
# rownames(pD) <- substr(as.character(pD$description),2,18)
# colnames(rgSet)<- substr(as.character(colnames(rgSet)),12,28)
# pD <- pD[sampleNames(rgSet),]
# identical(rownames(pD), colnames(rgSet))
# pData(rgSet) <- DataFrame(pD, row.names = rownames(pD))
# rgSet
# GSE77797<-rgSet
# 
# library(sesame)
# 
# GSE77797sesame<-sesamize(GSE77797)

load("I:/Dropbox (Christensen Lab)/Extended library/Raw_Data/Raw_Data/GSE77797.rda")

EPICmix<-GSE77797[,GSE77797$CellType%in%c("WBC", "MIX")]
table(EPICmix$CellType)

EPICpheno<-as.data.frame(pData(EPICmix))

identical(colnames(EPICmix), rownames(EPICpheno))
EPICmixsesame<-preprocessNoob(GSE77797)#GSE77797sesame
EPICmixsesame<-EPICmixsesame[,order(colnames(EPICmixsesame))]
EPICpheno<-EPICpheno[order(rownames(EPICpheno)),]

identical(colnames(EPICmixsesame), rownames(EPICpheno))
for (i in c("CD4T", "CD8T", "Mono", "Bcell", "NK", "Gran")){
  EPICpheno[,i]<-as.numeric(EPICpheno[,i])
}

#Only WBC
stack<-as.data.frame(EPICpheno[EPICpheno$CellType=="WBC",c("CD4T", "CD8T", "Mono", "Bcell", "NK", "Gran")])

cellcountsval<-projectCellType_CP(getBeta(EPICmixsesame)[IDOLOptimizedCpGsBloodExtended450k,EPICpheno$CellType=="WBC"], FlowSorted.BloodExtended.450klegacy.compTable, lessThanOne = TRUE)
boxplot(cellcountsval)
cellcountsval<-round(cellcountsval*100,2)
cellcountsval<-as.data.frame(cellcountsval)
cellcountsval$CD8T<-cellcountsval$CD8mem+cellcountsval$CD8nv
cellcountsval$CD4T<-cellcountsval$CD4mem+cellcountsval$CD4nv+cellcountsval$Treg
cellcountsval$Bcell<-cellcountsval$Bnv+cellcountsval$Bmem
cellcountsval$Gran<-cellcountsval$Neu+cellcountsval$Eos+cellcountsval$Bas

cellTypes<-c("CD4T", "CD8T", "Mono", "Bcell", "NK", "Gran" )
cellcolors<-c( Bcell="#1B9E77", 
                   CD4T="#D95F02", CD8T ="#7570B3",
           Gran="#66A61E", Mono="#E7298A",NK="#E6AB02")
truevalues<-stack[complete.cases(stack),]
truevalues<-round(truevalues,1)
cellcount_EPICIDOL<-cellcountsval[rownames(truevalues),]

K = length(cellTypes)  # number of cell types

matrixr<-data.frame(Celltypes=cellTypes, R2=rep(NA, length(cellTypes)), RMSE= rep(NA, length(cellTypes)), row.names=cellTypes)

par(mfrow=c(3,3))
for(k in 1:K) {
	pred = as.numeric(cellcount_EPICIDOL[,cellTypes[k]])
	obs = as.numeric(truevalues[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))
	matrixr$R2[k]<-round(R2,2)
	matrixr$RMSE[k]<-round(RMSE,2)

	rangexy = range(c(pred,obs))

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
	      cex.main = 2)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}

cellTypes<-c("CD4T", "CD8T", "Mono", "Bcell", "NK", "Gran" )
cellcolors<-c( Bas= "#E41A1C", Bcell="#1B9E77", CD4mem="darkblue", 
                   CD4nv="#D95F02", CD8nv ="#7570B3", CD8mem="maroon",  
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02", 
           Treg="black")

truevalues<-truevalues
cellcount_EPICIDOL<-cellcount_EPICIDOL
identical(rownames(truevalues), rownames(cellcount_EPICIDOL))



library(ggplot2)
library(scatterpie)
library(reshape2)
library(gridExtra)


stack2<-t(round(stack[complete.cases(stack),cellTypes],1))
dat2 <- melt(stack2)#, id.vars = "Sample")
dat2$X1<-as.character(dat2$X1)
dat2<-dat2[order(as.character(dat2$X2)),]
dat2<-dat2[order(as.character(dat2$X1)),]

stack3<-t(cellcount_EPICIDOL[,colnames(cellcount_EPICIDOL)%in%cellTypes])
dat3 <- melt(stack3)
dat3$X1<-as.character(dat3$X1)
dat3<-dat3[order(as.character(dat3$X2)),]
dat3<-dat3[order(as.character(dat3$X1)),]

identical(dat2$X2, dat3$X2)
identical(as.character(dat2$X1), as.character(dat3$X1))

dat2$obs<-dat2$value
dat2$pred<-dat3$value

groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
Bnv = "#1B9E77", Bmem = "magenta", 
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black",  
CD8nv = "#7570B3", CD8mem = "maroon", NK = "#E6AB02")

for (i in names(groups))
dat2[,i]<-0

for(i in c( "Mono",  "NK" ))
  dat2[dat2$X1==i,i]<-cellcount_EPICIDOL[,i]
dat2[dat2$X1=="CD4T",c("CD4nv","CD4mem", "Treg")]<-cellcount_EPICIDOL[,c("CD4nv","CD4mem", "Treg")]
dat2[dat2$X1=="CD8T",c("CD8nv","CD8mem")]<-cellcount_EPICIDOL[,c("CD8nv","CD8mem")]
dat2[dat2$X1=="Bcell",c("Bnv","Bmem")]<-cellcount_EPICIDOL[,c("Bnv","Bmem")]
dat2[dat2$X1=="Gran",c( "Bas","Eos","Neu")]<-cellcount_EPICIDOL[,c( "Bas","Eos","Neu")]

dput(colnames(dat2))

plotval7<-ggplot() + 
  geom_scatterpie(aes(x=obs, y=pred, r=2), data=dat2, cols=c("Bas", 
                                                               "Bnv", "Bmem", 
                                                               "CD4nv", "CD4mem",
                                                               "CD8nv", "CD8mem",
                                                               "Eos", "Mono", "Neu", 
                                                               "NK", "Treg"))+
  xlab("Observed (%)") +#xlab("\nSample") +
        ylab("Predicted (%)") +#ylab("Cell %\n") + 
  geom_smooth(method = "lm", se = FALSE)+
        geom_abline(intercept=0, slope=1, linetype="dashed", color = "black")+ 
        theme_classic(base_size = 20)+ 
        ggtitle("Adult peripheral blood with FACS (450k)") + scale_fill_manual(values=groups)+
        theme(legend.position = "none")+
  annotation_custom(tableGrob(matrixr, rows=NULL), 
                    xmin=40, xmax=60, ymin=0, ymax=18)
        

plotval7

#Artificial mixtures 450k

#Only WBC
stack<-as.data.frame(EPICpheno[EPICpheno$CellType=="MIX",c("CD4T", "CD8T", "Mono", "Bcell", "NK", "Gran")])

cellcountsval<-projectCellType_CP(getBeta(EPICmixsesame)[IDOLOptimizedCpGsBloodExtended450k,EPICpheno$CellType=="MIX"], FlowSorted.BloodExtended.450klegacy.compTable, lessThanOne = TRUE)
boxplot(cellcountsval)
cellcountsval<-round(cellcountsval*100,2)
cellcountsval<-as.data.frame(cellcountsval)
cellcountsval$CD8T<-cellcountsval$CD8mem+cellcountsval$CD8nv
cellcountsval$CD4T<-cellcountsval$CD4mem+cellcountsval$CD4nv+cellcountsval$Treg
cellcountsval$Bcell<-cellcountsval$Bnv+cellcountsval$Bmem
cellcountsval$Gran<-cellcountsval$Neu+cellcountsval$Eos+cellcountsval$Bas

cellTypes<-c("CD4T", "CD8T", "Mono", "Bcell", "NK", "Gran" )
cellcolors<-c( Bcell="#1B9E77", 
                   CD4T="#D95F02", CD8T ="#7570B3",
           Gran="#66A61E", Mono="#E7298A",NK="#E6AB02")
truevalues<-stack[complete.cases(stack),]
truevalues<-round(truevalues,1)
cellcount_EPICIDOL<-cellcountsval[rownames(truevalues),]

K = length(cellTypes)  # number of cell types

matrixr<-data.frame(Celltypes=cellTypes, R2=rep(NA, length(cellTypes)), RMSE= rep(NA, length(cellTypes)), row.names=cellTypes)

par(mfrow=c(3,3))
for(k in 1:K) {
	pred = as.numeric(cellcount_EPICIDOL[,cellTypes[k]])
	obs = as.numeric(truevalues[,cellTypes[k]])
	R2 = cor(obs, pred, method = "pearson")^2
	RMSE = sqrt(sum((pred - obs)^2/length(pred)))
	matrixr$R2[k]<-round(R2,2)
	matrixr$RMSE[k]<-round(RMSE,2)

	rangexy = range(c(pred,obs))

	par(mar = c(5,5,4,2))
	plot(obs, pred, cex = 2, col = cellcolors[cellTypes[k]], pch = 19, 
		xlim = rangexy, ylim = rangexy, main = "",
		xlab = "Observed (%)", ylab = "Predicted (%)",
		cex.main = 2, cex.lab = 1.5, cex.axis = 1.3)
	title(main = paste(cellTypes[k], "\nRMSE = ", 
		  round(RMSE, 2), "; R2 = ", round(R2, 2), sep = ""), 
	      cex.main = 2)
	abline(a = 0, b = 1, lwd = 2, lty = "dotted")
}

cellTypes<-c("CD4T", "CD8T", "Mono", "Bcell", "NK", "Gran" )
cellcolors<-c( Bas= "#E41A1C", Bcell="#1B9E77", CD4mem="darkblue", 
                   CD4nv="#D95F02", CD8nv ="#7570B3", CD8mem="maroon",  
           Eos="#3E8E93",
           Neu="#66A61E", Mono="#E7298A",NK="#E6AB02", 
           Treg="black")

truevalues<-truevalues
cellcount_EPICIDOL<-cellcount_EPICIDOL
identical(rownames(truevalues), rownames(cellcount_EPICIDOL))



library(ggplot2)
library(scatterpie)
library(reshape2)
library(gridExtra)


stack2<-t(round(stack[complete.cases(stack),cellTypes],1))
dat2 <- melt(stack2)#, id.vars = "Sample")
dat2$X1<-as.character(dat2$X1)
dat2<-dat2[order(as.character(dat2$X2)),]
dat2<-dat2[order(as.character(dat2$X1)),]

stack3<-t(cellcount_EPICIDOL[,colnames(cellcount_EPICIDOL)%in%cellTypes])
dat3 <- melt(stack3)
dat3$X1<-as.character(dat3$X1)
dat3<-dat3[order(as.character(dat3$X2)),]
dat3<-dat3[order(as.character(dat3$X1)),]

identical(dat2$X2, dat3$X2)
identical(as.character(dat2$X1), as.character(dat3$X1))

dat2$obs<-dat2$value
dat2$pred<-dat3$value

groups<-c(Bas = "#E41A1C", Eos = "#3E8E93", Neu = "#66A61E", Mono = "#E7298A", 
Bnv = "#1B9E77", Bmem = "magenta",  
CD4nv = "#D95F02", CD4mem = "darkblue", Treg = "black", 
CD8nv = "#7570B3", CD8mem = "maroon", NK = "#E6AB02")

for (i in names(groups))
dat2[,i]<-0

for(i in c( "Mono",  "NK" ))
  dat2[dat2$X1==i,i]<-cellcount_EPICIDOL[,i]
dat2[dat2$X1=="CD4T",c("CD4nv","CD4mem", "Treg")]<-cellcount_EPICIDOL[,c("CD4nv","CD4mem", "Treg")]
dat2[dat2$X1=="CD8T",c("CD8nv","CD8mem")]<-cellcount_EPICIDOL[,c("CD8nv","CD8mem")]
dat2[dat2$X1=="Bcell",c("Bnv","Bmem")]<-cellcount_EPICIDOL[,c("Bnv","Bmem")]
dat2[dat2$X1=="Gran",c( "Bas","Eos","Neu")]<-cellcount_EPICIDOL[,c( "Bas","Eos","Neu")]

dput(colnames(dat2))

plotval7a<-ggplot() + 
  geom_scatterpie(aes(x=obs, y=pred, r=2), data=dat2, cols=c("Bas", 
                                                               "Bnv", "Bmem", 
                                                               "CD4nv", "CD4mem",
                                                               "CD8nv", "CD8mem",
                                                               "Eos", "Mono", "Neu", 
                                                               "NK", "Treg"))+
  xlab("Observed (%)") +#xlab("\nSample") +
        ylab("Predicted (%)") +#ylab("Cell %\n") + 
  geom_smooth(method = "lm", se = FALSE)+
        geom_abline(intercept=0, slope=1, linetype="dashed", color = "black")+ 
        theme_classic(base_size = 20)+ 
        ggtitle("Artificial mixtures isolated blood cells adults (450k)") + scale_fill_manual(values=groups)+
        theme(legend.position = "none")+
  annotation_custom(tableGrob(matrixr, rows=NULL), 
                    xmin=40, xmax=60, ymin=0, ymax=18)
        

plotval7a



```


```{r}

library(readxl)
Extended_library_1200_EPIC_fix <- read_excel("Figures tables/Extended_library_1200_EPIC_fix.xlsx", 
    sheet = "Extended_library_1200_EPIC_fix")
Extended_library_1500_450k_fix <- read_excel("Figures tables/Extended_library_1200_EPIC_fix.xlsx", 
    sheet = "Extended_library_1500_450k_fix")
minfi_pickcompprobes_1200 <- read_excel("Figures tables/Extended_library_1200_EPIC_fix.xlsx", 
    sheet = "minfi_pickcompprobes_1200")
table(Extended_library_1200_EPIC_fix$Cell_deconv, Extended_library_1200_EPIC_fix$Relation_to_UCSC_CpG_Island, useNA = "ifany")
table(Extended_library_1500_450k_fix$Cell_deconv, Extended_library_1500_450k_fix$Relation_to_UCSC_CpG_Island, useNA = "ifany")
table(minfi_pickcompprobes_1200$Cell_deconv, minfi_pickcompprobes_1200$Relation_to_UCSC_CpG_Island, useNA = "ifany")

```

```{r}


library(BAGS)
library(missMethyl)

#read C2 or any of the curated DB gmt file downloaded from http://software.broadinstitute.org/gsea/msigdb/genesets.jsp?collection=C2
load("D:/OneDrive/OneDrive - Dartmouth College/Crossreactive/annotationEPICb5updated2.rda")
MSigDB_C7_CP <- ReadGMT("D:/OneDrive/OneDrive - Dartmouth College/GSEA/c7.all.v7.2.entrez.gmt")

#perform gsameth
dput(levels(as.factor(Extended_library_1200_EPIC_fix$Cell_deconv)))
for (i in cellTypes){
try(gsa.MSigDB <- gsameth(sig.cpg=Extended_library_1200_EPIC_fix$IlmnID[Extended_library_1200_EPIC_fix$Cell_deconv==i], all.cpg=rownames(Ext_deconv_betas3), collection=MSigDB_C7_CP,array.type = "EPIC", anno = annotDF))
#  Multiple symbols ignored for one or more aliases
#sort by FDR p-value & only keep <0.05
print(i)
try(gsa.MSigDB <- as.data.frame(gsa.MSigDB[order(gsa.MSigDB[,4]),]))
try(print(head(gsa.MSigDB)))
try(sigMSigDB <- gsa.MSigDB[gsa.MSigDB$FDR<0.05,])
try(print(head(sigMSigDB)))
#write.csv(gsa.MSigDB, file=paste0(i,"GSEACP7_IDOL_EPIC.csv"))
}



MSigDB_all_CP <- ReadGMT("D:/OneDrive/OneDrive - Dartmouth College/GSEA/msigdb.v7.2.entrez.gmt")
for (i in cellTypes){
try(gsa.MSigDB <- gsameth(sig.cpg=Extended_library_1200_EPIC_fix$IlmnID[Extended_library_1200_EPIC_fix$Cell_deconv==i], all.cpg=rownames(Ext_deconv_betas3), collection=MSigDB_all_CP,array.type = "EPIC", anno = annotDF))
#  Multiple symbols ignored for one or more aliases
#sort by FDR p-value & only keep <0.05
print(i)
try(gsa.MSigDB <- as.data.frame(gsa.MSigDB[order(gsa.MSigDB[,4]),]))
try(print(head(gsa.MSigDB)))
try(sigMSigDB <- gsa.MSigDB[gsa.MSigDB$FDR<0.05,])
try(print(head(sigMSigDB)))
#write.csv(gsa.MSigDB, file=paste0(i,"GSEAall_IDOL_EPIC.csv"))
}
#perform gsameth for the single-cell database (later known as curated 8)
MSigDB_sc<- ReadGMT("D:/OneDrive/OneDrive - Dartmouth College/GSEA/scsig.all.v1.0.1.entrez.gmt")
gsa.MSigDBsc <- gsameth(sig.cpg=Extended_library_1200_EPIC_fix$IlmnID, all.cpg=rownames(Ext_deconv_betas3), collection=MSigDB_sc)
#  Multiple symbols ignored for one or more aliases
#sort by FDR p-value & only keep <0.05
gsa.MSigDBsc <- as.data.frame(gsa.MSigDBsc[order(gsa.MSigDBsc[,4]),])
sigMSigDBsc <- gsa.MSigDBsc[gsa.MSigDBsc$FDR<0.05,]
head(sigMSigDBsc)
#write.csv(gsa.MSigDBsc, file="GSEAsc_ 150 filtered_BN2.csv")


```

```{r}
sessionInfo()
```
